---
layout: page
title: genome-similarity [artifact]
categories: [anvio]
comments: false
image:
  featurerelative: ../../../images/header.png
  display: true
---


{% include _toc.html %}


<img src="../../images/icons/CONCEPT.png" alt="CONCEPT" style="width:100px; border:none" />

A CONCEPT-type anvi'o artifact. This artifact is typically generated, used, and/or exported **by anvi'o** (and not provided by the user)..

Back to the **[main page](../../)** of anvi'o programs and artifacts.

## Provided by


<p style="text-align: left" markdown="1"><span class="artifact-p">[anvi-compute-genome-similarity](../../programs/anvi-compute-genome-similarity)</span> <span class="artifact-p">[anvi-script-compute-ani-for-fasta](../../programs/anvi-script-compute-ani-for-fasta)</span></p>


## Required or used by


<p style="text-align: left" markdown="1"><span class="artifact-r">[anvi-dereplicate-genomes](../../programs/anvi-dereplicate-genomes)</span></p>


## Description

This is the output of <span class="artifact-n">[anvi-compute-genome-similarity](/software/anvio/help/7/programs/anvi-compute-genome-similarity)</span> (which describes the level of similarity between all of the input genomes) or <span class="artifact-n">[anvi-script-compute-ani-for-fasta](/software/anvio/help/7/programs/anvi-script-compute-ani-for-fasta)</span> (which describes the level of similarity between contigs in a fasta file). 

{:.notice}
The output of <span class="artifact-n">[anvi-compute-genome-similarity](/software/anvio/help/7/programs/anvi-compute-genome-similarity)</span> will only be in this structure if you did not input a <span class="artifact-n">[pan-db](/software/anvio/help/7/artifacts/pan-db)</span>. Otherwise, the data will be put directly into the additional data tables of the <span class="artifact-n">[pan-db](/software/anvio/help/7/artifacts/pan-db)</span>. The same is true of <span class="artifact-n">[anvi-script-compute-ani-for-fasta](/software/anvio/help/7/programs/anvi-script-compute-ani-for-fasta)</span>. 

This is a directory (named by the user) that contains both a <span class="artifact-n">[dendrogram](/software/anvio/help/7/artifacts/dendrogram)</span> (NEWICK-tree) and a matrix of the similarity scores between each pair for a variety of metrics dependent on the program that you used to run <span class="artifact-n">[anvi-compute-genome-similarity](/software/anvio/help/7/programs/anvi-compute-genome-similarity)</span> or <span class="artifact-n">[anvi-script-compute-ani-for-fasta](/software/anvio/help/7/programs/anvi-script-compute-ani-for-fasta)</span> .

For example, if you used `pyANI`'s `ANIb` (the default program), the output directory will contain the following twelve files. These are directly created from the heatmaps generated by PyANI, just converted into matrices and newick files: 

-`ANIb_alignment_coverage.newick` and `ANIb_alignment_coverage.txt`: contains the percent coverage (for query and subject)

-`ANIb_percentage_identity.newick` and `ANIb_percentage_identity.txt`: contains the percent identity

-`ANIb_full_percentage_identity.newick` and `ANIb_full_percentage_identity.txt`: contains the percent identity in the context of the length of the entire query and subject sequences (not just the aligned segment)

-`ANIb_alignment_lengths.newick` and `ANIb_alignment_lengths.txt`: contians the total aligned lengths 

-`ANIb_similarity_errors.newick` and `ANIb_similarity_errors.txt`: contains similarity errors (total number of mismatches, not including indels)

-`ANIb_hadamard.newick` and `ANIb_hadamard.txt`: contians the hadamard matrix (dot product of identity and coverage matrices)



{:.notice}
Edit [this file](https://github.com/merenlab/anvio/tree/master/anvio/docs/artifacts/genome-similarity.md) to update this information.

