---
layout: post
title: "Going Deeper with anvi'o"
excerpt: "Anvi'o makes it easier for pro-users to access raw information."
modified: 2015-05-02
tags: []
categories: [anvio]
comments: true
---

{% include _toc.html %}

This ever-growing article demonstrates ways to use anvi'o for different purposes. If you can't get anvi'o do something for you, please post a comment down below, or open an <a href="https://github.com/meren/anvio/issues">issue</a>.

One of the goals we had in mind when we designed anvi'o was to create a platform that serves end-users as well as it serves to pro-users with more computational skills. The current design of anvi'o does it both. While there are some best practices we define for end-users, we do not impose anything to experts (on the contrary we hope to learn from them).

If you are a bioinformatician, you can see anvi'o simply as an abstraction layer between you and your data that saves you from the boring details of curating data into a structured form, so you can implement your ideas using anvi'o as a resource, or as a mere visualization tool.

For instance, these commands will work on any annotation database generated by anvi'o:

    $ sql='select length,gc_content from contigs_basic_info where length > 10000 limit 20000'
    $ sqlite3 annotation.db "$sql" | sed 's/|/,/g' > i.csv
    $ echo "set datafile separator \",\"; set logscale y; plot '-' using 2:1;" | cat - i.csv | gnuplot -p

and will show a scatter plot where the X-axis is the GC-content, and the Y-axis is the length for the first 20,000 contigs that are longer than 10,000 nucleotides in a given annotation database:

<a href="{{ site.url }}/images/anvio/tutorials/gnuplot-gc-content.png"><img src="{{ site.url }}/images/anvio/tutorials/gnuplot-gc-content.png" width="50%" /></a>


# Examples


This article will serve as a collection of different ways to query the data and use anvi'o for _ad hoc_ purposes to give inspiration anyone who may be interested.

## Exporting tables from databases

In an ideal world, you should never need to do this (because anvi'o has all the classes to access and work with its standard databases from within Python scripts). Howevever, you can export the information stored in any anvio database/table using `anvi-get-db-table-as-matrix` program.

You can see what tables are available in a given anvio dataase,

<pre>
$ anvi-get-db-table-as-matrix ANNOTATION.db --list
self
contig_sequences
kmer_contigs
kmer_splits
contigs_basic_info
splits_basic_info
hmm_hits_info
hmm_hits_in_splits
hmm_hits_in_contigs
genes_in_contigs
genes_in_splits_summary
genes_in_splits
collections_info
collections_colors
collections_of_contigs
collections_of_splits
</pre>

And export any one of them:

<pre>
$ anvi-get-db-table-as-matrix ANNOTATION.db -t hmm_hits_in_contigs
Database .....................................: "ANNOTATION.db" has been initiated with its 16 tables.
Table ........................................: "hmm_hits_in_contigs" has been read with 2 entries and 8 columns.
Output .......................................: hmm_hits_in_contigs.txt
</pre>

You can also export only certain fileds from a table. To see what fields are available, use `--list` with a table:

<pre>
$ anvi-get-db-table-as-matrix ANNOTATION.db -t hmm_hits_in_contigs --list
Database .....................................: "ANNOTATION.db" has been initiated with its 16 tables.
Table ........................................: "hmm_hits_in_contigs" has been read with 2 entries and 8 columns.
Table columns ................................: "entry_id, source, contig, start, stop, gene_name, gene_id, e_value"
</pre>

Then you can specify which columns you would like to be exported from the table of interest like this:

<pre>
$ anvi-get-db-table-as-matrix ANNOTATION.db -t hmm_hits_in_contigs --fields "contig, gene_name"
Database .....................................: "ANNOTATION.db" has been initiated with its 16 tables.
Table ........................................: "hmm_hits_in_contigs" has been read with 2 entries and 8 columns.
Columns to report ............................: "contig, gene_name"
Output .......................................: hmm_hits_in_contigs.txt
</pre>

This time `hmm_hits_in_contigs.txt` will contain only two of the columns you specified. 


## Running the visualization interface with manual input

You can use anvi'o to visualize almost any type of data, as `anvi-interactive` can also be run using text files alone without requiring any annotation or profile databases.

Here is an example command line that would achieve that:

    anvi-interactive -f FASTA.fa -m METADATA.txt -t TREE.txt -o OUTPUT_DIRECTORY

In the example above, `FASTA.fa` is a FASTA formatted file for your contigs, `METADATA.txt` is a TAB-delimited file that contains information for each entry appears in the FASTA file, and `TREE.txt` is the file where you describe the organization of your entries you have in your FASTA file. For this to work, the names of the leaves of the tree, the deflines found in your FASTA file and the first column of your metadata file must match, perfectly (and anvi'o will complain about it if they don't).

Here is an example FASTA file:

<pre>
>A
ATCG
>B
GTCA
>C
GACT
>D
TACG
>E
TTCG
>F
GGCA
>G
ACAC
>H
CATC
</pre>

And the corresponding example metadata file:


| contig    | C1    | C2            |
|--------   |------ |-----------    |
| A         | 0.5   | Cluster_1     |
| B         | 0.5   | Cluster_1     |
| C         | 1.2   | Cluster_0     |
| D         | 0.6   | Cluster_3     |
| E         | 0.7   | Cluster_3     |
| F         | 0.8   | Cluster_3     |
| G         | 0.9   | Cluster_4     |
| H         | 0.9   | Cluster_4     |


And finally an example newick-formatted tree that would work with the previous two:

    ((G:0.0122349,H:0.0122349)Int11:0.450561,((C:0.0103605,D:0.0103605)Int10:0.0791402,(E:0.0262071,(F:0.00364725,(A:0,B:0)Int8:0.00364725)Int9:0.0262071)Int12:0.0791402)Int13:0.450561);

If you put these examples into cognate files, the previous `anvi-interactive` will work, and welcome you with the interactive interface.

Files for a bit more realistic looking example is in the [`tests/manual_interactive`](https://github.com/meren/anvio/tree/master/tests/manual_interactive) directory in the codebase. Also, `run_manual_interactive.sh` in the `tests` directory demonstrates how those files should be run.


## Generating a newick tree file from an input matrix

Please see example files stored in the [`tests/manual_interactive`](https://github.com/meren/anvio/tree/master/tests/manual_interactive) directory. If you have a matrix file (such as [`data.txt`](https://github.com/meren/anvio/blob/master/tests/manual_interactive/data.txt)), you can get a newick file for it by running this anvi'o command:

    anvi-matrix-to-newick data.txt -o tree.txt


