---
layout: post
authors: [matt]
title: "Applications of the Agnostos Workflow in Genomes and Metagenomes with anvi'o"
excerpt: "Explore the unknown!"
modified: 2020-07-17
tags: []
categories: [anvio]
comments: true
redirect_from:
  - /export-loci/
image:
    feature: 
---

{% include _toc.html %}

How many times have you analyzed a metagenome and found that a significant proportion of the genes have no annotation? 

Have you ever been interested in a genomic locus but it contains critical gene that has no determined function?

The Agnostos workflow ([Vanni et al., 2020](https://www.biorxiv.org/content/10.1101/2020.06.30.180448v1)) is a great way to start to address the problems above.

Recently, Vanni et al., 2020 published the preprint [Light into the darkness: Unifying the known and unknown coding sequence space in microbiome analyses](https://www.biorxiv.org/content/10.1101/2020.06.30.180448v1). One of the goals of this manuscript was to provide a resource for the community to explore genes of unknown function in microbiomes. To create this dataset, briefly, the Agnostos workflow clustered open reading frames (ORFS) from 1,749 metagenomes and 28,941 bacterial and archaeal genomes, then systematically validated and categorized the clusters into Known and Unknown gene clusters. 

To accompany this manuscript, my former supervisor [Antonio Fernandez-Guerra](http://orcid.org/0000-0002-8679-490X) wrote a blog post that goes over the origins of the Agnostos worflow, an overview of the categories of gene-clusters resulting from the workflow, and the future of Unknown genes in microbiomes. Check it out here: [blog post](http://merenlab.org/2020/07/01/dark-side/#a-conceptual-framework-to-unify-the-known-and-the-unknown-in-microbiome-analyses).

"Ok great, we have this fancy resource... but how do I use it? I have metagenomes I'm analyzing right now!?"
 
This tutorial will demonstrate how to use the results from integrating ORFs from genomes and metagenomes into the Agnostos workflow from [Vanni et al., 2020](https://www.biorxiv.org/content/10.1101/2020.06.30.180448v1) using anvio! 

The first step for utilizing this workflow is to annotated your open reading frames (ORFs), genomes or metagenomes, with the Agnostos categories. The best practice method to do this is to integrate your ORFs into the existing Agnostos clusters. In other words, the workflow will attempt to insert a new ORF into an existing cluster or create new clusters ORFs from your sample. You can find the workflow [here](https://github.com/functional-dark-side/agnostos-wf).

To have some fun, [Chiara Vanni](https://orcid.org/0000-0002-1124-1147) kindly integrated all of the ORFs from the classic anvio metagenomic sandbox, the Infant Gut Dataset from [Sharon et al., 2013](http://www.ncbi.nlm.nih.gov/pubmed/22936250)(IGD). We will explore how the Agnostos categories can be used to get more out of both the IGD metagenomes and pangenomes of *E. faecalis*. The tutorial below is not an exhaustive description of how to use the Agnostos workflow, but merely a brief introduction.

# Download the goods 

First let's make a directory where we will work with the IGD Agnostos integrated data and download all the files we will need.
```bash
mkdir IGD_agnostos && cd IGD_agnostos/

wget https://ndownloader.figshare.com/files/18046139 -O INFANT-GUT-TUTORIAL.tar.gz && tar -zxvf INFANT-GUT-TUTORIAL.tar.gz

# Need to make a figshare for the integrated IGD data into agnostos clusters
# Download Chiaras integrated results from here: https://drive.google.com/file/d/1uhyoWVcABZlL0XkelPpBQHT2QIcQHlTO/view?usp=sharing
# needs to contain both of these files: IGD_genes_summary_info_red.tsv & IGD_genes_summary_info_exp.tsv
# and external genomes

# check if niche breadth is joined

# also need ext_genomes_gene_info.tsv for pangenome
```

# Metagenomics applications

Now let's get the Unknowns party started!

Metagenomic sequencing allows access to the uncultured majority of microbiomes all kinds of biomes including soil, ocean, to the human gut. These metagenomic samples are a huge source of genetic novelty. Here we will start out by analyzing the metagenomes from IGD after they have been integrated into the Agnostos worflow. The concept of ORF integration is key here because we are not passing annotations to the IGD ORFs through alignment based approaches (e.g. BLAST, hmmsearch), but are attempting to insert the IGD ORFs into the extant Agnostos clusters that already have metadata associated with them. Integration allows for the entire dataset of ORFs to be used rather than alignment based methods where one filter for best hits. Additionally, when IGD ORFs do no find a home in an Agnostos cluster, the worflow attempts to form new clusters (we will explore this in a second)!

There are many ways to analyze metagenomic data. First, we will explore the IGD ORF integration results of the Agnostos workflow in a read-based approach, then later using genome resolved metagenomics. 

## Read-based approach

Now let's explore the agnostos integrated IGD data. The first step will be to import the data as "functions" into the IGD [contigs DB](http://merenlab.org/software/anvio/help/artifacts/contigs-db/). Please check out this [post](http://merenlab.org/2016/06/18/importing-functions/) if you have more questions about anvio functions tables. This will make it easy to explore the Agnostos categories in the context of metagenomic assembled contigs! To accomplish this, we will clean up the Agnostos data in R to make it an anvio friendly functions table. 

### How much is Unknown?

Get the agnotos integrated IGD data into R
```r
library(tidyverse)

# Angostos categories
IGD_genes_summary_info_exp <- read_tsv("IGD_genes_summary_info_exp_new.tsv")

# Niche breadth data
IGD_niche_breadth <- read_tsv("IGD_genes_cluster_niche_breadth.tsv")

# Convert data into anvio friendly functions table format
IGD_gene_functions <- IGD_genes_summary_info_exp %>%
  mutate("source" = "agnostos_categories") %>%
  mutate(e_value = 0) %>% # just a filler value
  rename(accession = cl_name, "function" = category) %>%
  select(gene_callers_id, source, accession, `function`, e_value)

IGD_gene_sources <- IGD_genes_summary_info_exp %>%
  mutate("source" = "agnostos_source") %>%
  mutate(e_value = 0) %>% # just a filler value
  rename(accession = cl_name, "function" = db) %>%
  select(gene_callers_id, source, accession, `function`, e_value)
  
# Export table
write_tsv(IGD_gene_functions, path = "IGD_gene_functions.tsv", col_names = TRUE)
write_tsv(IGD_gene_sources, path = "IGD_gene_sources.tsv", col_names = TRUE)
```


Now that we have an anvio friendly functions table, we can import it into the IGD contigs DB. Next, the best part, we can export the data in the context of metagenomic contigs! This is a powerful method because we can look for contigs that are enriched in Genomic Unknowns and leveraged the genomic neighborhood to infer functionality. 

```bash
# Import functions into metagenome
anvi-import-functions -c INFANT-GUT-TUTORIAL/CONTIGS.db -i IGD_gene_functions.tsv
anvi-import-functions -c INFANT-GUT-TUTORIAL/CONTIGS.db -i IGD_gene_sources.tsv

# Make a collection of all contigs called EVERYTHING
anvi-script-add-default-collection -p INFANT-GUT-TUTORIAL/PROFILE.db \
                                   -c INFANT-GUT-TUTORIAL/CONTIGS.db \
                                   -C EVERYTHING

# Retrieve a ton of awesome tables for fun later
anvi-summarize -c INFANT-GUT-TUTORIAL/CONTIGS.db \
               -p INFANT-GUT-TUTORIAL/PROFILE.db \
               -C EVERYTHING \
               --report-aa-seqs-for-gene-calls \
               -o SUMMARY-EVERYTHING
```

Now we can get the super awesome `EVERYTHING-gene_calls.txt` table! Let's get it back into R and answer some basic questions.

A first question that the Agnostos categories quickly answers is the proportions of Known (K + KWP), Unknown (GU + EU), Singletons (SINGL), and Discarded (DISC) ORFs that are in your dataset. Here we are aggregating some of the Agnostos categories to get a first look at the data.
```r
# Import gene-calls data
all_gene_calls <- read_tsv("SUMMARY-EVERYTHING/bin_by_bin/EVERYTHING/EVERYTHING-gene_calls.txt")

# Aggregate Agnostos categories
all_gene_calls_aggregated <- all_gene_calls %>%
  mutate(agnostos_categories_aggregated  = case_when(grepl("GU", agnostos_categories) ~ 'unknown',
                                                     grepl("EU", agnostos_categories) ~ 'unknown',
                                                     grepl("K", agnostos_categories) ~ 'known',
                                                     grepl("KWP", agnostos_categories) ~ 'known',
                                                     grepl("DISC", agnostos_categories) ~ 'DISC',
                                                     grepl("SINGL", agnostos_categories) ~ 'SINGL',
                                                     is.na(agnostos_categories) ~ 'DISC')) 

library(kableExtra)

all_gene_calls_aggregated %>%
  count(agnostos_categories_aggregated) %>%
  mutate(total_ORFs = sum(n)) %>%
  mutate(Proportion = n/total_ORFs) %>%
  arrange(desc(Proportion)) %>%
  select(agnostos_categories_aggregated, Proportion) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"))
```

[![agnostos_cat_prop_metagenome](/images/agnostos_cat_prop_metagenome.jpeg)](images/agnostos_cat_prop_metagenome.jpeg){:.center-img .width-100}

Let's discuss these proportions.

Around 51.5% of the ORFs from IGD were classified as Known Agnostos categories. In other words, the Agnostos workflow was able to link these ORFs to previously validated functions and taxa. This means that either an ORF was successfully integrated into an existing Known cluster or the ORF of interest formed a new cluster with other IGD ORFs and was validated and classified as Known (novel clusters go through the same validation and classification steps that existing agnostos clusters when through). The infant gut is an relatively explored environment where we would expect to find microbes with cultured representatives. That fact that over 50% of the ORFs are Known makes sense in this biome. If this were to be a less sampled environment, say a hydrothermal vent, the proportion of Known ORFs could potentially be much lower. 

The next category of ORFs we should discuss are the Singletons (30.1%). The Singleton ORFs are a huge source of sequence diversity. The Agnostos clusters were created from a MASSIVE metagenomics dataset, including TARA Oceans and the Human Microbiome Project. If an ORF is classified as a Singleton that means it very unique in the sequence space and could not be put into context of sequence space of some of the largest metagenomic and genomic surveys available. These ORFs could be of viral origin or novel genes important for niche adaptation! Because they are not associated with any Agnostos clusters, there is no metadata associated with them. One reason why there are so many Singletons in this dataset is because of the presence of a yeast called *C. albicans*. Eukaryotic genes were not used to form the Agnostos dataset and can easily be a large source of Singletons. 

Next let's discuss the Unknowns which make up 9.48% of the ORFs in IGD. These ORFS were linked to Genomics Unknowns (GUs) and Environmental Unknowns (EUs) in the Agnostos dataset. The GU classification refers to genes of unknown function that are found in sequenced genomes while EUs are genes of unknown function found in the environment (via metagenomic sequencing). GUs and EUs are the most exciting Agnostos categories because if they are not dataset specific (more on this later), they come with metadata! (We will explore how to leverage the Agnostos metadata for the Unknowns later in the post.) Without GU and EU categories, genes of unknown function have no context and computational investigation of them basically stops there. Now, you will be able to learn about their lineage specificity and niche breadth!

Finally, 8% of the ORFs were classifed as Discarded. The agnostos workflow contains a stringent cluster validation step. Sequences may have been initially included in existing Agnostos clusters but after validation the cluster no longer met the HQ criteria and thus the new sequences were discarded. Another explanation is that an ORF is integrated into previously DISC clustered. Finally, ORFs could of been removed because of their low quality (i.e. spurious, shadow ORFs). If you are interested in how the Agnostos worflow discards sequences please refer to the [preprint](https://www.biorxiv.org/content/10.1101/2020.06.30.180448v1) for details.

### Exploring IGD specific clusters

As mentioned earlier, another way that ORFs from metagenomes can be classified in the Agnostos workflow is if a set of ORFs from the new data create their own new cluster! In other words, there could be dataset specific clusters. Let's see if there are any clusters unique to IGD.
```r
all_gene_calls %>%
  count(agnostos_source) %>%
  mutate(total_ORFs = sum(n)) %>%
  mutate(Proportion = n/total_ORFs) %>%
  arrange(desc(Proportion)) %>%
  select(agnostos_source, Proportion) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"))
```
[![agnostos_IGD_specific_clusters](/images/agnostos_IGD_specific_clusters.png)](images/agnostos_IGD_specific_clusters.png){:.center-img .width-100}

Whoa! 31% of the ORFs from IGD formed their own clusters! The next question to ask here is, what are the proportion of Agnostos categories in the IGD specific clusters?

```r
all_gene_calls %>%
  filter(agnostos_source == "IG") %>%
  count(agnostos_categories) %>%
  mutate(total_ORFs = sum(n)) %>%
  mutate(Proportion = n/total_ORFs) %>%
  arrange(desc(Proportion)) %>%
  select(agnostos_categories, Proportion)
```
[![agnostos_cat_2](/images/agnostos_cat_2.png)](images/agnostos_cat_2.png){:.center-img .width-100}


~96% of the IGD specific clusters were Singletons. This makes a lot of sense considering the high proportion of IGD specific clusters (31%). Luckily, there are a few IGD specific with agnostos annotations! (e.g. Knowns, Genomics Unknowns, and Knowns without Pfam annotations). Let's save these IGD specfic clusters for later ;)

```r
# IGD specific knowns
IGD_specific_knowns <- all_gene_calls %>%
  filter(agnostos_source == "IG") %>%
  filter(agnostos_categories == "K" | agnostos_categories == "KWP")

# IGD specific knowns
IGD_specific_gus <- all_gene_calls %>%
  filter(agnostos_source == "IG") %>%
  filter(agnostos_categories == "GU")
```

### Exploring Contigs

Now that we have a general idea of the distribution of agnostos categories in IGD, we can leverage genomic context of metagenomic assembled contigs to possibly infer functions of unknown genes by exploring their genomic context. To do this, we will filter for interesting contigs that have informative proportions of agnostos categories as well as contigs with long stretches of GUs.

Get a list of contigs with intersting proportions of agnostos categories
```r
contigs_with_agnostos_categories_prop <- all_gene_calls_aggregated %>%
  select(gene_callers_id, contig, start, stop, direction, agnostos_categories_aggregated, COG_FUNCTION) %>%
  group_by(contig) %>%
  mutate(contig_length = length(contig)) %>%
  group_by(contig, agnostos_categories_aggregated) %>%
  mutate(categ_sum = n()) %>%
  mutate(proportion_category = categ_sum/contig_length) %>%
  ungroup() %>%
  select(contig, contig_length, agnostos_categories_aggregated, proportion_category) %>%
  unique() %>%
  spread(agnostos_categories_aggregated, proportion_category, fill = 0)

# Contigs with that we can visualize with Knowns and a few Unknowns
contig_list <- contigs_with_agnostos_categories_prop %>%
  filter(contig_length > 8 & contig_length < 15) %>%
  filter(unknown > 0) %>%
  filter(DISC < 0.1) %>%
  .$contig
```

Find contigs with long stretches of GUs (need python... help!)
```r
# Look for longest stretch of GUs in a row (not from C_albicans)
# work with contigs > 10 genes
contigs_greater_10 <- contigs_with_agnostos_categories_prop %>%
  filter(contig_length > 10) %>%
  .$contig

test <- all_gene_calls_aggregated %>%
  filter(contig %in% contigs_greater_10) %>%
  select(gene_callers_id, contig, agnostos_categories_aggregated) %>%
  group_split(contig)

last <- test[[1]] %>%
  filter(row_number()==n()) %>%
  .$gene_callers_id %>% as.character()

test[[1]] %>%
  spread(gene_callers_id, agnostos_categories_aggregated) %>%
  unite(name, `0`:last)
```

Visualize potentially interesting contigs using [genoplotR](https://genoplotr.r-forge.r-project.org/)
```r
# Prepare contig data to visualize using genoplotR
contig_vis_info <- all_gene_calls %>%
  mutate(fill = case_when( grepl("GU", agnostos_categories)  ~ 'blue',
                           grepl("EU", agnostos_categories) ~ 'red',
                           grepl("K", agnostos_categories) ~ 'black',
                           grepl("SINGL", agnostos_categories) ~ 'white',
                           grepl("DISC", agnostos_categories) ~ 'white')) %>%
  select(contig, "COG_FUNCTION", start, stop, direction, fill) %>%
  rename(name = "COG_FUNCTION", strand = direction, end = stop) %>%
  mutate(col = "black",
         lty='1',
         lwd='1',
         pch='8',
         cex='1',
         gene_type='arrows') %>%
  mutate(strand = gsub("f", "1", strand)) %>%
  mutate(strand = gsub("r", "-1", strand)) %>%
  mutate(lwd = as.numeric(lwd)) %>%
  mutate(lty = as.numeric(lty)) %>%
  mutate(pch = as.numeric(pch)) %>%
  mutate(cex = as.numeric(cex))

library(genoPlotR)

# function for gene annotation
prep_gene_plot <- function(x){
  mid_pos <- middle(x)             # get position of where you want to put the annotation text over the gene
  annot1 <- annotation(x1=mid_pos, # get names of the genes
                       text=x$name, 
                       rot=30)     # rotate the text
  return(annot1)
}

# function for visualizing
visualize_contig <- function(X, contig_list){
  dna1 <- contig_vis_info %>%
    filter(contig %in% contig_list[X]) %>%
    select(-contig)

  dna2 <- contig_vis_info %>%
    filter(contig %in% contig_list[X]) %>%
    select(-contig)

  dna1_seg <- dna_seg(dna1)
  dna2_seg <- dna_seg(dna2)
  
  dna_list <-  list(dna1_seg,dna2_seg)
  anno_list <- lapply(dna_list[1:2], FUN = prep_gene_plot)

  plot_gene_map(dna_segs=dna_list,annotations = anno_list)
}

# Visualize contig
visualize_contig(X = 5, contig_list)
```

Here is an contig with a GU located right in the middle of a potential secretion operon!

[![agnostos_cat_2](/images/agnostos_contig.png)](images/agnostos_contig.png){:.center-img .width-100}

Let's explore the Agnostos metadata from one of the potentially important GU's from the interesting contig
```{r}
# Get cluster_ID
contig_name <- contig_list[[5]]

# Get gene_callers_ID
interesting_gu <- all_gene_calls %>%
  filter(contig == contig_name & gene_callers_id == "11303") %>%
  .$`agnostos_categories (ACCESSION)`

# Explore Lineage Specificity
IGD_genes_summary_info_exp %>%
  filter(cl_name == interesting_gu)

# Explore niche breadth
IGD_niche_breadth %>%
  filter(cl_name == interesting_gu)
```

Unfortunately the cluster that this particular ORF is associated with is not Lineage Specfic or has a significant ecological signal.

## Genome-resolved metagenomics

Now that we have explored the agnostos clusters on individual reads and contigs, let's leverage them and explore Unknowns in metagenomic assembled genomes!

To do this, we will find out the proportions of agnostos categories per bin and visualize it. (In the near future this will be a plug-and-play tool in anvio called **anvi-show-agnostos-in-bins**)

Grab all the bin information from the Infant Gut Tutorial.
```bash
# Important the pre-binned data, thanks Meren!
anvi-import-collection INFANT-GUT-TUTORIAL/additional-files/collections/merens.txt \
                       --bins-info INFANT-GUT-TUTORIAL/additional-files/collections/merens-info.txt \
                       -p INFANT-GUT-TUTORIAL/PROFILE.db \
                       -c INFANT-GUT-TUTORIAL/CONTIGS.db \
                       -C default

# Export tables with all the info we need
anvi-summarize -p INFANT-GUT-TUTORIAL/PROFILE.db \
               -c INFANT-GUT-TUTORIAL/CONTIGS.db \
               -C default \
               -o SUMMARY-BINS
```

Import Summaries per bin into R
```r
# construct path string
path <- paste0(getwd(), "/SUMMARY-BINS/bin_by_bin/")

# list the files we need
files <- list.files(path)

# Function to import files
load_bins <- function(files){

  name <- files
  name2 <- paste0(name,  "-gene_calls.txt") 

  path_to_gene_calls <- paste0(path, name, "/", name2)

  df <- read_tsv(path_to_gene_calls)

  df  %>%
  mutate(k_unk_category = case_when( grepl("GU", agnostos_categories)  ~ 'UNKNOWN',
                                     grepl("EU", agnostos_categories) ~ 'KNOWN',
                                     grepl("K", agnostos_categories) ~ 'KNOWN',
                                     grepl("SINGL", agnostos_categories) ~ 'SINGL',
                                     grepl("DISC", agnostos_categories) ~ 'DISC',
                                     is.na(agnostos_categories) ~ 'non_integrated')) %>%
  count(k_unk_category) %>%
  mutate(prop = prop.table(n)) %>%
  mutate(bin = name)
}

# Import list of files
bins <- lapply(files[1:13], load_bins)

# Combine into one dataframe
bins_df <- bind_rows(bins)
```

Make layer for metagenome where you show proportions of Agnostos categories per bin in the interactive interface. If you have any questions about adding layers to the anvio interactive interface check out his post [here](http://merenlab.org/2017/12/11/additional-data-tables/).
```r
bin_agnostos_proportions <- bins_df %>%
  group_by(bin) %>%
  mutate(prop = n/sum(n)) %>%
  select(-n) %>%
  spread( k_unk_category, prop, fill = 0) %>%
  rename("Agnostos_categories!DISC" = DISC,
         "Agnostos_categories!SINGL" = SINGL,
         "Agnostos_categories!KNOWN" = KNOWN,
         "Agnostos_categories!UNKNOWN" = UNKNOWN,
         "Agnostos_categories!non_integrated" = non_integrated) 
```

Find which contigs belong to which bin
```bash
anvi-export-table INFANT-GUT-TUTORIAL/PROFILE.db --table collections_of_splits -o collection_of_splits.tsv
```

```r
collection_of_splits <- read_tsv("collection_of_splits.tsv")

split_additional_data <- collection_of_splits %>%
  filter(collection_name == "default") %>%
  left_join(bin_agnostos_proportions %>% rename(bin_name = bin)) %>%
  select(split, `Agnostos_categories!DISC`, `Agnostos_categories!SINGL`, `Agnostos_categories!KNOWN`, `Agnostos_categories!non_integrated`, `Agnostos_categories!UNKNOWN`) %>%
  rename(contig = split) %>% 
  drop_na()

split_additional_data %>%
  write_tsv("split_additional_data.tsv")
```

Add to metagenome profile DB
```bash
anvi-import-misc-data split_additional_data.tsv \
                      -p INFANT-GUT-TUTORIAL/PROFILE.db \
                      -t items

# Visualize
anvi-interactive -p INFANT-GUT-TUTORIAL/PROFILE.db -c INFANT-GUT-TUTORIAL/CONTIGS.db
```
[![agnostos_binning](/images/agnostos_binning.png)](images/agnostos_binning.png){:.center-img .width-100}

Looks like bins with the most coverage tend to have the most Known genes. This maybe because they are from more ecological dominant population of bacteria that have more cultured representatives. Additionally, *E. faecalis* and *P. avidum* have a good amount of Unknown ORFs (GUs + EUs). One could refine these MAGs and explore these genes now!

If visualizing is not providing the quanitative insight to bins you are looking for, you can also filter for bins that are the most Known or Unknown.

Which bins have the most agnostos Unknowns?
```r
# most unknowns
bins_df %>% 
  filter(k_unk_category == "UNKNOWN") %>%
  arrange(desc(prop)) 
```

Which bins have the most agnostos Knowns?
```r
# most knowns
bins_df %>% 
  filter(k_unk_category == "KNOWN") %>%
  arrange(desc(prop))
```

Which bins have the most DISC (singleton or discarded) ORFs?
```r
# most DISC
bins_df %>% 
  filter(k_unk_category == "DISC") %>%
  arrange(desc(prop))

bins_df %>% 
  filter(bin == "P_acnes") 

bins_df %>% 
  filter(bin == "P_rhinitidis") 

bins_df %>% 
  filter(bin == "E_facealis") 
```

# Pangenomic applications

The Agnostos workflow is a great way to get more out of genes of unknown function when comparing genomes. Maybe there are a set of accessory genes that separate genomes into ecotypes? [Chiara Vanni](https://orcid.org/0000-0002-1124-1147) also kindly integrated a set of *E. faecalis* external genomes for us to work with. Let's calculate a pangenomes of 6 *E. faecalis* genomes and 1 *E. faecalis* bin.

## Prepare External genomes

First, we will import agnostos categories as functions to the set of 6 external E. faecalis genomes. To start, we will get the data anvio ready in R. 
```r
# Get Chiara's integrated external genome data
external_genomes_agnosots_categories <- read_tsv("ext_genomes_genes_summary_info_exp.tsv")

# Separate the ORF data by genome
genome_list <- external_genomes_agnosots_categories %>%
  select(-gene) %>%
  group_split(genome)

# Convert data into anvio friendly function import format
export_functions <- function(DF) {

  file_name <- DF$genome %>% unique()

  path_name <- paste0("INFANT-GUT-TUTORIAL/additional-files/pangenomics/external-genomes/", file_name, "_agnostos_categories.tsv")

  DF %>%
    mutate("source" = "Agnostos_categories") %>%
    select(-genome) %>%
    rename(accession = cl_name, "function" = category) %>%
    mutate("e_value" = 0) %>%
    select(gene_callers_id, source, accession, "function", e_value) %>%
    write_tsv(path_name, col_names = TRUE)
}

# Run it
lapply(genome_list, export_functions)
```

Now we will import agnostos categories as (functions) to each contigs DB for the external genomes.
```bash
for GENOME in `ls INFANT-GUT-TUTORIAL/additional-files/pangenomics/external-genomes/*.db`;
do
  anvi-migrate "${GENOME}" --migrate-dbs-safely # because we've probably updated either contigs or profile dbs
  fname=$(basename "${GENOME}" .db)
  anvi-import-functions -c  "${GENOME}" -i INFANT-GUT-TUTORIAL/additional-files/pangenomics/external-genomes/"${fname}"_agnostos_categories.tsv;
done
```

## Internal genomes

Now that we have finished with the external genomes, let's prepare the E. faecalis bin Meren retrieved from the IGD metagenome. To do this, we will separate the E. faecalis bin from the IGD metagenome and import Agnostos categories to it.

First, let's get the E. Faecalis bin from the IGD dataset. Meren already did the binning for you :)

```bash
anvi-import-collection INFANT-GUT-TUTORIAL/additional-files/collections/e-faecalis.txt \
                       --bins-info INFANT-GUT-TUTORIAL/additional-files/collections/e-faecalis-info.txt \
                       -p INFANT-GUT-TUTORIAL/PROFILE.db \
                       -c INFANT-GUT-TUTORIAL/CONTIGS.db \
                       -C E_faecalis
```

Now we will enter Gene-mode for E. faecalis to get the specific gene-callers-ids with agnostos categories.

```bash
anvi-script-add-default-collection -p INFANT-GUT-TUTORIAL/PROFILE.db

anvi-interactive -p INFANT-GUT-TUTORIAL/PROFILE.db \
                 -c INFANT-GUT-TUTORIAL/CONTIGS.db \
                 -C default \
                 -b E_facealis \
                 --gene-mode

anvi-export-table INFANT-GUT-TUTORIAL/GENES/default-E_facealis.db \
                  --table gene_level_coverage_stats \
                  -o INFANT-GUT-TUTORIAL/GENES/default-E_facealis.tsv
```

## Integrating Agnostos categories into E. faecalis pangenome

Calculate the pangenome of E. faecalis
```bash
# Focus just on E. faecalis
grep -v "faecium" INFANT-GUT-TUTORIAL/additional-files/pangenomics/external-genomes.txt > INFANT-GUT-TUTORIAL/additional-files/pangenomics/external-genomes_efaecalis.txt

# Generate genome storage 
anvi-gen-genomes-storage -i INFANT-GUT-TUTORIAL/additional-files/pangenomics/internal-genomes.txt \
                         -e INFANT-GUT-TUTORIAL/additional-files/pangenomics/external-genomes_efaecalis.txt \
                         -o Enterococcus-faecalis-GENOMES.db

# Calculate pangenome
anvi-pan-genome -g Enterococcus-faecalis-GENOMES.db \
                -n Enterococcus_faecalis \
                -o Enterococcus_faecalis \
                --num-threads 14
```

Now that we have calculated a E. faecalis pangenome, let's add in the agnostos category info!

Get gene-cluster info
```bash
# anvi-export-table Enterococcus_faecalis/Enterococcus_faecalis-PAN.db --table gene_clusters -o Enterococcus_faecalis_gene_clusters.tsv

# Add default script to pangenome
anvi-script-add-default-collection -p Enterococcus_faecalis/Enterococcus_faecalis-PAN.db

# Get all the awesome tables
anvi-summarize -g Enterococcus-faecalis-GENOMES.db -p Enterococcus_faecalis/Enterococcus_faecalis-PAN.db -C DEFAULT -o SUMMARY-E_faecalis

# unzip the goods
gunzip SUMMARY-E_faecalis/Enterococcus_faecalis_gene_clusters_summary.txt.gz
```

Join with agnostos annotations
```r
 gene cluster info
# Enterococcus_gene_clusters <- read_tsv("Enterococcus_faecalis_gene_clusters.tsv") %>%
#   mutate(genome_name = gsub("E", "Enterococcus", genome_name))

Enterococcus_gene_clusters <- read_tsv("SUMMARY-E_faecalis/Enterococcus_faecalis_gene_clusters_summary.txt") %>%
  mutate(genome_name = gsub("E", "Enterococcus", genome_name)) %>%
  rename(gene_caller_id = "gene_callers_id")

# Join info for external gene calls
external_genomes_integrated_agnostos <- read_tsv("ext_genomes_gene_info.tsv") %>% select(-lowest_level, -lowest_rank, -project, -norfs, -is.LS)

# add proportions to external genomes
gene_cluster_additional_data <- Enterococcus_gene_clusters %>%
  unite("gene", c("genome_name", "gene_caller_id"), remove = FALSE, sep = "_") %>%
  left_join(external_genomes_integrated_agnostos) %>%
  select(gene_cluster_id, gene_caller_id, category, genome_name) %>%
  filter(!grepl("SHA", genome_name)) # just get the external genomes

# add proportions to internal genome (MAG)
internal_genome_annotations <- Enterococcus_gene_clusters %>%
  filter(grepl("SHA", genome_name)) %>%
  left_join(IGD_genes_summary_info_exp %>% rename(gene_caller_id = gene_callers_id)) %>%
  select(gene_cluster_id, gene_caller_id, category, genome_name)

gene_cluster_agnostos_categories <- gene_cluster_additional_data %>%
  bind_rows(internal_genome_annotations) %>%
  group_by(gene_cluster_id) %>%
  select(gene_cluster_id, gene_caller_id, category) %>%
  count(category) %>%
  mutate(prop = n/sum(n)) %>%
  select(-n) %>%
  spread(category, prop, fill = 0) %>%
  rename("Agnostos_catergories!DISC" = DISC,
         "Agnostos_catergories!SINGL" = SINGL,
         "Agnostos_catergories!GU" = GU,
         "Agnostos_catergories!K" = K,
         "Agnostos_catergories!KWP" = KWP) 
  
gene_cluster_agnostos_categories %>%
  write_tsv(path = "Enterococcus_faecalis/gene_cluster_additional_data.tsv")
```

```bash
anvi-import-misc-data Enterococcus_faecalis/gene_cluster_additional_data.tsv \
                      -p Enterococcus_faecalis/Enterococcus_faecalis-PAN.db \
                      --target-data-table items \
                      --just-do-it
```

and finally visualize
```bash
anvi-display-pan -g Enterococcus-faecalis-GENOMES.db \
                 -p Enterococcus_faecalis/Enterococcus_faecalis-PAN.db \
                 --title "Enterococccus Pan"\
```
[![agnostos_pan](/images/agnostos_pan.png)](images/agnostos_pan.png){:.center-img .width-100}

Above we can see the E. faecalis pangenome with an extra layer designating the Agnostos categories 


[![agnostos_pan_syn](/images/agnostos_pan_syn.png)](images/agnostos_pan_syn.png){:.center-img .width-100}

[![agnostos_pan_syn_gus](/images/agnostos_pan_syn_gus.png)](images/agnostos_pan_syn_gus.png){:.center-img .width-100}

