---
layout: post
authors: [sam]
title: "Anvi'o tRNA-seq capabilities"
excerpt: "Welcome to the future"
modified: 2021-01-18
tags: []
categories: [anvio]
comments: true
---

{% include _toc.html %}

{% include _project-anvio-version.html %}

{% capture images %}{{site.url}}/images/anvio/2021-01-18-anvio-trna-seq-capabilities{% endcapture %}

{:.warning}
tRNA-seq capabilities are available in anvi'o `v7` and later. You can identify which version you have on your computer by typing `anvi-self-test --version` in your terminal. Some details mentioned in this post are not included in the `v7` release but can be used if you are tracking the anvi'o `dev` branch.

Here is a diagram of the anvi'o tRNA-seq workflow that will be explained...
[![trnaseq-workflow-diagram]({{images}}/trnaseq_workflow_diagram.png)]({{images}}/trnaseq_workflow_diagram.png){:.center-img .width-100}

## Disclaimer

High-throughput tRNA sequencing (tRNA-seq) is a cutting edge technique that is currently used to study diverse environmental and clinical samples by the [Pan](https://openwetware.org/wiki/Pan_Lab) and [Meren](https://merenlab.org/) labs at the University of Chicago. Given the variety and richness of data generated by tRNA-seq -- as well as the low cost of sequencing relative to metagenomics and metatranscriptomics -- we anticipate that many other scientists will soon add this approach to their multi-omics toolkit. If you read on and find tRNA-seq could be worth trying in a project of your own, we encourage you to [email the author](mailto:samuelmiller@uchicago.edu) to explore setting up analyses.

## Background

Transfer RNA (tRNA) is an essential class of molecules that transports and adds amino acids to proteins being synthesized in the ribosome. tRNA is 10 times more abundant than mRNA by mass, and is the second most abundant class of RNA after rRNA. Each tRNA decodes a certain amino acid, as it has an anticodon complementary to one or more codons encoding that amino acid. For example, the amino acid, tryptophan, is encoded by a single codon, **UGG**, which is decoded by the anticodon, **CCA** (the reverse complement of **UGG**). The amino acid, leucine, is encoded by two codons, **UUA** and **UUG**. **UUA** is decoded by the anticodon, **UAA** (the reverse complement). **UUG** is decoded by two anticodons, **CAA** (the reverse complement) and **UAA**, which forms a G-U wobble base pair.

[![trna-structure]({{images}}/trna_structure.png)]({{images}}/trna_structure.png){:.center-img .width-50}

---
*The tertiary and secondary structure of tRNA, with the 3'-CCA <span style="color:DarkKhaki">acceptor tail</span> in yellow, the <span style="color:DarkViolet">acceptor stem</span> in purple, the <span style="color:DarkOrange">variable loop</span> in orange, the <span style="color:DarkRed">D arm</span> in red, the <span style="color:RoyalBlue">anticodon arm</span> in blue with the anticodon in black, and the <span style="color:DarkGreen">T arm</span> in green. Source: [Wikipedia](https://bit.ly/35RsqVz).*

---
[![trna-ribosome-translocation]({{images}}/trna_ribosome_translocation.gif)]({{images}}/trna_ribosome_translocation.gif){:.center-img .width-50}

---
*Translation, showing the codon-anticodon pairing of mRNA and tRNA, and the elongation and termination of an amino acid chain. tRNA is a flexible molecule that passes through multiple binding states during translation, interacting with the ribosome, mRNA, tRNA in other states, and protein factors. Source: [Noller lab](http://rna.ucsc.edu/rnacenter/ribosome_movies.html).*

---
During translation, tRNA molecules nestled in the ribosome connect an mRNA codon to the corresponding amino acid added to a polypeptide chain. To facilitate these interactions, ~10-20% of a tRNA molecule's ~74-95 nucleotides (A, C, G, and U) are modified with a wide range of chemical groups. Organisms adapt to environmental perturbation over the course of minutes to hours by adding and subtracting nucleotide modifications and by synthesizing and degrading tRNA molecules. Changes in modification levels and tRNA abundances alter the rate of decoding and thereby protein production. Genes containing certain codons can be favored under normal conditions and others containing a contrasting set of codons favored under stress. Measurement of the relative abundances of tRNAs and their modifications opens the domain of "metaepitranscriptomics," enabling the study of how cells adapt to the environment short of death and natural selection.

[![human-trna-modifications]({{images}}/human_trna_modifications.png)]({{images}}/human_trna_modifications.png){:.center-img .width-60}

---
*Known modifications in human cytoplasmic tRNA. Source: [de Cr√©cy-Legard et al., 2019](https://academic.oup.com/nar/article/47/5/2143/5304329).*

---
High-throughput tRNA sequencing (tRNA-seq), as developed by Tao Pan's lab at the University of Chicago, overcomes challenges that stymied single-shot analysis of in situ tRNA pools. The compact, stable structure of tRNA necessitated the invention of an adapter to the universal 3'-CCA tail of the molecule that allows a reverse transcriptase (RT) to gain a foothold and synthesize cDNA from the template. No PCR amplification is involved in this process, eliminating a common source of error in marker gene datasets. Many types of modifications are detected as mutations (substitutions and indels) in the cDNA, as RT overcomes the obstruction caused by the unexpected chemical group by inserting a semi-random nucleotide or skipping over the site. RT can also fail to overcome the modification, falling off the template and producing cDNA only from a 3' fragment of the tRNA. Furthermore, a small number of untemplated nucleotides (~1-3) is commonly added to the 5' or 3' end of the cDNA product. tRNA-seq data is therefore unique and complex, requiring specialized analytical tools.

Our anvi'o tRNA-seq workflow takes in raw sequence libraries and generates data products that fit into the rest of the anvi'o ecosystem. The workflow includes novel tools for the **de novo** prediction of tRNA sequences and modifications -- you do not need to know what organisms and tRNAs are in your sample ahead of time, and no external references are used in these predictions. The workflow finds biological tRNA fragments, which are increasingly recognized as important regulatory molecules. It assigns taxonomy to tRNA sequences, leveraging tRNA gene sequences in GTDB, a giant, taxonomically organized database of prokaryotic genomes. Taxonomy can be assigned to eukaryotic tRNA as well, as there are no taxonomic constraints on which tRNAs can be sampled and sequenced, unlike marker genes such as 16S rRNA.

## Snakemake workflow

An analytical pipeline for a tRNA-seq experiment can be run using a Snakemake workflow, as adapted by the program, `anvi-run-workflow`. Snakemake is designed for reproducible and scalable data analysis, seamlessly connecting the inputs and outputs of programs in the pipeline. `anvi-run-workflow` takes a config file as its input. Use the following command to write a config file of default workflow parameterizations in the working directory (here, the file is named `config.json`).

```
anvi-run-workflow -w trnaseq --get-default-config config.json
```

Now you need to look through and modify the config file. The value of `samples_txt` must refer to another file that you create. The `samples_txt` file contains information on each input sample. We in the Pan and Meren labs have been dealing with paired-end sequence libraries; the workflow merges and quality-controls the forward and reverse reads in FASTQ files using the program, [Illumina-utils](https://github.com/merenlab/illumina-utils). We added functionality to this program to handle both full and partial overlap of the cDNA insert by the paired-end reads. Despite the shortness of tRNAs -- typically less than 96 nt -- partially overlapping reads can result from the inclusion of barcodes and unique molecular identifiers in 100 bp reads. Also, troves of other small RNAs longer than mature tRNA are captured in the samples, including pre-tRNA, signal recognition particle (SRP) RNA, and 5S rRNA. If you wish to use a program other than Illumina-utils for read merging and QC, then you will use FASTA files rather than FASTQ files as inputs to the workflow. `samples_txt` allows for both types of inputs.

Here is an example of what a `samples_txt` file should look like with **FASTQ** inputs:

| sample | treatment | r1 | r2 | r1_prefix | r2_prefix |
| :--- | :--- | :--- | :--- | :--- | :--- |
| stool_1A_noDM | untreated | FASTQ/stool_1A_noDM.r1.fq.gz | FASTQ/stool_1A_noDM.r2.fq.gz | NNNNNN | ACGTACG |
| stool_1A_DM | demethylase | FASTQ/stool_1A_DM.r1.fq.gz | FASTQ/stool_1A_DM.r2.fq.gz |  | TGCAACG |
| stool_1B_noDM | untreated | FASTQ/stool_1B_noDM.r1.fq.gz | FASTQ/stool_1B_noDM.r2.fq.gz | NNNNNN | GACTACG |
| (...) | (...) | (...) | (...) | (...) | (...) |

{:.warning}
Illumina-utils only recognizes FASTQ deflines with the [Casava 1.8 format](https://en.wikipedia.org/wiki/FASTQ_format#Illumina_sequence_identifiers). Check that the FASTQ files produced by your demultiplexing tool have this format.

Here is an example of what a `samples_txt` file should look like with **FASTA** inputs:

| sample | treatment | fasta |
| :--- | :--- | :--- |
| ecoli_1A_noDM | untreated | FASTA/ecoli_1A_noDM.fa.gz |
| ecoli_1A_DM | demethylase | FASTA/ecoli_1A_DM.fa.gz |
| ecoli_1A_DM_PD | demethylase_periodate | FASTA/ecoli_1A_DM_BE.fa.gz |
| (...) | (...) | (...) |

The same column headers should always be used (6 with FASTQ and 3 with FASTA inputs), even if you do not need to enter values in the `treatment`, `r1_prefix`, and `r2_prefix` columns.

`treatment` refers to the chemical treatment used in processing the sample, such as `demethylase`. If any entries in the column are kept blank, the samples will be assumed to be `untreated`.

`r1_prefix` and `r2_prefix` refer to sequences found at the beginning of the reads in the `r1` and `r2` files. In the example above, a unique molecular identifier of 6 random nucleotides (NNNNNN) was placed at the start of r1 in 2 of the 3 samples. By setting the flag `--report-r1-prefix` -- found, like all accepted program parameters, in the config file -- you can instruct `Illumina-utils` to report this sequence from each merged read in a supplementary FASTA file. There is similarly a flag called `--report-r2-prefix`. In the example above, a sample barcode of 7 nucleotides was placed at the start of r2; it doesn't need to be reported since it is already known.

Other parameters in the config file should be set in addition to `samples_txt`. You need to provide a `--project-name` for `anvi_convert_trnaseq_database`, a program which processes all of your input samples as "one experiment" to find common tRNA seed sequences and other results. If for whatever reason you are processing samples from multiple experiments in a single workflow and so the samples -- say, from soil and water -- should not all be compared, then set the `run` parameter to `false` for `anvi_convert_trnaseq_database` and the following rule, `anvi_run_trna_taxonomy`. These programs can later be run on the desired subsets of samples. If you are starting with FASTA rather than FASTQ files, make sure to set the `run` parameter to `false` for `iu_merge_pairs`. The `run` parameter for `anvi_reformat_fasta` should be set to `true` unless you are certain that your FASTA deflines are anvi'o-compliant.

The following rules are multithreaded, so the number of `--threads` that Snakemake can allocate to the program should be adjusted accordingly given your computational resources: `iu_merge_pairs`, `anvi_trnaseq`, `anvi_convert_trnaseq_database`, and `anvi_run_trna_taxonomy`. The general parameter, `max_threads`, must also be adjusted.
