---
layout: post
authors: [matt, chiara]
title: "Integrating AGNOSTOS gene categories into anvi'o projects"
excerpt: "Explore the unknown!"
modified: 2021-05-28
tags: []
categories: [anvio]
comments: true
redirect_from:
  - /agnostos-tutorial/
image:
    feature: https://merenlab.org/images/agnostos/summary.png
    display: false
---


Anyone who studies microbial life through genomes or metagenomes often finds out that a large proportion of their genes do not have any functional annotation. As of today, 40% to 60% of predicted genes from genomes or metagenomes have no significant homology to genes with known or predicted functions in public nucleotide databases - which is only getting worse as we sequence more and more microbiomes from around the globe. To improve accessiblity of genes of unknown functions, [Vanni *et al*., 2021](https://elifesciences.org/articles/67667), have recently introduced [AGNOSTOS](https://github.com/functional-dark-side/agnostos-wf), a computational framework that aims to help microbiologists integrate all genes found in genomes and metagenomes into their ecological and evolutionary investigations. We strongly recommend for anyone who is interested in this topic to also read [this blog post](http://merenlab.org/2020/07/01/dark-side/) by [Antonio Fernandez-Guerra](
Antonio Fernandez-Guerra) ([orcid](http://orcid.org/0000-0002-8679-490X)), where he describes the origins of AGNOSTOS, the categories of gene-clusters (GCs) it produces, and discusses the future of unknown genes in microbiomes.

[Vanni *et al*., 2021](https://elifesciences.org/articles/67667) demonstrated that [the AGNOSTOS workflow](https://github.com/functional-dark-side/agnostos-wf) was able to shed light on the unknown sequence landscape of microbiomes from oceans and the human body by revealing thousands of lineage-specific unknown genes and their ecology. Now, we are motivated to turn AGNOSTOS into a resource that empowers researchers to explore genes of unknown function in their own datasets. To accomplish this, we created this tutorial to demonstrate how to integrate AGNOSTOS GC categories into anvi'o projects using the database [AGNOSTOS-DB](https://doi.org/10.1101/2021.06.07.447314).

Even though the tutorial explains the key steps in this integration for anvi'o projects, we are hoping that the example input and output files will enable anyone to be able to implement this workflow independent of tools.

## Setting up the stage

{:.warning}
**The sole purpose of this subsection is to help you download the necessary data to follow the tutorial**. If you want to go straight to data analysis, feel free to skip this section.

Throughout the tutorial, we will use [the Infant Gut Dataset](https://merenlab.org/tutorials/infant-gut/) (IGD) as an example project from [Sharon et al., 2013](http://www.ncbi.nlm.nih.gov/pubmed/22936250). Please first follow the download instructions in the IGD tutorial [here](https://merenlab.org/tutorials/infant-gut/#downloading-the-pre-packaged-infant-gut-dataset) to get the {% include ARTIFACT name="contigs-db" text="contigs database" %} in your working directory.

```bash
$ ls
AUXILIARY-DATA.db  CONTIGS.db  PROFILE.db  additional-files/
```

If the above is what you see in your work directory, then you are good to follow the rest of the tutorial!

## Running AGNOSTOS with data from anvi'o

The purpose of this section is to demonstrate how to extract the necessary data files from an anvi'o project to run the AGNOSTOS workflow on them. If you want to skip straight to analyzing AGNOSTOS integrated data in anvi'o then please skip to [Download the goods](#download-the-goods). 

AGNOSTOS needs two things from anvi'o: {% include ARTIFACT name="genes-fasta" text="genes-fasta" %} of gene calls and {% include ARTIFACT name="gene-calls-txt" text="gene-calls-txt" %} 


### Extracting data from anvi'o
	
The first step is to export the gene sequences from the {% include ARTIFACT name="contigs-db" text="contigs database" %} using the anvi'o program {% include PROGRAM name="anvi-get-sequences-for-gene-calls" %}:

```bash
anvi-get-sequences-for-gene-calls -c CONTIGS.db \
                                  --get-aa-sequences \
                                  -o infant_gut_genes.fasta
```

Next, we need the gene completion information i.e. does a gene contain both a start and stop codon? This can be found in the {% include ARTIFACT name="gene-calls-txt" %} which can be extracted from your {% include ARTIFACT name="contigs-db" text="contigs database" %}. To do this, we can use the program {% include PROGRAM name="anvi-export-gene-calls" %}:

```bash
anvi-export-gene-calls -c CONTIGS.db \
                       --gene-caller prodigal \
                       -o infant_gut_gene_calls.tsv
```

And that's it!

At this point, we have all the necessary files to run the [AGNOSTOS-workflow](https://github.com/functional-dark-side/agnostos-wf), namely the gene amino acid sequences (`infant_gut_genes.fasta`) and the gene completion information (`infant_gut_gene_calls.tsv`).


### Running AGNOSTOS

<div class="extra-info" markdown="1">

<span class="extra-info-header">AGNOSTOS gene categories</span>

Before we start, let's have a refresher on AGNOSTOS gene annotation categories discussed in Antonio's [blog post](https://merenlab.org/2020/07/01/dark-side/):

* Known with Pfam annotations (**K**): genes annotated to contain one or more Pfam entries (domain, family, repeats or motifs) but excluding the domains of unknown function (DUF).
* Known without Pfam annotations (**KWP**): which contains the genes that have a known function but lack a Pfam annotation. Here we can find intrinsically disordered proteins or small proteins, among others.
* Genomic unknown (**GU**): genes that have an unknown function (DUF are included here) and found in sequenced or draft genomes
* Environmental unknown (**EU**): genes of unknown function not detected in sequenced or draft genomes, but only in environmental metagenomes or metagenome-assembled genomes.

</div>

The purpose of this section is to demonstrate different ways of utilizing AGNOSTOS depending on the computational resources you have available and the kind of information you want to learn about your unknown sequences. The easiest way to take advantage of AGNOSTOS is to run a profile search of the AGNOSTOS-DB clusters against predicted ORFs from your sequencing data. This will quickly annotate your sequences and tell you about the landscape of your unknown sequencing space. Additionally, you can leverage the metadata associated with each AGNOSTOS cluster annotation, including its lineage specificity, niche breadth, etc.

However, running [AGNOSTOS](https://github.com/functional-dark-side/agnostos-wf) integrates your sequences into the AGNOSTOS-DB clusters and potentially creates new clusters if your genomes and metagenomes have novel homologous sequences (go read about that [here](https://merenlab.org/2020/07/01/dark-side/#why-gene-clusters)). The concept of ORF integration is key here because instead of annotating the IGD metagenomes ORFs, AGNOSTOS will attempt to insert new ORFs into the extant AGNOSTOS clusters that already have metadata associated with them (read more about it [here](http://merenlab.org/2020/07/01/dark-side/#why-gene-clusters)), otherwise it will create new GCs instead of leaving gene unannotated. Thus, integration allows for the entire dataset of ORFs to be used rather than alignment-based methods where one filters for best hits.

The [AGNOSTOS-workflow](https://github.com/functional-dark-side/agnostos-wf) is a complex workflow that relies on many external dependencies. Furthermore, to achieve high sensitivity levels, some of their steps are computationally expensive for large datasets. Currently, [AGNOSTOS](https://github.com/functional-dark-side/agnostos-wf), although fully functional, should be considered proof of concept of what can be done. At the moment, we are working to make it more accessible and less resource-intensive. This tutorial is the starting point for our plans to integrate an optimized version of AGNOSTOS into anvi'o.

Regardless of which way you decided to use AGNOSTOS we will visualize the data in the next section.

To get started please clone the AGNOSTOS repo:
```bash
git clone https://github.com/functional-dark-side/agnostos-wf.git && cd agnostos-wf
```

#### Easy mode - quickly annotated sequences with AGNOSTOS categories

Here we will quickly annotate the IGD dataset with AGNOSTOS categories using profile based searching.

First, download the AGNOSTOS-DB profiles and cluster categories here:

```bash
# Make a home for the AGNOSTOS data
mkdir -p IGD_agnostos

# Download and unzip AGNOSTOS profiles
wget https://ndownloader.figshare.com/files/23066963 -O IGD_agnostos/GC_profiles.tar.gz 
tar -zxvf IGD_agnostos/GC_profiles.tar.gz

# Download and unzip AGNOSTOS cluster categories
wget https://ndownloader.figshare.com/files/23067140 -O IGD_agnostos/cluster_ids_categ.tsv.gz
gunzip IGD_agnostos/cluster_ids_categ.tsv.gz
```

Next, run the profile search to annotate your sequences with AGNOSTOS GCs.

{:.warning}
Please conda install [mmseqs2](https://github.com/soedinglab/MMseqs2#installation) to run this profile search

```bash
Profile_search/profile_search.sh --query infant_gut_genes.fasta \
			         --clu_hmm IGD_agnostos/GC_profiles/clu_hmm_db \
			         --clu_cat IGD_agnostos/cluster_ids_categ.tsv \
			         --threads 10
```

#### Hard mode - integrate sequences into AGNOSTOS-DB

{:.notice}
Some steps of AGNOSTOS uses MPI compiled programs to be able to process large data sets. It also uses SLURM to distribute many small jobs across many nodes. Although it might work in commodity hardware after having all dependencies installed, we only have tested it in a cloud or HPC infrastructure. Soon we hope to provide [Singularity](https://sylabs.io/singularity/) images to make the whole process way more accessible.

Installation instructions for [AGNOSTOS](https://github.com/functional-dark-side/agnostos-wf) can be found [here](https://github.com/functional-dark-side/agnostos-wf/blob/master/AGNOSTOS_usage.md). Once you've completed the installation on your HPC come back here and continue the tutorial. If you have any trouble installing AGNOSTOS, don't hesitate to open an issue [here](https://github.com/functional-dark-side/agnostos-wf/issues)

## Download the goods

Let's get back to the tutorial exploring unknown genes in IGD!

This section will go over how to integrate and visualize AGNOSTOS categories in your anvi'o projects. To have some fun, [Chiara Vanni](https://orcid.org/0000-0002-1124-1147) kindly integrated all of the ORFs from the IGD metagenomes using [AGNOSTOS](https://github.com/functional-dark-side/agnostos-wf). Additionally, she integrated the ORFs from 6 *E. faecalis* so we can investigate AGNOSTOS categories in the context of a pangenome.

Please download these files to start the analysis:
```bash
# Make sure you are back in Infant gut
cd INFANT-GUT-TUTORIAL

# Make a home for the AGNOSTOS data
mkdir -p IGD_agnostos

# FIXME: make figshare link for necessary data from [git clone https://github.com/genomewalker/ebame6.git](https://github.com/genomewalker/ebame6/tree/main/data)

wget figsharelink -O IGD_AGNOSTOS.tar.gz
```

You should now see this in your working directory

```bash
$ ls
AUXILIARY-DATA.db CONTIGS.db        IGD_agnostos      PROFILE.db        additional-files
```

{:.notice}
AGNOSTOS produces a very extensive output. [Here](https://github.com/functional-dark-side/agnostos-wf/blob/master/Output_README.md) you can find a description of all the fields reported. For this tutorial, we cherrypicked some of them.

Now you are ready to import the AGNOSTOS output for your genes back into your anvi'o project.

## What type of results does AGNOSTOS provide?

Let's have a look at what is inside of the file `IGD_agnostos/IGD_genes_summary_info_exp_coverage.tsv`. If you have installed the awesome [csvtk](https://github.com/shenwei356/csvtk) you can easily check the contents with:

```bash
csvtk csv2md -t <(head -n4 IGD_agnostos/IGD_genes_summary_info_exp_coverage.tsv)
```

| cl_name  | gene_callers_id | contig           | gene_x_contig | cl_size | category | pfam            | is.HQ | is.LS | lowest_rank | lowest_level | niche_breadth_sign | is.singleton | ground_coverage      |
| :------- | :-------------- | :--------------- | :------------ | :------ | :------- | :-------------- | :---- | :---- | :---------- | :----------- | :----------------- | :----------- | :------------------- |
| 40423550 | 0               | Day17a_QCcontig1 | 1051          | 1       | KWP      | NA              | FALSE | FALSE | NA          | NA           | NA                 | TRUE         | NA                   |
| 32886021 | 1               | Day17a_QCcontig1 | 1051          | 3       | K        | Glycos_transf   | FALSE | FALSE | NA          | NA           | NA                 | FALSE        | 0.000257075416579027 |
| 16622933 | 2               | Day17a_QCcontig1 | 1051          | 3606    | K        | Glycos_transf_1 | FALSE | FALSE | NA          | NA           | Broad              | FALSE        | 0.0153310430250765   |


Some of the interesting columns to check are the cluster size (**cl_size**); to which category the gene cluster (GC) belongs (**category**); if the GC is high-quality (**is.HQ**); if it is lineage-specific, and which rank and level (**is.LS**, **lowest_rank**, **lowest_level**); and its environmental distribution based on the metagenomes used to build the seed DB (**niche_breadth_sign**). 

<div class="extra-info" markdown="1"> 
We added an extra column that is not yet integrated into the workflow (**ground_coverage**) that provides the taxonomic coverage of a GC related to a base taxonomy; the larger the number, the more taxa that have this GC. This metric is one of the outputs of an experimental application of AGNOSTOS to identify contamination in MAGs, where we flag potential contaminants at the contig level (and even within a contig, identifying misassemblies, relevant for ancient metagenomics). The method exploits the information contained in all genes in a MAG, and it considers things like the presence of prophage regions or mobile genetic elements. This is the <a href="https://merenlab.org/2020/07/01/dark-side/#why-gene-clusters"> beauty of GCs</a>, where one can combine function and taxonomy. We will write soon another blog post describing the method and how it can help to automatize the manual curation of bins.
</div>


Now you are probably wondering, how many KNOWNS and UNKNOWNS do we have? Let's figure it out with a little of awk-fu (you will need GAWK 4):

```bash
$ gawk 'NR>1{a[$6]++;}END { PROCINFO["sorted_in"] = "@val_num_desc"; for(n in a ){printf "%s: %s (%.2f%)\n", n, a[n], 100*(a[n]/(NR-1))}}' IGD_agnostos/IGD_genes_summary_info_exp_coverage.tsv
K: 18576 (57.56%)
GU: 4506 (13.96%)
DISC: 2851 (8.83%)
KWP: 2406 (7.45%)
NA: 2095 (6.49%)
EU: 1840 (5.70%)
```

As expected, the majority of the ORFs (57%) are part of the Known (K) fraction. Then the next category with more ORFs is the Genomic unknowns (GU). The category **DISC** are those ORFs that have been discarded during the [cluster validation](https://dark.metagenomics.eu/cluster-validation) step, we are quite conservative here. Then, there is a category with the name **NA** that contains all ORFs that are most likely gene miscalls. Some of the contigs have many __Ns__ and we also integrated the PRODIGAL predictions for __Candida__, in both cases, we might have many spurious ORFs. AGNOSTOS will not discard them, but they will end up as singletons. As a side note, for the tutorial we integrated the singletons after being [classified](https://dark.metagenomics.eu/cluster-classification) and [refined](https://dark.metagenomics.eu/cluster-refinement).

## Import AGNOSTOS into an anvi'o

Whether you have your AGNOSTOS data tables or have downloaded the pre-calculated datasets above, now it's time to get the AGNOSTOS data back into anvi'o. The first step will be to import the data as {% include ARTIFACT name="functions" text="functions" %} into the IGD {% include ARTIFACT name="contigs-db" text="contigs database" %} (check out this [post](http://merenlab.org/2016/06/18/importing-functions/) if you have more questions about anvi'o functions tables). This will make it easy to explore the AGNOSTOS categories in the context of assembled contigs and read recruitment results in the anvi'o interactive interface!
 

Import AGNOSTOS data into the IGD {% include ARTIFACT name="contigs-db" text="contigs database" %}:

Metagenomes
```bash
anvi-import-functions -c CONTIGS.db -p AGNOSTOS -i IGD_agnostos/IGD_genes_summary_info_exp_coverage.tsv
```

Pangenomes:
```bash
# Generate genome storage
anvi-gen-genomes-storage -e external-genomes.txt -o Enterococcus-GENOMES.db

# Calculate pangenome 
anvi-pan-genome -g Enterococcus-GENOMES.db \
                -n Enterococcus \
                -o PAN \
                --num-threads 10

# Visualize
anvi-display-pan -g Enterococcus-GENOMES.db \
                 -p PAN/Enterococcus-PAN.db \
                 --title "Enterococccus Pan"
```

## Metagenomics applications

Now that we have imported the AGNOSTOS categories into our {% include ARTIFACT name="contigs-db" text="contigs database" %}, let's check out the distribution of categories in metagenomic-assembled genomes (MAGs) from the IGD metagenome dataset:

First, let's import some bins Meren made:
```
anvi-import-collection additional-files/collections/merens.txt \
                       --bins-info additional-files/collections/merens-info.txt \
                       -p PROFILE.db \
                       -c CONTIGS.db \
                       -C default
```


Since we already imported the AGNOSTOS output into anvi'o, we can immediately visualize the results like this:
```
anvi-interactive -p PROFILE.db -c CONTIGS.db -F AGNOSTOS
```
[![agnostos_binning](/images/agnostos/agnostos_binning.png)](){:.center-img .width-2}

[![agnostos_binning_legend](/images/agnostos/agnostos_legend.jpeg)](images/agnostos_legend.jpeg){:.center-img .width-2}

So now we have the data integrated and ready to be used in anvi'o. For example, you can do a global analysis like the one we did with [Tom Delmont](https://twitter.com/tomodelmont) with the [SMAGS](https://www.biorxiv.org/content/10.1101/2020.10.15.341214v2) or you can use some of the associated metadata to the GCs. Below we use the AGNOSTOS metadata for environmental distribution and GC taxonomic coverage to select which ORFs with unknown functions (Genomic unknowns) would be interisting to explore first:

[![agnostos_taxcov](/images/agnostos/agnostos_taxcov.png)](){:.center-img .width-2}

For example, if we are interested in finding potential new marker genes, we can select the Genomic unknowns that have a broad environmental distribution and a large taxonomic coverage. Another direction we can go is to find ORFs with unknown function that might be associated with very specific taxonomic groups and give them a fitness advantage allowing for broad distribution. To find these ORFs, we can select those with a broad environmental distribution but very limited taxonomic coverage. These are just random thoughts for the tutorial, but we hope that you can see the potential of being able to access to the unknown fraction when you are exploring your data.

## Pangenomic applications

In this section we will continue with our sandbox [INFANT-GUT-DATAST](https://merenlab.org/tutorials/infant-gut/#chapter-iv-pangenomics) and go over how to utilize AGNOSTOS in the context of pangenomics. First, we will learn how to integrate the AGNOSTOS results into an anvi'o {% include ARTIFACT name="pan-db" %}. Next, we will explore the differences between the GCs produced by anvi'o with {% include PROGRAM name="anvi-pan-genome" %} versus AGNOSTOS - both methods use different approaches to cluster and validate GCs. Finally, we will select [single-copy core genes](https://anvio.org/vocabulary/#single-copy-core-gene-scg) found in the pangenome that were categorized as Genomic Unknowns (GUs) and explore their phylogenetic breadth across the bacterial tree of life. Our main goal with this section is to demonstrate what new information AGNOSTOS gives you about genes of unknown function which usually end up discarded in our analyses.

First we will integrate AGNOSTOS data into an anvi'o pangenome.
```bash
anvi-import-misc-data IGD_agnostos/gene_cluster_additional_data.tsv \
                      -p IGD_agnostos/PAN/Enterococcus-PAN.db \
                      --target-data-table items \
                      --just-do-it

anvi-import-state -p IGD_agnostos/PAN/Enterococcus-PAN.db \
                  --state IGD_agnostos/state-pan.json \
                  --name default
```

Then we can display the pangenome with AGNOSTOS categories

{:.notice}
If you are interested in learning more about pangenomes in anvi'o please refer this [tutorial](https://merenlab.org/2016/11/08/pangenomics-v2/).

```bash
anvi-display-pan -g IGD_agnostos/Enterococcus-GENOMES.db \
                 -p IGD_agnostos/PAN/Enterococcus-PAN.db \
                 --title "Enterococccus Pan"
```

### anvi'o vs AGNOSTOS gene clusters

Just as a reminder, the anvi'o pangenome calculated GCs to learn about the extent of gene content shared between Enterococcus genomes, but remember, AGNOSTOS has it's own gene clusters (we know that's confusing, we'll explain in a second). AGNOSTOS has two layers in the pangenome you should consider `AGNOSTOS number of clusters` and `Agnostos categories`. The first layer contains the number of AGNOSTOS GCs contained in a anvi'o GC. The second layer includes the proportions of AGNOSTOS GC categories in a AGNOSTOS GC  (what... a GC can have multiple categories... yes C'est la vie, we will come back to this later).

To start, let's find out the degree of agreement between AGNOSTOS and anvi'o GCs. This is an important step if we want to use the information that AGNOSTOS provides. If the agreement is low, then we have a problem (FIXME: need explanation). Let's search the pangenome to highlight anvi'o GCs that contain more than one ANGOSTOS GC:

{% include IMAGE path="/images/agnostos/agnostos_search_clusters.png" caption="Figure x: test test test" %}

We found 1,387 GCs out of 6,412 i.e. ~20% of anvio GCs contain more than one AGNOSTOS GC. This is something we would expect. In AGNOSTOS, we define pretty stricter parameters for GCs than anvi'o i.e. AGNOSTOS will more often then not great more GCs than an anvi'o pangenome. This is because in AGNOSTSO we use strict alignment parameters with the clusering algorithm [MMseqs2](https://github.com/soedinglab/MMseqs2). AGNOSTOS tried to (1) maintain the multi-domain structure in a GC and (2) avoid lumping together genes that might be unrelated but have partial matches. To achieve this, two genes will have a connection, if they have a [bidirectional coverage](https://github.com/soedinglab/MMseqs2/wiki#bidirectional-coverage) of at least 80% of their sequence. You can have a look [here](https://dark.metagenomics.eu/cluster-validation) if you are curious about how we validate and refine our GCs.

Let's look at this a little bit more in detail. Let's sort the GCs by the [combined GC homogeneity index](https://merenlab.org/2016/11/08/pangenomics-v2/):

{:.notice}
A GC may contain amino acid sequences from different genomes that are almost identical, which would be a highly homogeneous GC. Another GC may contain highly divergent amino acid sequences from different genomes, which would then be a highly non-homogeneous one, and so on.

{% include IMAGE path="/images/agnostos/agnostos_fig_b.png" caption="Figure x: test test test" %}

Check out this pattern. The GCs with low [homogeneity](https://merenlab.org/2016/11/08/pangenomics-v2/#inferring-the-homogeneity-of-gene-clusters) tend to contain more AGNOSTOS GCs.

{% include IMAGE path="/images/agnostos/agnostos-pan-9.png" caption="Figure asdf" %}

Here is where you can see that gene clustering is not an easy process. This tutorial, is a tiny example of gene clustering, but when scaling it up to millions or billions of genes, it gets very complicated for the computational methods to set the boundaries of where to partition; sometimes, these methods will pull entirely unrelated sequences into one GC. This is why it is essential to understand (or partly understand) what we want to do and the aim of our clustering. In this specific case, the clustering algorithm used by anvi'o has a couple of parameters we can tune to improve the clustering results. In AGNOSTOS, we use the same clustering algorithm to group GCs in GC communities, and we developed a [semi-automated way](https://dark.metagenomics.eu/cluster-communities) to identify the best parameters for our purposes. Always explore your data and test (if possible) different values for the parameters of your programs of choice.

To put the nail in the coffin RE clustering differences between anvi'o and AGNOSTOS, let's explore the AGNOSTOS categorie Knowns:

{% include IMAGE path="/images/agnostos/agnostos-pan-1.png" caption="Figure asdf" %}

This will result in the following display:

{% include IMAGE path="/images/agnostos/agnostos-pan-2.png" caption="Figure asdf" %}

Do you see something weird? Let's look at it a bit closer...

{% include IMAGE path="/images/agnostos/agnostos-pan-3.png" caption="Figure asdf" %}

Besides the issue described previously, we see a mix of categories in a GC. While most of the time is due to the unspecific clustering, sometimes AGNOSTOS struggles to assign the right category to a GC. This is a common problem anytime one wants to do an annotation. Although we try to not rely on any other source for annotation than PFAM, we follow a thorough [process](https://dark.metagenomics.eu/cluster-classification) to decide if a GC is known or unknown and its [refinement](https://dark.metagenomics.eu/category-refinement). Although we are extremely careful, it fails sometimes. Furthermore, the fact that we are very strict on the thresholds we use, sometimes we might miss some potential annotations.

### Exploring genes of unknown functions

Time to dive a bit into the **unknown**... Let's now sort the splits by the Genomic Unknown categories:

{% include IMAGE path="/images/agnostos/agnostos-pan-6.png" caption="Figure asdf" %}

We will focus on the group of GCs that are part of the Genomic Unknowns and also classified by anvi'o as SCGs.

{% include IMAGE path="/images/agnostos/agnostos-pan-5.png" caption="Figure asdf" %}

Let's see if we can learn something from these genes of unknown function, i.e., how are they distributed across the bacterial tree of life. Are they just restricted to a certain group of organisms or are they broadly distributed? We will use R for the last part of the tutorial and some of the data we generated for [Vanni et al., 2022](https://www.biorxiv.org/content/10.1101/2020.06.30.180448v5). Let's start cloning the project from:

{:.warning}
This next section will be data analysis in R and may require you to install multiple packages. Please follow the instructions [here]() to install the proper R environment. Also, please note that any statistical software can perform the same analyses including Python. 

```bash
git clone https://github.com/genomewalker/ebame6.git

cd ebame6/
```

If you have RStudio open the `.Rproj`, if not you will find the code we will use in this [file](https://github.com/genomewalker/ebame6/blob/main/scripts/ebame6-tutorial.R).

Follow the instructions in the [README](https://github.com/genomewalker/ebame6/) to start exploring the Unknown.

Let's import the data into R

```r
# Get the data from the anvi'o pan DB
# This data has is a combination of the results from the integration and anvi'o
# pan db summary

# FIXME: clearly this is the anvio pangenome summary file joined with agnostos annotations but I cannot find how it's made in the repo ebame6
agnostos_data <- read_tsv(file = "data/agnostos-pan-data.tsv.gz")

# Subset of Vanni et al., 2022 GTDB integrated data
agnostos_gtdb <- read_tsv(file = "data/agnostos-gtdb.tsv.gz")
```

Now let's explore those Single-copy core genes that are Genomic Unknowns by leveraging the metadata that comes with AGNOSTOS. One question we can ask is how many times a GU can be found in a specific taxonomic rank
:
```r
# Filter for GUs that are SCGs
gu_sgc <- agnostos_data %>%
  filter(category == "GU", SCG==1)

# Define taxonomic ranks
rnks <- c("phylum", "class", "order", "family", "genus")

# Let's calculate how many times a GC is seen in a specific rank
get_ranks <- function(df, rnk="genus"){
  df %>%
    select(cl_name, all_of(rnk)) %>%
    distinct() %>%
    group_by(cl_name) %>%
    count(name = "counts", sort = T) %>%
    mutate(rank = rnk)
}

gu_sgc_rnk_count <- map_dfr(rnks, function(X){get_ranks(df = agnostos_gtdb, rnk = X)}) %>%
  mutate(rank = fct_relevel(rank, rnks))

# As you can see, we have some of these genomic unknown SCGs that are broadly distributed over
# the bacterial tree of life, while others are pretty have a pretty narrow distribution
ggplot(gu_sgc_rnk_count,aes(rank, counts, group = cl_name)) +
  geom_line(size = 0.8, alpha = 0.7) +
  theme_light() +
  xlab("") +
  ylab("Counts")

# FIXME: add graph
```

Another insight AGNOSTOS can provide is how phylogenetically distributed a GC can be:

# FIXME: I am having trouble describing the differences between these two code blocks, they both feel like they are describing phylogenetic breadth to me

```r
# Which of these GCs have a more narrow distribution
agnostos_gtdb %>%
  filter(cl_name %in% gu_sgc$cl_name) %>%
  select(cl_name, genus, family) %>%
  distinct() %>%
  group_by(cl_name, family) %>%
  count(name = "n", sort = T) %>%
  inner_join(gu_sgc_rnk_count %>% filter(rank == "genus")) %>%
  mutate(diff = abs(n-counts)) %>%
  arrange(cl_name, family) %>%
  filter(diff == 0) %>%
  head(25) %>%
  knitr::kable()

# And with a broader distribution?
# How do you read the table?
# For example for 275106:
# The GC 275106 has been seen in 10 genera, which 1 is in Clostridiaceae,
# 3 in Enterococcaceae and so on..
agnostos_gtdb %>%
  filter(cl_name %in% gu_sgc$cl_name) %>%
  select(cl_name, genus, family) %>%
  distinct() %>%
  group_by(cl_name, family) %>%
  count(name = "n", sort = T) %>%
  inner_join(gu_sgc_rnk_count %>% filter(rank == "genus")) %>%
  mutate(diff = abs(n-counts)) %>%
  arrange(cl_name, family) %>%
  filter(diff > 0) %>%
  head(25) %>%
  knitr::kable()
```

```
# FIXME fix table formatting
``` 
```
| cl_name|family               |  n| counts|rank  | diff|
|  275106|Clostridiaceae       |  1|     10|genus |    9|
|  275106|Enterococcaceae      |  3|     10|genus |    7|
|  275106|Lactobacillaceae     |  3|     10|genus |    7|
|  275106|Listeriaceae         |  1|     10|genus |    9|
|  275106|Planococcaceae       |  1|     10|genus |    9|
|  275106|Staphylococcaceae    |  1|     10|genus |    9|
| 1285886|Aerococcaceae        | 12|     31|genus |   19|
| 1285886|Carnobacteriaceae    |  6|     31|genus |   25|
| 1285886|Enterococcaceae      | 13|     31|genus |   18|
| 1775574|Enterococcaceae      | 14|     19|genus |    5|
| 1775574|Listeriaceae         |  1|     19|genus |   18|
| 1775574|Streptococcaceae     |  4|     19|genus |   15|
| 2284829|1G12                 |  3|    997|genus |  994|
| 2284829|2-12-FULL-37-12      |  1|    997|genus |  996|
| 2284829|70-9                 |  2|    997|genus |  995|
| 2284829|A4b                  |  2|    997|genus |  995|
| 2284829|Acetivibrionaceae    |  3|    997|genus |  994|
| 2284829|Acholeplasmataceae   |  4|    997|genus |  993|
| 2284829|Acholeplasmataceae_A |  1|    997|genus |  996|
| 2284829|Acidaminobacteraceae |  1|    997|genus |  996|
| 2284829|Acidihalobacteraceae |  1|    997|genus |  996|
| 2284829|Actinomycetaceae     |  6|    997|genus |  991|
| 2284829|Aerococcaceae        |  7|    997|genus |  990|
| 2284829|Akkermansiaceae      |  1|    997|genus |  996|
| 2284829|Alkalibacillaceae    |  4|    997|genus |  993|
```

```
# FIXME: add output table  description
```

# Conclusion

```
# FIXME: need to review
```

The AGNOSTOS workflow from [Vanni et al., 2020](https://www.biorxiv.org/content/10.1101/2020.06.30.180448v1) provides a path forward to utilizing all genes from microbiomes and integrates seamlessly into your standard metagenomic or pangenomic workflow as seen above using anvi'o. We hope this brief introduction will catalyze your future microbiomes analyses!

Please refer to our recent preprint [Unifying the known and unknown microbial coding sequence space](https://www.biorxiv.org/content/10.1101/2020.06.30.180448v1) and [blog post](http://merenlab.org/2020/07/01/dark-side/#a-conceptual-framework-to-unify-the-known-and-the-unknown-in-microbiome-analyses) by [Antonio](http://orcid.org/0000-0002-8679-490X) for background and technical details on how to illuminate the genes of unknown function.

"To help you stop having to sweep the unknown under the carpet, we are simply removing the carpet." - Antonio


