---
layout: page
title: metapangenome [artifact]
categories: [anvio]
comments: false
redirect_from: /m/metapangenome
image:
  featurerelative: ../../../images/header.png
  display: true
---


{% include _toc.html %}


<img src="../../images/icons/CONCEPT.png" alt="CONCEPT" style="width:100px; border:none" />

A CONCEPT-type anvi'o artifact. This artifact is typically generated, used, and/or exported **by anvi'o** (and not provided by the user)..

ðŸ”™ **[To the main page](../../)** of anvi'o programs and artifacts.

## Provided by


<p style="text-align: left" markdown="1"><span class="artifact-p">[anvi-meta-pan-genome](../../programs/anvi-meta-pan-genome)</span></p>


## Required or used by


There are no anvi'o tools that use or require this artifact directly, which means it is most likely an end-product for the user.


## Description

A metapangenome is a way of comparing a bunch of genomes based both on the gene clusters they contain (like in a pangenome) and their abundances in different environments/samples (using metagenomic read recruitment).

Here is a [wonderful workflow](http://merenlab.org/data/prochlorococcus-metapangenome/) that takes you through the analysis of 31 *Prochlorococcus* isolate genomes and 93 TARA Oceans metagenomes, based on [this paper](https://peerj.com/articles/4320/). 


{:.notice}
Edit [this file](https://github.com/merenlab/anvio/tree/master/anvio/docs/artifacts/metapangenome.md) to update this information.

