---
layout: post
title: "An anvi'o workshop at the University of Chicago"
excerpt: "Just for <i>you</i>. Well, also for others. And totally free! Except you pay for your travel. But then it may be a lot of fun. MAYBE."
modified: 2017-03-23
tags: []
categories: [miscellaneous]
comments: true
authors: [meren]
image:
  feature: http://merenlab.org/images/miscellaneous/2017-03-23-anvio-workshop-at-uchicago/workshop.png
  display: false
---


{% capture images %}{{site.url}}/images/miscellaneous/2017-03-23-anvio-workshop-at-uchicago{% endcapture %}

![Workshop]({{images}}/workshop.png)

## Aftermath

On behalf of our lab, I thank everyone who took their time to come all the way here to attend to the workshop. During the two days we tried to cover a wide range of topics that spanned from metagenomic binning to the refinement of MAGs, reporting and analysis of single-nucleotide variants, and pangenomic analyses in anvi'o. Although how well we did to go through these topics is open to discussion, we probably were most successful at making the most out of the cozy space that was available to us:

![Workshop]({{images}}/first-day.jpg)

We all owe a big thanks to the Section of Gastroenterology, Hepatology, and Nutrition at the Department of Medicine for their support, and to our colleagues on the 9th floor of the KCBD for their patience with us -- not only we took the entire conference room for two full days, but we also occupied their lunch area on both days:

![Workshop]({{images}}/lunch.jpg)

If we do this again, we probably will try to do it in a bigger room, run it for more than two days, be more determined to stick with the schedule, and include more down time for data analyses and social interactions.

## Attendees

||Person|Affiliation|
|:--:|:--|:--|
|01|Nate Hubert|[Chang Lab](http://changlab.uchicago.edu/), Medicine, **University of Chicago**|
|02|[Sharon Grim](https://twitter.com/microgrim)|[Dick Lab](http://www.earth.lsa.umich.edu/geomicrobiology/), Earth and Environmental Sciences, **University of Michigan**|
|03|Ibrahim Farag|[Elshahed Lab](https://microbiology.okstate.edu/faculty/mostafa/elshahed.html), Microbiology and Molecular Genetic, **Oklahoma State University**|
|04|Alyson Yee|[Gilbert Lab](https://biomedsciences.uchicago.edu/page/jack-gilbert), Surgery, **University of Chicago**|
|05|Cesar Cardona|[Gilbert Lab](https://biomedsciences.uchicago.edu/page/jack-gilbert), Surgery, **University of Chicago**|
|06|Melissa Dsouza|[Gilbert Lab](https://biomedsciences.uchicago.edu/page/jack-gilbert), Surgery, **University of Chicago**|
|07|Tito David Peña Montenegro|[Joye Lab](http://joyeresearchgroup.uga.edu), Marine Sciences, **University of Georgia**|
|08|Orlando DeLeon|[Kirby Lab](http://www.mcw.edu/Microbiology/John-Kirby-PhD.htm), Microbiology and Immunology, **Medical College of Wisconsin**|
|09|Samantha Atkinson|[Kirby Lab](http://www.mcw.edu/Microbiology/John-Kirby-PhD.htm), Microbiology and Immunology, **Medical College of Wisconsin**|
|10|Jason Koval|[Antonopoulos Lab](http://www.igsb.org/labs/dionysios-antonopoulos), Biosciences Division, **Argonne National Laboratory**|
|11|Richard Kevorkian|[Lloyd Lab](http://lloydlab.utk.edu), Microbiology, **University of Tennessee**|
|12|Joy Buongiorno|[Lloyd Lab](http://lloydlab.utk.edu), Microbiology, **University of Tennessee**|
|13|Linnea Kop|[Jetten Lab](https://scholar.google.com/citations?user=iXjCKTgAAAAJ), Ecological Microbiology, **Radboud University Nijmegen**|
|14|Jessica Mark Welch|[Mark Welch Lab](http://www.mbl.edu/jbpc/staff/jmarkwelch/), **Marine Biological Laboratory**|
|15|[Sarah Stevens](http://sarahlrstevens.info)|[McMahon Lab](https://mcmahonlab.wisc.edu), Bacteriology, **University of Wisconsin-Madison**|
|16|[Benjamin Peterson](https://twitter.com/BigMicrobeBen)|[McMahon Lab](https://mcmahonlab.wisc.edu), Bacteriology, **University of Wisconsin-Madison**|
|17|Elizabeth McDaniel|[McMahon Lab](https://mcmahonlab.wisc.edu), Bacteriology, **University of Wisconsin-Madison**|
|18|Brooke Weigel|[Pfister Lab](http://pfisterlab.uchicago.edu), Ecology and Evolution, **University of Chicago**|
|19|Bing Ma|[Ravel Lab](http://ravel-lab.org), School of Medicine, **University of Maryland**|
|20|Johanna Holm|[Ravel Lab](http://ravel-lab.org), School of Medicine, **University of Maryland**|
|21|Pawel Gajer|[Ravel Lab](http://ravel-lab.org), School of Medicine, **University of Maryland**|
|22|Bruno Gomez-Gil Rodriguez Sala|[Sala Lab](http://www.ciad.mx/coordinaciones/unidad-mazatlan/personal/dr-bruno-gomez-gil-rodriguez-sala.html), Aquaculture, **Unidad Mazatlán en Acuicultura y Manejo Ambiental**|
|23|Bryan Merrill|[Sonnenburg Lab](http://sonnenburglab.stanford.edu), Microbiology and Immunology, **Stanford University**|
|24|Ann Gregory|[Sullivan Lab](http://u.osu.edu/viruslab), Civil, Env., and Geodetic Eng., **Ohio State University**|
|25|Hervé Tettelin|[Tettelin Lab](http://www.medschool.umaryland.edu/profiles/Tettelin-Herve/), Microbiology and Immunology, **University of Maryland**|
|26|Evan Kiefl|[Unaffiliated Grad Student](http://biophysics.uchicago.edu), Biophysical Sciences, **University of Chicago**|
|27|Veronika Kivenson|[Valentine Lab](http://methane.geol.ucsb.edu/Home.html), Earth Science, **UC Santa Barbara**|
|28|Devin Doud|[Woyke Lab](http://jgi.doe.gov/our-science/scientists-jgi/single-cells/), **Joint Genome Institute**|
|29|Robert Bowers|[Woyke Lab](http://jgi.doe.gov/our-science/scientists-jgi/single-cells/), **Joint Genome Institute**|
|30|Michael Dills|[Young Lab](http://plantsciences.montana.edu/facultyorstaff/faculty/young/), Microbiology and Immunology, **Montana State University**|

## The first anvi'o workshop at the UChicago

During the last year we traveled to [many places]({% post_url anvio/2016-08-18-events %}) to talk about [anvi'o](http://merenlab.org/software/anvio), and now we are inviting you to come to us.

This was possible thanks to the [Department of Medicine](http://medicine.uchicago.edu/), [Section of Gastroenterology, Hepatology, and Nutrition](http://medicine.uchicago.edu/gast/index.html), as well as [the Digestive Diseases Research Core Center (DDRCC)](http://www.uchicagoddrcc.org/) at the [University of Chicago](http://www.uchicago.edu). We are thankful for their support.

We will run a two-days anvi'o workshop at the University of Chicago on the **26<sup>th</sup> and 27<sup>th</sup> of April, 2017**. **As of today this is almost only a month away, so if you are planning to attend, you should act quickly**.

Here are some quick information for people who are still reading this page:

* **The workshop is free, but the space is limited** (if you are reading this, it means there is space, and if you are interested in coming, please send an e-mail). **We are planning to host about 30 people**.

* **Yes**, we indeed strategically planned this event to take place right after the [2017 Microbiome Center Research Symposium](https://www.eventbrite.com/e/2017-microbiome-center-research-symposium-registration-31297813603), so you can plan accordingly if you want and feel much better about your spectacular pragmatism.

* If you end up coming, in a cozy room with a great view, we will tell you about how to **visualize complex data** using anvi'o, as well as **working with genomes, metagenomes, metatranscriptomes, and pangenomes** in anvi'o. You will also hear about other things including **single-nucleotide variant detection and profiling, or refining of metagenome-assembled genomes**. Depending on the interest, there may be spontaneous discussions about any set of issues regarding 'omics data analysis and visualization.

* We are hoping this will be an opportunity to learn from each other in a friendly environment. Although we will keep things as simple and as clear as possible, **we encourage you to consider attending this workshop if you are a software developer or a power user** (because we won't hold back and will be ready to dive into the code whenever it is relevant).

* You are a perfect candidate to attend to this **if you have genomic / metagenomic / metatranscriptomic data, and you explore best ways to analyze them**. This workshop wouldn't be the best use of your time if your current endeavors exclusively revolve around marker gene amplicon data. We will have example datasets for you to follow everything on your anvi'o installed computers, but feel free to bring your own datasets. Meanwhile, certain steps of genomic / metagenomic data analysis takes time, and focusing on concepts and asking specific questions about how to apply some of them to your data may be much more useful way to spend your time.

* We will offer you lunch, snacks, water, tea, coffee, internet, and electricity (we also have chairs, desks, and restrooms), but you will need to take care of your travel (if you are coming from outside of Chicago). There is a [Hyatt Hotel in Hyde Park](https://chicagosouthuniversity.place.hyatt.com/en/hotel/home.html), and [Airbnb in Hyde Park](https://www.airbnb.com/s/Hyde-Park--Chicago--IL--United-States/homes?checkin=2017-04-26&checkout=2017-04-27&allow_override%5B%5D=&s_tag=epAeIoZ-) may be a good option (**if you will go with this option, do not book anything outside of Hyde Park area**). Or you can think about finding a place in the downtown area. It is not a terrible commute. [Alon](http://merenlab.org/people/) does it every day, ask him if you want.

* Although we are located in the Department of Medicine, we have interest in exploring the microbial life not only in mammalian guts or only in the context of health and disease, but also in oceans, soils, and other ecosystems. Do not feel that your environments or questions may be irrelevant. Ecology is happening everywhere, we are not afraid of anything, and you should just _bring it on_.

This will likely be a lot of fun. Or it will be "*just OK*". Talk to your supervisor, and show them a couple of anvi'o figures. Are you the supervisor? Well, then you know the right thing to do.

---

We will keep this page up-to-date as things progress, but please use the comments section to ask questions.

Thank you, and see you soon!

<blockquote class="twitter-tweet" data-lang="en"><p lang="en" dir="ltr">We just decided to do a 2-day <a href="https://twitter.com/hashtag/anvio?src=hash">#anvio</a> workshop next month in Chicago: <a href="https://t.co/x4BfJmuGQr">https://t.co/x4BfJmuGQr</a>. Please let your colleagues know.</p>&mdash; A. Murat Eren (@merenbey) <a href="https://twitter.com/merenbey/status/845258630173769731">March 24, 2017</a></blockquote>
<script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script>
