---
layout: page
title: Publicly available data for Eren et al. 2015 PeerJ
modified: 2015-07-30
excerpt: "An advanced analysis and visualization platform for 'omics data"
comments: true
image:
  feature: http://i.imgur.com/FjNWEGZ.png
---

{% include _toc.html %}

The manuscript, published on 2015-10-08: [Anvi'o: An advanced analysis and visualization platform for 'omics data](https://peerj.com/articles/1319/).

{:.notice}
The anvi'o profiles here will run with a much earlier version of anvi'o. If you would like to work with them, please checkout your anvi'o codebase to [this commit](https://github.com/meren/anvio/commit/265318856301ab4f7a71911aeaefb524339efbb3). Please don't hesitate to write us if you need assistance.

### Daily Infant Gut Samples by Sharon _et al._

_Raw data and anvi'o results for the section on supervised binning and the analysis of the variability in genome bins_.

* Visit [this address](http://umkk2268fc06.merenbey.koding.io:8080) to **try the anvi'o interactive interface** on the infant gut data, which provides the basis for **Figure 2**.
* You can [view the summary of the 13 bins](http://anvio.org/data/INFANT-CLC-SUMMARY-SUPERVISED), or you can [download the browsable output](http://anvio.org/data/INFANT-CLC-SUMMARY-SUPERVISED.tar.gz).
* [This Github repository](https://github.com/meren/anvio-methods-paper-analyses) gives access to the code that generates **Figure 3** (see the [relevant](https://github.com/meren/anvio-methods-paper-analyses/tree/master/SHARON_et_al/VARIABILITY_REPORTS) directory).
* You can download the output of `anvi-merge` (the merged profile db, and the annotation db) for the infant gut metagenomes [from here](http://dx.doi.org/10.6084/m9.figshare.1499236).

### Pensacola Beach Samples by Overholt _et al._ and Rodriguez-R _et al._

_Raw data and anvi'o results for the section on linking cultivar genomes with metagenomes_.

* While [this address](http://anvio.org/data/OVERHOLT-CULTIVARS-SUMMARY) gives access to the anvi'o summary of the ten cultivar genomes ([download](http://anvio.org/data/OVERHOLT-CULTIVARS-SUMMARY.tar.gz)), [this one](http://anvio.org/data/RODRIGUEZ-R-MG-SUMMARY) serves the 56 metagenomic bins ([download](http://anvio.org/data/RODRIGUEZ-R-MG-SUMMARY.tar.gz)) shown in **Figure 4**.
* You can download the output of `anvi-merge` for the mapping of metagenomes to Overholt cultivars [from here](http://dx.doi.org/10.6084/m9.figshare.1499234), and the output for `anvi-merge` for metagenomic bins is available [here](http://dx.doi.org/10.6084/m9.figshare.1499248).

### Gulf of Mexico Samples by Mason _et al._, and Yergeau _et al._

_Results for the section on linking metagenomes, metatranscriptomes, and single-cell genomes_.

* You can [view the summary of the two bins](http://anvio.org/data/MASON-SAGs-SUMMARY-SUPERVISED) identified in the assembly of the single-cell genomes ([download](http://anvio.org/data/MASON-SAGs-SUMMARY-SUPERVISED.tar.gz)) (**Figure 5 panel A**).
* You can [view the summary of the three bins](http://anvio.org/data/MASON-YERGEAU-MG-SUMMARY-SUPERVISED) identified in the metagenomic assembly ([download](http://anvio.org/data/MASON-YERGEAU-MG-SUMMARY-SUPERVISED.tar.gz)) (**Figure 5 panel B**).
* [This Github repository](https://github.com/meren/anvio-methods-paper-analyses) also gives access to the code that generates **Figure 5 panel C** (see the [relevant](https://github.com/meren/anvio-methods-paper-analyses/tree/master/MASON_et_al/SCATTER_PLOTS) directory).
* You can download the output of `anvi-merge` for the mapping of all samples against the assembly of SAGs [from here](http://dx.doi.org/10.6084/m9.figshare.1499235), and the output for `anvi-merge` for mapping to metagenomic contigs is available [here](http://dx.doi.org/10.6084/m9.figshare.1499246).


### Media and Supplementary files

* Tables: [Additional file 1](http://dx.doi.org/10.6084/m9.figshare.1499237), [Additional file 2](http://dx.doi.org/10.6084/m9.figshare.1499238), [Additional file 3](http://dx.doi.org/10.6084/m9.figshare.1499239).
* Figures: [Figure 1](http://dx.doi.org/10.6084/m9.figshare.1499240), [Figure 2](http://dx.doi.org/10.6084/m9.figshare.1499241), [Figure 3](http://dx.doi.org/10.6084/m9.figshare.1499242), [Figure 4](http://dx.doi.org/10.6084/m9.figshare.1499243), [Figure 5](http://dx.doi.org/10.6084/m9.figshare.1499244), [Figure S1](http://dx.doi.org/10.6084/m9.figshare.1499245), [Figure S2](http://dx.doi.org/10.6084/m9.figshare.1499247).




<div style="display: block; height: 200px;">&nbsp;</div>
