---
layout: page
noleftpanel: true
title: "Introduction to Popular `Omics Strategies"
author: "Course Plan"
authors: [iva, meren]
date: October, 2023
geometry: "left=3cm,right=3cm,top=2cm,bottom=2cm"
output: syllabus.pdf
colorlinks: true
urlcolor: blue
---

## Preface

The purpose of this document is to share the details of the course "**Introduction to Popular `Omics Strategies**". In the following sections you will find the course description, learning objectives, plan, schedule, attendance and grading policies, as well as other key information that is critical for the course attendees to consider.

## Course Details

This course is a part of the module "**Applied Molecular Ecology**" (**mar260**) taught by [Dr. Iva Veseli](https://hifmb.de/people/iva-veseli/) and [Prof. Dr. Iliana Baums](https://www.baumslab.org/).

|Course Details||
|:---|:---|
|Name|**Introduction to Popular ‘Omics Strategies**|
|Meeting Location|**[W16A 010](https://www.openstreetmap.org/#map=19/53.15263/8.16827)**|
|Number|5.12.262|
|Type|Lecture|
|Credits|2|
|Language|English|

In addition to this course, the module **mar260** contains three additional components that each participant of this course is expected to also attend,

* **Applied Microbial &#39;Omics** (5.12.263, Seminar / Exercise, Iva, Course Plan: [HTML](https://merenlab.org/courses/AMO/), [PDF](https://merenlab.org/courses/AMO/syllabus.pdf))
* **Coastal Conservation in the &#39;Omics Age** (5.12.260, Lecture, Iliana Baums)
* **Readings/Exercises in Coastal Conservation** (5.12.261, Seminar, Iliana Baums)

Please familiarize yourself with the details of the remaining courses using the appropriate documentation provided for the other components.

## Faculty and Communication

The lectures in the **Introduction to Popular &#39;Omics Strategies** will be primarily delivered by Iva. However, additional experts will take part in the design and/or delivery of various sections. The following table lists individuals who will be involved in the course, and their contact information:

|Name|Role|Expertise|Contact information|
|:---|:---|:---|:---|
|**Iva Veseli**|Postdoc|Microbial Ecology, Computer Science|iva.veseli@hifmb.de|
|**Meren**|Professor|Microbial Ecology, Computer Science|meren@hifmb.de|
|**Jessika Füssel**|Postdoc|Microbial Metabolism, Biogeochemistry|jessika.fuessel@uol.de|
|**Florian Trigodet**|Postdoc|Microbiology, Bioinformatics|florian.trigodet@hifmb.de|

Throughout the semester (and beyond) you can reach out via email with any question to Iva, who should be your first contact for anything related to the course activities unless specified otherwise anytime throughout the semester.

Iva's office is **Room 2118** at the Helmholtz Institute for Functional Marine Biodiversity. If you need to meet with her here in person, please schedule an appointment by email first.

## Description and Learning Objectives

Every ecological niche our planet has to offer is home to an astonishing number of microbial cells that form complex communities. The last several years witnessed tremendous advances in molecular and computational approaches which now offer unprecedented access to these communities through new &#39;omics strategies. Developing an overall understanding of these strategies, including the ability to identify their appropriate applications and shortcomings, has quietly become a de facto necessity in the journey of an independent life scientist. **The primary aim of this course** is to empower its participants and enable them to study the ecology, evolution, and functioning of naturally occurring microbial populations and recognize the current **conceptual framework** that helps us wrap our collective mind around the most diverse form of life on our planet.

**Introduction to Popular &#39;Omics Strategies** is a *lecture* with a companion *seminar*, **Applied Microbial &#39;Omics**. The lecture is designed to introduce its participants to the extent of microbial diversity on Earth and its impact on key biogeochemical processes, and strategies by which we characterize and study microbial life in naturally occurring systems to answer fundamental questions in microbial ecology and evolution. Participants will learn about the theoretical underpinnings of popular &#39;omics data types and their contemporary applications. These data types include **genomics** and **metagenomics** as well as various &#39;omics data analysis approaches such as **metabolic reconstruction** in genomes and metagenomes, **metagenomic read recruitment**, **pangenomics**, **phylogenomics**, and **microbial population genetics**.

The learning objectives of the course includes the following:

* To explain **microbial diversity** in naturally occurring systems and recognize mechanisms and outcomes of microbial **evolution**.

* To recognize data-enabled means to study **microbial ecosystems**.

* To introduce **state-of-the-art ‘omics approaches** and data types to characterize naturally occurring microbial diversity.

* To improve discussion, analytical, presentation and writing skills.


## Prerequisites

To maximize benefit, the participants of this course are expected to be familiar with the [central dogma of molecular biology](https://en.wikipedia.org/wiki/Central_dogma_of_molecular_biology), and be able to answer what is a **gene**, a **genome**, a **transcript**, or a **protein**, and have at least a **preliminary understanding of the principles in ecology and evolution**, such as the basics of **taxonomy** and broad ecological principles that maintain complex ecosystems.

The course will require its participants to read and understand contemporary literature written in English.

## Content Delivery

The primary mode of course content delivery will be through slides, where Iva will explain **core concepts**, **data types**, and **analysis strategies**. There will often be extensive discussions over these slides, which will require **active, verbal participation** by the attendees. Slides will be available on Stud.IP after each lecture.

After each lecture, the attendees will be given **suggested readings** from the recent literature that covers relevant topics and/or their real-world applications to contemporary questions in marine microbiology, oceanography, marine conservation, and beyond.

Please note that **preparation** and **participation** will play a key role in your success. For an effective learning experience please consider (1) taking a brief look at the suggested reading material of a given week in the course plan *before* coming to the lecture to familiarize yourself with the topic through online resources (~1 hour of study), (2) participate in the lecture actively by asking questions and attending discussions, and (3) read the suggested material carefully soon after the lecture (~3 hours of study).

### Suggested Readings: Sentence Highlights

Reading academic literature can be tedious (even for those who are used to it), and it is easy to read it passively without retaining any information. To help maintain an active mode of reading, please do the following:

Each time you read one of the suggested materials, highlight (or write down) **one sentence** that stands out to you. It can be something that intrigues you, confuses you, makes you excited, or raises a question. At the beginning of the next lecture, we will spend five minutes discussing one or two of these 'sentence highlights' together. Sharing your sentence highlights will be voluntary (but highly encouraged, as this will be an opportunity to clarify concepts and engage with the material in a way that follows your interests).

## Attendance Policy

Each participant is expected to attend each lecture in person (unless a legitimate reason for absence that is recognized by the University is in effect). The attendance will be recorded by a strategy that we call **class citizenship**, which aims to help the course director to have an overall understanding of the evolution of the course.

The class citizenship demands every participant to send a **class citizenship** email to iva.veseli@hifmb.de at the end of **each lecture**. The class citizenship email must be composed of two parts:

* A **brief summary** of the main concepts discussed during the class, interpreted by the attendee in their own words.

* A **short question** that is relevant to a concept or idea discussed during the lecture, yet remained unclear.

**The last 5 minutes of every lecture will be dedicated to class citizenship emails**, therefore the attendees will leave the class without having to remember doing it later. The class citizenship emails that are sent after the end of the class will not be taken into consideration as a mark of attendance.

The subject of the class citizenship email must follow this pattern **word-by-word**:

> **ITPOS Citizenship: DD/MM/YY**

For instance, the following would be the appropriate title for this email for the first lecture:

> **ITPOS Citizenship: 14/10/24**

The best class citizenship emails are those that are brief, genuine, and insightful. In an ideal world the emails should be no less than 50 words, and no more than 150 words. Please do not send notes you take throughout the class. You should use the last 5 minutes of the lecture to gather your thoughts, and come up with a summary of what you can remember. Here is an example class citizenship email:

> Summary: Today we discussed what is phylogenomics, how phylogenomic trees are built, and why single-copy core genes are suitable for building phylogenomic trees. We also discussed the relationship between phylogenetics, phylogenomics, and pangenomics with respect to the fraction of genome used and the evolutionary distance that they can cover.
>
> Question: Since phylogenomics and pangenomics are both useful for inferring evolutionary distances, it seems to me that integrating both methods in a systematic way would yield a more reliable tree. But it looks like the field only uses phylogenomics and pangenomics separately, is there a reason for that?


## Course Plan

Please note that each lecture takes place on **Mondays**, between **10:15 - 12:00**, at **W16A 010**.

### 14/10/24 :: Course logistics

* *Discussion Leaders*:

  * Iva

* *Discussion Topic*. **Course Logistics and Introductions**. We will meet each other and go over what to expect from the course. If we have time, we will start with **a brief introduction to microbial life on Earth**, and will continue next week.

* *Suggested Reading*:

  * The Course Syllabus :)


### 21/10/24 :: A brief introduction to **microbial life on Earth**

* *Discussion Leaders*:

  * Iva

* *Learning Objectives*:

  * Describe the extent of microbial diversity on Earth and microbial involvement in key biogeochemical processes, as well as human health and disease
  * Remember seminal studies that contributed to our understanding of the diversity, functioning, and metabolic potential of naturally occurring microbial communities and approaches to study them
  * Explain the old and new intellectual and technical challenges that prevent us from defining fundamental units of microbial life, and the *art* of moving forward without any answers

* *Suggested Reading*:

  * Gilbert JA, Neufeld JD (2014) [Life in a World without Microbes](https://doi.org/10.1371/journal.pbio.1002020). *PLoS Biology*.
  * Pace NR (2018) [The small things can matter](https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.3000009). *PLoS Biology*.
  * Falkowski, PG, Fenchel T, Delong EF (2008). [The microbial engines that drive Earth's biogeochemical cycles](https://doi.org/10.1126/science.1153213). *Science*.

* *Even More Suggested Resources for the Ambitious*:

  * [Seeing the Invisible](https://www.youtube.com/watch?v=mTzHxNMK0bU), Op-Docs, The New York Times (a short and cute video on microbial life for a lay audience).
  * [Meet Your Microbes](https://www.ted.com/talks/jonathan_eisen_meet_your_microbes), Jonathan Eisen, TED Talk (an easy-to-follow introduction to microbes for a lay audience).
  * [How Giant Tube Worms Survive at Hydrothermal Vents](https://www.youtube.com/watch?v=8W_ywzhkR90), Ed Yong, PBS Digital Studies (Ed Yong is a very successful science journalist, who talks about hydrothermal vent microbes with Colleen Cavanaugh, who made significant contributions to our understanding of microbial symbioses).
  * [Overview of how next-generation sequencing works](https://www.youtube.com/watch?v=mI0Fo9kaWqo), Eric Chow (a very useful and easy-to-follow lecture on the general principles of sequencing with Illumina, Oxford Nanopore, and PacBio).

### 28/10/24 :: An introduction to the **command-line environment**

* *Discussion Leaders*:

  * Iva

* *Learning Objectives*:

  * Apply basic commands for navigating and working in the command-line environment, specifically in the **Unix shell (BASH)**
  * Recognize **common vocabulary terms and errors** to make searching the Internet for help easier
  * Practice useful programming structures like **loops and variables** and techniques like **file searching and editing** that will be helpful in our data analysis exercises
  * Design a shell script to automate a practical task 

* *Suggested Reading*:

  * [Beginner's Guide to the Bash Terminal](https://www.youtube.com/watch?v=oxuRxtrO2Ag), Joe Collins, YouTube (a video introduction to the Linux command line environment).
  * [Learning the Shell](http://linuxcommand.org/lc3_learning_the_shell.php), William Shotts (a chapter from the open book "*The Linux Command Line*").
  * [The Unix Shell](https://swcarpentry.github.io/shell-novice/), Software Carpentry (an online tutorial)

### 04/11/24 :: An overview of **data-driven strategies** to survey environmental microbiomes

* *Discussion Leaders*:

  * Iva

* *Learning Objectives*:

  * Recognize currently available &#39;omics data types (such as metagenomics and metatranscriptomics), approaches (such as pangenomics and phylogenomics), and questions they can *and* can not answer
  * Recognize the available computational solutions to gain insights into fundamental questions in microbiology and their brief history
  * Explain the power of **metagenomic read recruitment** and interpret the ecological and evolutionary insights that we can infer through this strategy

* *Suggested Reading*:

  * Franzosa EA, et al (2015). [Sequencing and beyond: integrating molecular ‘omics’ for microbial community profiling](https://www.nature.com/articles/nrmicro3451). *Nature Reviews Microbiology*.
  * Eren AM and Banfield JF (2024). [Modern microbiology: Embracing complexity through integration across scales](https://www.cell.com/cell/fulltext/S0092-8674(24)00955-3). *Cell*.

### 11/11/24 :: **Genome-resolved metagenomics**: opportunities and pitfalls

* *Discussion Leaders*:

  * Iva

* *Learning Objectives*:

  * Recognize the difference between microbial isolates, enrichments, single-cell amplified genomes, and metagenome-assembled genomes
  * Explain the importance of the ability to acquire genomic information from microbes we have not yet cultivated
  * Summarize the basics of algorithms and strategies to reconstruct microbial genomes directly from metagenomes
  * Appreciate the limitations and opportunities associated with genome-resolved workflows

* *Suggested Reading*:

  * Paoli L, et al (2022). [Biosynthetic potential of the global ocean microbiome](https://www.nature.com/articles/s41586-022-04862-3). *Nature*.
  * Chen LX, et al (2020). [Accurate and complete genomes from metagenomes](https://genome.cshlp.org/content/30/3/315.full.pdf+html). *Genome Research*.
  * Shaiber A, Eren AM (2019). [Composite metagenome-assembled genomes reduce the quality of public genome repositories](https://doi.org/10.1128/mBio.00725-19). *mBio*.
  * Meren and Scott JJ (2020). [Visualizing the fate of contigs across metagenomic binning algorithms](https://merenlab.org/2020/01/02/visualizing-metagenomic-bins/). *A blog post on merenlab.org*.

### 18/11/24 :: **Pangenomics**: comparative genomics in the era of genomic explosion

* *Discussion Leaders*:

  * Iva

* *Learning Objectives*:

  * Explain the concepts of core and accessory genome, as well as open and closed pangenomes
  * Define gene clusters in pangenomes through sequence homology
  * Interpret the ecological and evolutionary insights that pangenomes offer

* *Suggested Reading*:

  * Tettelin H, et al (2005). [Genome analysis of multiple pathogenic isolates of Streptococcus agalactiae: Implications for the microbial "pan-genome"](https://doi.org/10.1073/pnas.0506758102). *PNAS*.
  * McInerney MO, et al (2017). [Why prokaryotes have pangenomes](https://doi.org/10.1038/nmicrobiol.2017.40). *Nature Microbiology*.
  * Zhou Z, et al (2018). [Pan-genome Analysis of Ancient and Modern Salmonella enterica Demonstrates Genomic Stability of the Invasive Para C Lineage for Millennia](https://doi.org/10.1016/j.cub.2018.05.058). *Current Biology*.

### 25/11/24 :: A Group Discussion Over Course Proposals

* *Discussion Leaders*:

  * Meren
  * Iva

* *Learning Objectives*:

  * Learn how to write a compelling proposal

### 02/12/24 :: **Phylogenomics**: inferring evolutionary relationships between microorganisms

* *Discussion Leaders*:

  * Iva

* *Learning Objectives*:

  * Identify commonly used genes, statistics, and heuristics to infer phylogenomic relationships across distantly related organisms
  * Recognize historical events that led to the emergence of the current Tree of Life, and why scientists can't even
  * Appreciate technical and theoretical limitations of inferring deep branching patterns confidently

* *Suggested Reading*:

  * Woese CR, Fox GE (1977). [Phylogenetic structure of the prokaryotic domain: The primary kingdoms](https://doi.org/10.1073/pnas.74.11.5088). *PNAS*.
  * Hug LA, et al (2016). [A new view of the tree of life](https://www.nature.com/articles/nmicrobiol201648). Nature Microbiology.
  * Shaiber A, et al (2020). [Functional and genetic markers of niche partitioning among enigmatic members of the human oral microbiome](https://doi.org/10.1186/s13059-020-02195-w). *Genome Biology*.
  * Gaïa M, et al (2023). [Mirusviruses link herpesviruses to giant viruses](https://www.nature.com/articles/s41586-023-05962-4). *Nature*.

* *Even More Suggested Reading for the Ambitious*:

  * Spang A, et al (2015). [Complex Archaea that bridge the gap between prokaryotes and eukaryotes](https://www.nature.com/articles/nature14447). *Nature*.
  * Da Cunha V, et al (2017). [Lokiarchaea are close relatives of Euryarchaeota, not bridging the gap between prokaryotes and eukaryotes](https://doi.org/10.1371/journal.pgen.1006810). *PLOS Genetics*.
  * Spang A, et al (2018). [Asgard archaea are the closest prokaryotic relatives of eukaryotes](https://doi.org/10.1371/journal.pgen.1007080). *PLOS Genetics*.

### 09/12/24 :: Inferring **microbial metabolism** in genomes and metagenomes

* *Discussion Leaders*:

  * Iva

* *Learning Objectives*:

  * Recognize the difference between microbial genes, functions, and metabolism.
  * Explain the ways by which microbial metabolism can be recovered from genomes and metagenomes
  * Tell the difference between understanding microbial diversity and understanding metabolic potential in a given environment

* *Suggested Reading*:

  * Watson AR, Füssel J, Veseli I, et al (2023). [Metabolic independence drives gut microbial colonization and resilience in health and disease](https://doi.org/10.1186/s13059-023-02924-x). *Genome Biology*.
  * van Kessel MAHJ, et al (2015). [Complete nitrification by a single microorganism](https://doi.org/10.1038/nature16459). *Nature*.
  * Liu R, et al (2022). [Novel Chloroflexi genomes from the deepest ocean reveal metabolic strategies for the adaptation to deep-sea habitats](https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-022-01263-6). *Microbiome*.

### 16/12/24 :: Linking **metabolomics and (meta)genomics**: opportunities and limitations

* *Discussion Leaders*:

  * Jessika
  * Iva

* *Learning Objectives*:

  * Recognize how measurements of metabolites from the environment are linked to metabolic pathways in genomes
  * Understand the limitations of combining metagenomics and metabolomics in simple and complex environments
  * Identify the limitations of metagenomics to gain insights into the chemical currencies of life

* *Suggested Reading*:

  * Moran MA and Kujawinski EB, et al (2022). [Microbial metabolites in the marine carbon cycle](https://rdcu.be/cKnTP). *Nature Microbiology*.
  * Kujawinski EB, et al (2023). [Metabolite diversity among representatives of divergent *Prochlorococcus* ecotypes](https://doi.org/10.1128/msystems.01261-22). *mSystems*.
  * Wienhausen G, et al (2017). [The exometabolome of two model strains of the *Roseobacter* group: a marketplace of microbial metabolites](https://www.frontiersin.org/articles/10.3389/fmicb.2017.01985/full). *Frontiers in Microbiology*.
  * Noriega-Ortega BE, et al (2019). [Does the chemodiversity of bacterial exometabolomes sustain the chemodiversity of marine dissolved organic matter?](https://doi.org/10.3389/fmicb.2019.00215). *Frontiers in microbiology*.

### 06/01/25 :: **Metapangenomics**: integrated interpretations of pangenomes and metagenomes

* *Discussion Leaders*:

  * Iva

* *Learning Objectives*:

  * Explain the emerging opportunities to investigate the functioning and the ecology of microbial populations by linking pangenomes and metagenomes
  * Comprehend the power of characterizing a single genome across metagenomes

* *Suggested Reading*:

  * Delmont TO, Eren AM (2018). [Linking pangenomes and metagenomes: the Prochlorococcus metapangenome](https://peerj.com/articles/4320/). *PeerJ*.
  * Utter DR, et al (2020). [Metapangenomics of the oral microbiome provides insights into habitat adaptation and cultivar diversity](https://doi.org/10.1186/s13059-020-02200-2). *Genome Biology*.
  * Boeuf D, et al (2021). [Metapangenomics reveals depth-dependent shifts in metabolic potential for the ubiquitous marine bacterial SAR324 lineage](https://doi.org/10.1186/s40168-021-01119-5). *Microbiome*.

### 13/01/25 :: **Microbial population genetics**: tools, terminology, and open questions

* *Discussion Leaders*:

  * Iva

* *Learning Objectives*:

  * Learn ecological and evolutionary implications of clonality and heterogeneity within environmental populations
  * Identify approaches to study single-nucleotide variants, and methods to reconstruct *haplotypes*
  * Comprehend differences and overlaps between population genetics approaches in animal populations and microbial populations
  * Characterize variation within a metagenomic sample and make use of it for exploratory analyses or hypothesis testing.

* *Suggested Reading*:

  * Simmons SL and Dibartolo G, et al (2008). [Population genomic analysis of strain variation in Leptospirillum group II Bacteria involved in acid mine drainage formation](https://doi.org/10.1371/journal.pbio.0060177). *PLOS Biology*.
  * Denef VJ (2018). [Peering into the genetic make up of natural microbial populations using metagenomics](https://doi.org/10.1007/13836_2018_14). *Springer Publishing*.
  * Delmont TO, et al (2019). [Single-amino acid variants reveal evolutionary processes that shape the biogeography of a global SAR11 subclade](https://doi.org/10.7554/eLife.46497). *eLife*.

### 20/01/25 :: **Genomic dynamism**: inversions, diversity generating retroelements, and more

* *Discussion Leaders*:

  * Florian
  * Iva

* *Learning Objectives*:

  * Learn about the processes of genomic diversification that are driven by biology, as opposed to stochasticity
  * Comprehend the site-specific nature of these processes and their associated mechanisms
  * Describe the approaches to find and report the dynamics of genomes across metagenomic samples

* *Suggested Reading*:

  * Chanin RB and West PT, et al (2024). [Intragenic DNA inversions expand bacterial coding capacity](https://www.nature.com/articles/s41586-024-07970-4). *Nature*.
  * Jiang X and Hall AB, et al (2019). [Invertible promoters mediate bacterial phase variation, antibiotic resistance, and host adaptation in the gut](https://doi.org/10.1126%2Fscience.aau5238). *Science*.
  * Paul BG, Eren AM (2022). [Eco-evolutionary significance of domesticated retroelements in microbial genomes](https://mobilednajournal.biomedcentral.com/articles/10.1186/s13100-022-00262-6). *Mobile DNA*.
  * Roux S, et al (2021). [Ecology and molecular targets of hypermutation in the global microbiome](https://www.nature.com/articles/s41467-021-23402-7). *Nature Communications*.


### 27/01/25 :: **Structure-informed** interpretations of microbial population genetics

* *Discussion Leaders*:

  * Iva

* *Learning Objectives*:

  * Learn about the new generation of computational strategies to predict protein structures from sequences
  * Comprehend the implications of structure-informed interpretations of genomic variation in our ability to determine targets of distinct evolutionary processes

* *Suggested Reading*:

  * Jumper J, et al (2021). [Highly accurate protein structure prediction with AlphaFold](https://www.nature.com/articles/s41586-021-03819-2). *Nature*.
  * AlQuraishi M (2021). [Protein-structure prediction revolutionized](https://www.nature.com/articles/d41586-021-02265-4). *Nature News and Views*.
  * Robinson SL (2023). [Structure-guided metagenome mining to tap microbial functional diversity](https://doi.org/10.1016/j.mib.2023.102382). *Current Opinion in Microbiology*.
  * Kiefl E, et al (2023). [Structure-informed microbial population genetics elucidate selective pressures that shape protein evolution](https://www.science.org/doi/full/10.1126/sciadv.abq4632). *Science Advances*.


## Evaluation and Grading

The evaluation of the attendee performance in this course (along with all the other three in the module "**Applied Molecular Ecology**" (**mar260**) will be based on two items to be returned by each attendee individually: (1) a **research pre-proposal** (which will provide the basis for the full proposals due at the end of class) and a final **research proposal**.

Writing research proposals provides one with a critical skill that will be useful regardless of the profession one chooses to pursue after an undergraduate education. The primary purpose of a research proposal is to persuade others that *your* idea (in this case, science) is worthy of committing resources to, and you are the best person to implement it. The final research proposals will be graded based on their ability to address the following questions:

* Does the proposal contain **a novel research objective** that is relevant to the "**Applied Molecular Ecology**" (**mar260**) module content?
* Does it make a **strong case** given what is **known** and what is **unknown**?
* Does it adequately and accurately **cite the existing literature**?
* Does it suggest the use of methods that are **relevant** and **effective** to address the research question?
* Does the investigator demonstrate their **ability to use the proposed methods**?
* Does the proposal includes a discussion of **expected outcomes**, **potential risks**, and how to **mitigate** risks?
* Does it include a **reasonable budget** and a **meaningful timeline** to carry out the proposed research objectives?
* Does the proposal written in a **clear**, **concise**, and **accurate** manner that is expected of scientific work?

Upon the submission of the pre-proposal, smaller teams of students will serve as a "peer-review panel" and discuss each pre-proposal. The names of participants who wrote these proposals will not be known to the class. The panel will take place in class for all participants. The purpose of the panel is to provide feedback on how well the proposal was able to convey ideas. Participants are expected to improve their short pre-proposals based on the feedback they received from the peer-review panels and submit their full proposal as their final project.

**The participants will be provided with detailed instructions on how to come up with a proposal idea, what resources are available to carry out the projects, how to format the pre-proposal and final proposals, and how to evaluate them in panel.**

It is challenging to think of a topic for research that can be done in a short amount of time with limited resources. Therefore, the course directors (Iva and Iliana) will meet with each of the course participants individually to discuss their proposal idea, tell them whether it is feasible, and help them with resources. To initiate this discussion, you will hand in a one-paragraph research idea early in the semester.

The grading scale for this module is as follows:

|Grade|Threshold|
|:---|:---|
|1.0	|95% |
|1.3	|90% |
|1.7	|85% |
|2.0	|80% |
|2.3	|75% |
|2.7	|70% |
|3.0	|65% |
|3.3	|60% |
|3.7	|55% |
|4.0	|50% |

### Summary of important deadlines

* week of **26.11.24** :: we discuss the course proposals in class. Individual meetings with Iliana and Iva to discuss proposal ideas take place outside of class.
* **10.12.24** :: one paragraph research idea is due.
* **07.01.25** :: pre-proposal is due
* week of **20.01.25** :: peer-review panel takes place in class
* **15.02.25** :: final proposal is due
* **31.03.25** :: end of semester (grades are due)

## Examination Policy

Please find all relevant university policies here: [https://uol.de/studiengang/pruefungen/umweltwissenschaften-fach-bachelor-136](https://uol.de/studiengang/pruefungen/umweltwissenschaften-fach-bachelor-136)

## Academic Integrity
**All University policies regarding academic integrity, ethics and honorable behavior apply to this course.** Academic integrity is the pursuit of scholarly activity free from fraud and deception and is an educational objective of this class. Academic dishonesty includes, but is not limited to, cheating, plagiarizing, fabricating of information or citations, facilitating acts of academic dishonesty by others, having unauthorized possession of examinations, submitting work of another person or work previously used without informing the instructor, or tampering with the academic work of other students. For any material or ideas obtained from other sources, such as the course reading materials or things you see on the web, in the library, etc., a source reference must be given. Direct quotes from any source must be identified as such.

## Disability Access Statement

UOL welcomes students with disabilities and students with care obligations for their children or close relatives into the University's educational programs. In order to receive consideration for reasonable accommodations, you must contact the *Prüfungsausschuss*. Please let Iva and/or Iliana know at the beginning of the semester what accommodations were approved for you.
