---
layout: publications
modified: 2024-09-20
comments: false
image:
   display: true
   feature: eel-pond.jpg
---

<script type='text/javascript' src='https://d1bxh8uas1mnw7.cloudfront.net/assets/embed.js'></script>

<script async src="https://badge.dimensions.ai/badge.js" charset="utf-8"></script>

{:.notice}

This page lists publications that are most reflective of our interests. For a complete list, please see <a href='https://scholar.google.com/citations?user=GtLLuxoAAAAJ&view_op=list_works&sortby=pubdate' target='_blank'>Meren's Google Scholar page</a>.

## 2024

<a id="10.1016/j.cell.2024.08.028">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1016/j.cell.2024.08.028"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1016/j.cell.2024.08.028" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1016/j.cell.2024.08.028" target="_new">Modern microbiology: Embracing complexity through integration across scales</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Banfield JF</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/eren-banfield-cell-cover.jpg"><img src="/images/pubs/eren-banfield-cell-cover.jpg" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A review article that considers an inevitably incomplete list of emergent themes in microbiology and highlights those that are among the archetypes of its modern era.</span><br><span style="display: inline-block; padding-bottom: 5px;">- In which we claimed that <i>microbiology is arguably the most important science of the mid-21st century</i> and survived to tell the story.</span><br><span style="display: inline-block; padding-bottom: 5px;">- This paper was also the cover of <i>Cell</i>'s <a href="https://www.cell.com/cell/issue?pii=S0092-8674(23)X0020-8" target="_blank">anniversary focus issue</a> on microbes. See this <a href="https://merenlab.org/story-of-a-cover/" target="_blank">blog post</a> by Meren to see the evolution of the cover.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Cell</b>, 187(19):5151-5170 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Modern+microbiology:+Embracing+complexity+through+integration+across+scales" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1016/j.cell.2024.08.028" target="_blank">doi:10.1016/j.cell.2024.08.028</a></span>
</div>

<a id="10.1038/s41597-024-03778-z">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/s41597-024-03778-z"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/s41597-024-03778-z" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1038/s41597-024-03778-z" target="_new">Digital Microbe: A genome-informed data integration framework for collaborative research on emerging model organisms</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Veseli I<sup>☯</sup></span>, Cooper ZS<sup>☯</sup>, DeMers MA<sup>☯</sup>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Schechter MS</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Miller SE</span>, Weber L, Smith CB, Rodriguez LT, Schroer WF, McIlvin MR, Lopez PZ, Saito M, Dyhrman S, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM<sup>‡</sup></span>, Moran MA<sup>‡</sup>, Braakman R<sup>‡</sup></span>
    <span class="pub-co-first-authors"><sup>☯</sup>Co-first authors; <sup>‡</sup>Co-senior authors</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/veseli_cooper_demers_et_al_2024.png"><img src="/images/pubs/veseli_cooper_demers_et_al_2024.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- Description of a data-driven concept, 'Digital Microbe', that offers decentralized, reproducible, and interoperable means for multi-investigator teams to work on the same model organism collaboratively.</span><br><span style="display: inline-block; padding-bottom: 5px;">- The need and solution emerged as a funciton of our collaborative reserach in <a href="https://ccomp-stc.org/" target="_blank">C-CoMP</a>, the NSF-funded Science and Technology Center for Currencies of a Microbial Planet, and this publication aims to communicate our experience in this front to the community.</span><br><span style="display: inline-block; padding-bottom: 5px;">- The current implementation of this concept is simply a set of citable <a href="https://anvio.org">anvi'o</a> data products, such as <a href="https://anvio.org/help/main/artifacts/contigs-db/" target="_blank">contigs-db</a> and <a href="https://anvio.org/help/main/artifacts/pan-db/" target="_blank">pan-db</a> files, shared on public repositories, such as Zeonodo or FigShare.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Reproducible bioinformatics workflows are available for the <a href="https://github.com/C-CoMP-STC/digital-microbe/blob/main/R_pom_reproducible_workflow.md">generation of the <i>Ruegeria pomeroyi</i> digital microbe</a> and <a href="https://github.com/C-CoMP-STC/digital-microbe/tree/main/alteromonas/pangenomeCuration">curation of the <i>Altermonas</i> pangenome</a> as examples covered in the paper.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Scientific Data</b>, 11(967) | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Digital+Microbe:+A+genome-informed+data+integration+framework+for+collaborative+research+on+emerging+model+organisms" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/s41597-024-03778-z" target="_blank">doi:10.1038/s41597-024-03778-z</a></span>
</div>

<a id="10.1038/d41586-024-02661-6">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/d41586-024-02661-6"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/d41586-024-02661-6" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://rdcu.be/dSZVz" target="_new">Bioprospecting marine microbial genomes to improve biotechnology</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, <span class="pub-collaborator-author" title='A key collaborator of the lab at the time of publication'>Delmont TO</span></span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/eren_and_delmont_et_al_2024.png"><img src="/images/pubs/eren_and_delmont_et_al_2024.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A Nature News & Views piece to discuss the implications of the article published byaChen, Jia, Sun, and Liu et al., "<a href='https://www.nature.com/articles/s41586-024-07891-2'>Global marine microbial diversity and its potential in bioprospecting</a>".</span><br><span style="display: inline-block; padding-bottom: 5px;">- We point out that the ever-increasing number of environmental genomes reported from marine systems are ready for bioprospecting efforts and to those who wish to push the boundaries of microbiology by connecting its classical and modern means.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Nature</b> | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Bioprospecting+marine+microbial+genomes+to+improve+biotechnology" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/d41586-024-02661-6" target="_blank">doi:10.1038/d41586-024-02661-6</a></span>
</div>

<a id="10.1016/j.cell.2024.01.039">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1016/j.cell.2024.01.039"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1016/j.cell.2024.01.039" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1016/j.cell.2024.01.039" target="_new">A cryptic plasmid is among the most numerous genetic elements in the human gut</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Fogarty EC</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Schechter MS</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Lolans K</span>, Sheahan ML, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Veseli I</span>, Moore RM, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Kiefl E</span>, Moody T, Rice PA, <span class="pub-collaborator-author" title='A key collaborator of the lab at the time of publication'>Yu MK</span>, Mimee M, Chang EB, Ruscheweyh H, Sunagawa S, Mclellan SL, Willis AD, Comstock LE, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span></span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/data/pBI143/images/pBI143_across_the_globe.png"><img src="/data/pBI143/images/pBI143_across_the_globe.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A study that describes one of the <b>most prevalent and numerous cryptic plasmid</b> in the gut microbiomes of people who live in the industrialized world that is composed of only two genes (for its own replication and mobilization) in its native form.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Here is <a href="https://twitter.com/merenbey/status/1763282323309490546" target="_blank">a Twitter thread</a> that explains some of the interesting aspects of pBI143 ecology as well as the practical implications having <b>a human gut-specific and highly conserved genetic entity, copy-number of which responds to stress</b>.</span><br><span style="display: inline-block; padding-bottom: 5px;">- You can find the plasmid sequences to look for pBI143 in your metagenomes, reproducible data items to re-investigate metagenomic read recruitment results, and bioinformatics workflows to elucidate population genetics of pBI143 <a href="/data/pBI143">here</a>.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Cell</b>, 187(5):1206-1222.e16 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=A+cryptic+plasmid+is+among+the+most+numerous+genetic+elements+in+the+human+gut" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1016/j.cell.2024.01.039" target="_blank">doi:10.1016/j.cell.2024.01.039</a></span>
</div>

<a id="10.1038/s41564-024-01610-3">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/s41564-024-01610-3"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/s41564-024-01610-3" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1038/s41564-024-01610-3" target="_new">Diverse plasmid systems and their ecology across human gut metagenomes revealed by PlasX and MobMess</a></span>
    <span class="pub-authors"><span class="pub-collaborator-author" title='A key collaborator of the lab at the time of publication'>Yu MK<sup>☯</sup></span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Fogarty EC<sup>☯</sup></span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span></span>
    <span class="pub-co-first-authors"><sup>☯</sup>Co-first authors</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/plasmid_systems.png"><img src="/images/pubs/plasmid_systems.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A study that aims to shed light on <b>the ecology and evolution of one of the most critical yet poorly studied aspects of microbial life -- naturally occurring plasmids</b>.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Uses state-of-the-art machine learning strategies to identify <b>over 60,000 plasmids</b> from human gut metagenomes, which represents a <b>200-fold increase</b> in the number of known plasmids to date that were detectable in healthy humans.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Defines hundreds of '<b>plasmid systems</b>', and demonstrates that naturally occurring plasmids are not static entities, but <b>their evolution is driven by the need to respond to the environment, and their ecology cannot be simply explained by bacterial taxonomy and distribution patterns of their putative hosts</b>.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Here is <a href="https://twitter.com/merenbey/status/1764675816204513473" target="_blank">a Twitter thread</a> that goes through some of the interesting aspects if this work.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Nature Microbiology</b>, 9:830-847 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Diverse+plasmid+systems+and+their+ecology+across+human+gut+metagenomes+revealed+by+PlasX+and+MobMess" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/s41564-024-01610-3" target="_blank">doi:10.1038/s41564-024-01610-3</a></span>
</div>

<a id="10.1038/s41564-023-01560-2">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/s41564-023-01560-2"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/s41564-023-01560-2" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://rdcu.be/du8Up" target="_new">Dietary- and host-derived metabolites are used by diverse gut bacteria for anaerobic respiration</a></span>
    <span class="pub-authors">Little AS, Younker IT, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Schechter MS</span>, Bernardino PN, Méheust R, Stemczynski J, Scorza K, Mullowney MW, Sharan D, Waligurski E, Smith R, Ramanswamy R, Leiter W, Moran D, McMillin M, Odenwald MA, Iavarone AT, Sidebottom AM, Sundararajan A, Pamer EG, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Light SH</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/little_et_al_2024.png"><img src="/images/pubs/little_et_al_2024.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A study that discovers three distinct taxa that encode over fifty respiratory reductases per genome that enable the use of a diverse array of metabolites as electron acceptors, establishing a respiratory strategy that utilizes the metabolite pool of the anaerobic human gut environment.</span><br><span style="display: inline-block; padding-bottom: 5px;">- An application of <a href="https://anvio.org/help/main/workflows/ecophylo/" target="_blank">EcoPhylo</a> to recover reductases from genomes and explain their phylogeny.</span><br><span style="display: inline-block; padding-bottom: 5px;">- A news article on this work from Matt Wood is also available <a href="https://biologicalsciences.uchicago.edu/news/surprisingly-resourceful-ways-bacteria-thrive-human-gut" target="_blank">here</a>.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Nature Microbiology</b>, 9:55-69 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Dietary-+and+host-derived+metabolites+are+used+by+diverse+gut+bacteria+for+anaerobic+respiration" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/s41564-023-01560-2" target="_blank">doi:10.1038/s41564-023-01560-2</a></span>
</div>


## 2023

<a id="10.1126/sciadv.abq4632">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1126/sciadv.abq4632"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1126/sciadv.abq4632" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1126/sciadv.abq4632" target="_new">Structure-informed microbial population genetics elucidate selective pressures that shape protein evolution</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Kiefl E</span>, Esen ÖC, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Miller SE</span>, Kroll KL, Willis AD, Rappé MS, Pan T, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span></span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/kiefl_et_al_metagenomics_plus_protein_structures.png"><img src="/images/pubs/kiefl_et_al_metagenomics_plus_protein_structures.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A study that describes an approach to integrate <b>environmental microbiology</b> with recent advances in <b>protein structure prediction</b>, and illustrates the tight association between <b>intra-population genetic variants</b>, <b>environmental selective pressures</b>, and <b>structural properties of proteins</b>.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Demonstrates a quantifiable link between (1) the magnitude of selective pressures over key metabolic <b>genes</b> (e.g., glutamine synthase of the central nitrogen metabolism), (2) the availability of key <b>nutrients</b> in the environment (e.g., nitrate), and (3) the maintenance of nonsynonymous <b>variants</b> near protein active sites.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Comes with a <a href="https://merenlab.org/data/anvio-structure/chapter-I/"><b>reproducible bioinformatics workflow</b></a> that offers detailed access to computational steps used in the study that spans from metagenomic read recruitment and profiling to the integration of environmental variants and predicted protein structures.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Science Advances</b>, 9(8):eabq4632 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Structure-informed+microbial+population+genetics+elucidate+selective+pressures+that+shape+protein+evolution" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1126/sciadv.abq4632" target="_blank">doi:10.1126/sciadv.abq4632</a></span>
</div>

<a id="10.7554/eLife.89862">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.7554/eLife.89862"></div>
<div class="__dimensions_badge_embed__" data-doi="10.7554/eLife.89862" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.7554/eLife.89862" target="_new">Microbes with higher metabolic independence are enriched in human gut microbiomes under stress</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Veseli I</span>, Chen YT, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Schechter MS</span>, Vanni C, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Fogarty EC</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Watson AR</span>, Jabri B, Blekhman R, Willis AD, <span class="pub-collaborator-author" title='A key collaborator of the lab at the time of publication'>Yu MK</span>, Fernàndez-Guerra A, <span class="pub-collaborator-author" title='A key collaborator of the lab at the time of publication'>Füssel J</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span></span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/veseli_et_al_ibd_metabolism.png"><img src="/images/pubs/veseli_et_al_ibd_metabolism.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A study of microbial metabolic enrichment in human gut metagenomes that shows high metabolic independence (HMI) is a distinguishing characteristic of microbial communities associated with individuals diagnosed with IBD.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Furthermore, it shows that the enrichment of metabolic features that are predictive of HMI and that were enriched in IBD were also enriched in gut microbiome following antibiotic treatment, suggesting that HMI is a hallmark of microbial communities in stressed gut environments.</span><br><span style="display: inline-block; padding-bottom: 5px;">- An insight article written by Vanessa Rossetto Marcelino, <a href="https://elifesciences.org/articles/92319">Gut Health&#58; The value of connections</a>, accompanies this work with additional perspectives.</span><br><span style="display: inline-block; padding-bottom: 5px;">- <a href="https://elifesciences.org/reviewed-preprints/89862/reviews">Peer reviews</a>. <a href="https://merenlab.org/data/ibd-gut-metabolism/">Reproducible bioinformatics workflow</a>.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>eLife</b>, 12(RP89862) | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Microbes+with+higher+metabolic+independence+are+enriched+in+human+gut+microbiomes+under+stress" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.7554/eLife.89862" target="_blank">doi:10.7554/eLife.89862</a></span>
</div>

<a id="10.1186/s13059-023-02924-x">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1186/s13059-023-02924-x"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1186/s13059-023-02924-x" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1186/s13059-023-02924-x" target="_new">Metabolic independence drives gut microbial colonization and resilience in health and disease</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Watson AR<sup>☯</sup></span>, <span class="pub-collaborator-author" title='A key collaborator of the lab at the time of publication'>Füssel J<sup>☯</sup></span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Veseli I<sup>☯</sup></span>, DeLongchamp JZ, Silva M, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Trigodet F</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Lolans K</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Shaiber A</span>, Fogarty E, Runde JM, Quince C, <span class="pub-collaborator-author" title='A key collaborator of the lab at the time of publication'>Yu MK</span>, Söylev A, Morrison HG, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Lee STM</span>, Kao D, Rubin DT, Jabri B, Louie T, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span></span>
    <span class="pub-co-first-authors"><sup>☯</sup>Co-first authors</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/watson_et_al_fmt.jpeg"><img src="/images/pubs/watson_et_al_fmt.jpeg" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A Fecal Microbiota Transplantation (FMT) study that reveals unexpected parallels between the <b>adaptive ecological processes</b> that shape the recipient gut <b>microbial composition after FMT</b> and those that influence <b>microbial diversity in patients with Inflammatory Bowel Disease (IBD)</b>.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Includes an observation that links <b>the presence of superior metabolic competence in bacterial populations to their expansion in IBD</b>.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Here is a <a href="https://twitter.com/awatsontxt/status/1648420614737371136">Twitter thread</a> by Andrea that explains the key points of the study, and here is <a href="https://twitter.com/merenbey/status/1648848404456390658">another one</a> by Meren that details what are the critical learnings from it.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Genome Biology</b>, 24(78) | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Metabolic+independence+drives+gut+microbial+colonization+and+resilience+in+health+and+disease" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1186/s13059-023-02924-x" target="_blank">doi:10.1186/s13059-023-02924-x</a></span>
</div>

<a id="10.1038/s41592-023-02025-4">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/s41592-023-02025-4"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/s41592-023-02025-4" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://rdcu.be/dopGM" target="_new">Microbial-enrichment method enables high-throughput metagenomic characterization from host-rich samples</a></span>
    <span class="pub-authors">Wu-Woods NJ, Barlow JT, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Trigodet F</span>, Shaw DG, Romano AE, Jabri B, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Ismagilov RF</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/wo-woods_et_al_mem.png"><img src="/images/pubs/wo-woods_et_al_mem.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- <b>A microbial-enrichment protocol</b> that removes nucleic acids from complex samples that belong to the eukaryotic host, while not substantially perturbing the host associated microbial community composition.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Reduce host DNA from a wide range of sample types (including human saliva, stool, intestinal scrapings, and intestinal mucosal biopsies) <b>more than 1,000 fold</b> with negligible changes in microbial community structures.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Enables comprehensive surveys of microbial populations using shotgun metagenomics even in samples that have been extremely challenging to study due to extensive host contamination.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Nature Methods</b> | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Microbial-enrichment+method+enables+high-throughput+metagenomic+characterization+from+host-rich+samples" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/s41592-023-02025-4" target="_blank">doi:10.1038/s41592-023-02025-4</a></span>
</div>


## 2022

<a id="10.1038/s41564-022-01090-3">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/s41564-022-01090-3"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/s41564-022-01090-3" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://rdcu.be/cKnTP" target="_new">Microbial metabolites in the marine carbon cycle</a></span>
    <span class="pub-authors">Moran MA<sup>‡</sup>, Kujawinski EB<sup>‡</sup>, Schroer WF, Amin SA, Bates NR, Bertrand EM, Braakman R, Brown CT, Covert MW, Doney SC, Dyhrman ST, Edison AS, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Levine NM, Li L, Ross AC, Saito MA, Santoro AE, Segrè D, Shade A, Sullivan MB, Vardi A</span>
    <span class="pub-co-first-authors"><sup>‡</sup>Co-senior authors</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/moran_et_al_carbon_cycle.png"><img src="/images/pubs/moran_et_al_carbon_cycle.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A review that happens to be <b>the study number 001</b> for the <a href='https://ccomp-stc.org/' target='_blank'>Center for Chemical Currencies of a Microbial Planet</a>, or C-CoMP, an NSF-funded Science and Technology Center that aims to promote 'a deeper understanding of chemicals and chemical processes that underpin ocean ecosystems and the global carbon cycle by leveraging recent advances in analytical and data sciences, incorporating new ocean sampling technologies, using an open-science framework, and engaging educators and policy-makers'.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Nature Microbiology</b>, 7:508–523 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Microbial+metabolites+in+the+marine+carbon+cycle" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/s41564-022-01090-3" target="_blank">doi:10.1038/s41564-022-01090-3</a></span>
</div>

<a id="10.1186/s13100-022-00262-6">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1186/s13100-022-00262-6"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1186/s13100-022-00262-6" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1186/s13100-022-00262-6" target="_new">Eco-evolutionary significance of domesticated retroelements in microbial genomes</a></span>
    <span class="pub-authors">Paul BG, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span></span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/dgrs-in-metagenomes.png"><img src="/images/pubs/dgrs-in-metagenomes.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A short review on retrons and diversity-generating retro elements, some of the most beautiful and mysterious ways for life to beat the boring means of evolution and skip ahead.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Demonstrates a workflow that gives access to the extent intra-population hypervariability of DGRs and their ecology through the analysis of metagenomes.</span><br><span style="display: inline-block; padding-bottom: 5px;">- A complete bioinformatics workflow that uses anvi'o and oligotyping to study DGR activity in metageomes is <a href="https://merenlab.org/tutorials/dgrs-in-metagenomes/" target="_blank">available here</a>.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Mobile DNA</b>, 13(6) | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Eco-evolutionary+significance+of+domesticated+retroelements+in+microbial+genomes" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1186/s13100-022-00262-6" target="_blank">doi:10.1186/s13100-022-00262-6</a></span>
</div>

<a id="10.1016/j.molcel.2021.12.007">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1016/j.molcel.2021.12.007"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1016/j.molcel.2021.12.007" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1016/j.molcel.2021.12.007" target="_new">tRNA modification dynamics from individual organisms to metaepitranscriptomics of microbiomes</a></span>
    <span class="pub-authors">Zhang W, Foo M, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Pan T</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/zhang_et_al_2022.png"><img src="/images/pubs/zhang_et_al_2022.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A <a href="/files/Zhang_et_al_2022.pdf">review</a> that summarizes recent advances in the studies of <b>tRNA modification dynamics</b> in biological processes</span><br><span style="display: inline-block; padding-bottom: 5px;">- Defines '<b>metaepitranscriptomics</b>' as a strategy to study modification dynamics in complex environmental populations.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Molecular Cell</b> | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=tRNA+modification+dynamics+from+individual+organisms+to+metaepitranscriptomics+of+microbiomes" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1016/j.molcel.2021.12.007" target="_blank">doi:10.1016/j.molcel.2021.12.007</a></span>
</div>


## 2021

<a id="10.1038/s41564-020-00834-3">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/s41564-020-00834-3"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/s41564-020-00834-3" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://rdcu.be/cKMmi" target="_new">Community-led, integrated, reproducible multi-omics with anvi'o</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Kiefl E</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Shaiber A</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Veseli I</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Miller SE</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Schechter MS</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Fink I</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Pan JN</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Yousef M</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Fogarty EC</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Trigodet F</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Watson AR</span>, Esen ÖC, Moore RM, Clayssen Q, Lee MD, Kivenson V, Graham ED, Merrill BD, Karkman A, Blankenberg D, Eppley JM, Sjödin A, Scott JJ, Vázquez-Campos X, McKay LJ, McDaniel EA, Stevens SLR, Anderson RE, <span class="pub-collaborator-author" title='A key collaborator of the lab at the time of publication'>Füssel J</span>, Fernandez-Guerra A, Maignien L, <span class="pub-collaborator-author" title='A key collaborator of the lab at the time of publication'>Delmont TO</span>, Willis AD</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/anvio-integrated-omics.png"><img src="/images/pubs/anvio-integrated-omics.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A summary of the progress of <a href="https://anvio.org" target="_blank">anvi'o</a> during the past five years.</span><br><span style="display: inline-block; padding-bottom: 5px;">- The <a href='https://anvio.org/files/anvio-2021.pdf'>PDF</a> of the publication (which was supposed to be open access, yet somehow the Nature Publishing Group was unable to do so, so here it is).</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Nature Microbiology</b>, 6(1):186 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Community-led,+integrated,+reproducible+multi-omics+with+anvi'o" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/s41564-020-00834-3" target="_blank">doi:10.1038/s41564-020-00834-3</a></span>
</div>

<a id="10.1111/1755-0998.13588">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1111/1755-0998.13588"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1111/1755-0998.13588" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1111/1755-0998.13588" target="_new">High molecular weight DNA extraction strategies for long-read sequencing of complex metagenomes</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Trigodet F</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Lolans K</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Fogarty EC</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Shaiber A</span>, Morrison HG, Barreiro L, Jabri B, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span></span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/Trigodet_and_Lolans_et_al_HMW.png"><img src="/images/pubs/Trigodet_and_Lolans_et_al_HMW.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A study that <b>benchmarks six high molecular weight DNA extraction strategies</b> (commercially available kits, phenol-chloroform extraction, and agarose encasement followed by agarase digestion) <b>for long-read sequencing of metagenomes</b> with MinION.</span><br><span style="display: inline-block; padding-bottom: 5px;">- It turns out the protocol that works best for sequencing DNA from microbial isolates may not be the most effetive method for long-read sequencing of metagenomes ¯\_(ツ)_/¯</span><br><span style="display: inline-block; padding-bottom: 5px;">- A reproducible <b>bioinformatics workflow</b> is available <a href="https://merenlab.org/data/hmw-dna-extraction-strategies/" target="_blank">here</a>. Detailed <b>lab protocols</b> for HMW DNA extraction methods mentioned in the study are <a href="https://merenlab.org/data/hmw-dna-extraction-strategies/protocol/" target="_blank">here</a>.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Molecular Ecology Resources</b>, 22(5):1786-1802 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=High+molecular+weight+DNA+extraction+strategies+for+long-read+sequencing+of+complex+metagenomes" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1111/1755-0998.13588" target="_blank">doi:10.1111/1755-0998.13588</a></span>
</div>


## 2020

<a id="10.1186/s13059-020-02195-w">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1186/s13059-020-02195-w"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1186/s13059-020-02195-w" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1186/s13059-020-02195-w" target="_new">Functional and genetic markers of niche partitioning among enigmatic members of the human oral microbiome</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Shaiber A</span>, Willis AD, <span class="pub-collaborator-author" title='A key collaborator of the lab at the time of publication'>Delmont TO</span>, Roux S, Chen L, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Schmid AC</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Yousef M</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Watson AR</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Lolans K</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Esen ÖC</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Lee STM</span>, Downey N, Morrison HG, Dewhirst FE, Welch JLM<sup>‡</sup>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM<sup>‡</sup></span></span>
    <span class="pub-co-first-authors"><sup>‡</sup>Co-senior authors</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/shaiber_et_al_tm7_phylogenomics.png"><img src="/images/pubs/shaiber_et_al_tm7_phylogenomics.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A multi-omics study that combines <b>genome-resolved metagenomics</b>, <b>pangenomics</b>, <b>phylogenomics</b>, and <b>microbial population genetics</b> to investigate the ecology and evolution of Saccharibacteria (TM7) in the human oral cavity, and offers a formal description of '<b>functional enrichment</b>' statistic for phylogenomics and pangenomics.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Demonstrates that TM7s split into tongue specialists and plaque specialists, and plaque TM7s group with environmental TM7s, leading to an hypothesis that <b>the dental plaque may have served as a stepping stone for environmental microbes to adapt to host environments</b> at least for some clades of microbes</span><br><span style="display: inline-block; padding-bottom: 5px;">- A news article by <a href='https://twitter.com/alie_astrocyte' target='_blank'>Alison Caldwell, PhD</a>: <a href='https://www.uchicagomedicine.org/forefront/biological-sciences-articles/microbes-in-dental-plaque' target='_blank'>Microbes in dental plaque look more like relatives in soil than those on the tongue</a>.</span><br><span style="display: inline-block; padding-bottom: 5px;">- <a href="/data/#niche-partitioning-in-the-human-oral-cavity" target="_blank">Public raw and intermediate data</a>. <a href="https://static-content.springer.com/esm/art%3A10.1186%2Fs13059-020-02195-w/MediaObjects/13059_2020_2195_MOESM13_ESM.docx" target="_blank">Reviewer comments and responses</a>.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Genome Biology</b>, 21(292) | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Functional+and+genetic+markers+of+niche+partitioning+among+enigmatic+members+of+the+human+oral+microbiome" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1186/s13059-020-02195-w" target="_blank">doi:10.1186/s13059-020-02195-w</a></span>
</div>

<a id="10.1186/s13059-020-02200-2">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1186/s13059-020-02200-2"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1186/s13059-020-02200-2" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1186/s13059-020-02200-2" target="_new">Metapangenomics of the oral microbiome provides insights into habitat adaptation and cultivar diversity</a></span>
    <span class="pub-authors">Utter DR, Borisy GG, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Cavanaugh CM, Welch JLM</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/utter_et_al_metapangenome.png"><img src="/images/pubs/utter_et_al_metapangenome.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- An application of <b>metapangenomics</b> that links the gene pool of two major oral microbial taxa, <i>Haemophilus parainfluenzae</i> and the genus <i>Rothia</i>, to their ecology using the Human Microbiome Project metageonomes generated from tongue, buccal mucosa, and dental plaque samples.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Reveals that seemingly generalist organisms are composed of cryptic subpopulations with distinct ecology that is associated with only a small number of genes.</span><br><span style="display: inline-block; padding-bottom: 5px;">- <a href="https://static-content.springer.com/esm/art%3A10.1186%2Fs13059-020-02200-2/MediaObjects/13059_2020_2200_MOESM7_ESM.pdf" target="_blank">Reviewer comments and responses</a>. <a href="https://dutter.github.io/projects/oral_metapan">Reproducible bioinformatics workflow</a>.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Genome Biology</b>, 21(293) | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Metapangenomics+of+the+oral+microbiome+provides+insights+into+habitat+adaptation+and+cultivar+diversity" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1186/s13059-020-02200-2" target="_blank">doi:10.1186/s13059-020-02200-2</a></span>
</div>

<a id="10.7554/eLife.56998">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.7554/eLife.56998"></div>
<div class="__dimensions_badge_embed__" data-doi="10.7554/eLife.56998" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.7554/eLife.56998" target="_new">Droplet-based high-throughput cultivation for accurate screening of antibiotic resistant gut microbes</a></span>
    <span class="pub-authors">Watterson WJ, Tanyeri M, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Watson AR</span>, Cham CM, Shan Y, Chang EB, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Tay S</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/watterson_et_al_droplets.jpg"><img src="/images/pubs/watterson_et_al_droplets.jpg" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- An aneorobic microfluidics platform for high-throughput cultivation of microbes that grows <b>single microbial cells in millions of picoliter droplets</b>.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Populations of gut microbes that compete poorly in plates grow well in droplets <b>regardless of culture media</b>. Furthermore, taxonomic profile of droplets resembles droplets much better than plate scrapings even at the level of oligotypes.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Demonstrates that <b>not detecting microbes in plate-based screening of antibiotic resistance may not mean that the original sample does not contain microbes resistant to antibiotics</b>.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>eLife</b>, 9(e56998) | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Droplet-based+high-throughput+cultivation+for+accurate+screening+of+antibiotic+resistant+gut+microbes" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.7554/eLife.56998" target="_blank">doi:10.7554/eLife.56998</a></span>
</div>

<a id="10.1101/gr.258640.119">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1101/gr.258640.119"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1101/gr.258640.119" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1101/gr.258640.119" target="_new">Accurate and complete genomes from metagenomes</a></span>
    <span class="pub-authors">Chen L, Anantharaman K, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Shaiber A</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM<sup>‡</sup></span>, Banfield JF<sup>‡</sup></span>
    <span class="pub-co-first-authors"><sup>‡</sup>Co-senior authors</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/chen_accurate_genomes.png"><img src="/images/pubs/chen_accurate_genomes.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A <b>review on genome-resolved metagenomics</b> that discusses the importance of using assembly and careful binning strategies to study metagenomes.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Case studies include a demonstration of how <b>single-copy core genes can fail to predict the quality of metagenome-assembled genomes</b>, and <b>automated strategies that yield tens of thousands of metagenome-assembled genomes will include extensive contamination</b>.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Promotes approaches to reconstruct <b>'complete' genomes from metagenomes</b> and the use of <b>GC skew as a metric for checking genome correctness</b>.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Tutorial on <a href="https://ggkbase-help.berkeley.edu/genome_curation/scaffold-extension-and-gap-closing/" target="_blank">scaffold extension and gap closing</a>, reproducible workflow for <a href="http://merenlab.org/data/parcubacterium-in-hbcfdna/" target="_blank">binning and phylogenomics of a Parcubacterium genome from human blood metagenomes</a>.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Genome Research</b>, 30(3):315-333 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Accurate+and+complete+genomes+from+metagenomes" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1101/gr.258640.119" target="_blank">doi:10.1101/gr.258640.119</a></span>
</div>


## 2019

<a id="10.7554/eLife.46497">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.7554/eLife.46497"></div>
<div class="__dimensions_badge_embed__" data-doi="10.7554/eLife.46497" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.7554/eLife.46497" target="_new">Single-amino acid variants reveal evolutionary processes that shape the biogeography of a global SAR11 subclade</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Delmont TO<sup>☯</sup></span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Kiefl E<sup>☯</sup></span>, Kilinc O, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Esen ÖC</span>, Uysal I, Rappé MS, Giovannoni S, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span></span>
    <span class="pub-co-first-authors"><sup>☯</sup>Co-first authors</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/delmond_and_kiefl_sar11_saavs.png"><img src="/images/pubs/delmond_and_kiefl_sar11_saavs.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A study that introduces <a href="http://merenlab.org/2015/07/20/analyzing-variability/#an-intro-to-single-nucleotidecodonamino-acid-variation" target="_blank">'single-amino acid variants'</a> (SAAVs) and demonstrates <b>the use of SAAVs to tease apart evolutionary processes that shape the biogeography and genomic heterogeneity within a SAR11 population</b> through metagenomics.</span><br><span style="display: inline-block; padding-bottom: 5px;">- A first attempt to link population genetics and the predicted protein structures to explore <i>in silico</i> <b>the intersection beetween protein biochemistry and evolutionary processes</b> acting on an environmental microbe.</span><br><span style="display: inline-block; padding-bottom: 5px;">- An application of <a href="https://peerj.com/articles/4320/" target="_blank">metapangenomics</a> to define <b>subclades of SAR11 based on gene content and ecology</b>.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Reproducible bioinformatics workflow is <a href="http://merenlab.org/data/sar11-saavs/" target="_blank">here</a>. <a href="https://doi.org/10.7554/eLife.46497.040" target="_blank">Reviewer criticism</a> and <a href="https://doi.org/10.7554/eLife.46497.041" target="_blank">our responses</a> are also available.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>eLife</b>, 8(e46497) | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Single-amino+acid+variants+reveal+evolutionary+processes+that+shape+the+biogeography+of+a+global+SAR11+subclade" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.7554/eLife.46497" target="_blank">doi:10.7554/eLife.46497</a></span>
</div>

<a id="10.1128/mBio.00725-19">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1128/mBio.00725-19"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1128/mBio.00725-19" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1128/mBio.00725-19" target="_new">Composite Metagenome-Assembled Genomes Reduce the Quality of Public Genome Repositories</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Shaiber A</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span></span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/shaiber_eren_MAG_refinement.png"><img src="/images/pubs/shaiber_eren_MAG_refinement.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A letter that stresses that the composite metagenome-assembled genomes influence phylogenomic, pangenomic, and ecological insights (<a href="../files/Shaiber_Eren_2019_reviewer_response.pdf" target="_blank">peer reviews and responses</a>).</span><br><span style="display: inline-block; padding-bottom: 5px;">- <a href="http://merenlab.org/data/refining-espinoza-mags/" target="_blank">A reproducible workflow</a> to detail the steps of genome refinement and make available the refined versions of some key genomes.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>mBio</b>, 10(3):e00725-19 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Composite+Metagenome-Assembled+Genomes+Reduce+the+Quality+of+Public+Genome+Repositories" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1128/mBio.00725-19" target="_blank">doi:10.1128/mBio.00725-19</a></span>
</div>

<a id="10.1038/s41467-019-08973-w">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/s41467-019-08973-w"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/s41467-019-08973-w" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1038/s41467-019-08973-w" target="_new">The Wolbachia mobilome in Culex pipiens includes a putative plasmid</a></span>
    <span class="pub-authors"><span class="pub-collaborator-author" title='A key collaborator of the lab at the time of publication'>Reveillaud J<sup>☯</sup></span>, Bordenstein SR<sup>☯</sup>, Cruaud C, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Shaiber A</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Esen ÖC</span>, Weill M, Makoundou P, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Lolans K</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Watson AR</span>, Rakotoarivony I, Bordenstein SR<sup>☯</sup>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span></span>
    <span class="pub-co-first-authors"><sup>☯</sup>Co-first authors</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/miscellaneous/2019-03-05-wolbachia-plasmid/minion-metagenomics.png"><img src="/images/miscellaneous/2019-03-05-wolbachia-plasmid/minion-metagenomics.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- The first report of a <i>Wolbachia</i> plasmid through genome-resolved metagenomics on microsurgically removed individual mosquito ovary samples (<a href="https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-019-08973-w/MediaObjects/41467_2019_8973_MOESM2_ESM.pdf">peer reviews and responses</a>).</span><br><span style="display: inline-block; padding-bottom: 5px;">- Yet another application of <a href="https://doi.org/10.7717/peerj.4320">metapangenomics</a> and an applicatoin of minION long-read sequencing on extremely low-biomass samples.</span><br><span style="display: inline-block; padding-bottom: 5px;">- <a href="http://merenlab.org/data/wolbachia-plasmid">Reproducible bioinformatics workflow</a> with all data items, a 'behind the paper' <a href="http://merenlab.org/2019/03/05/wolbachia-plasmid/">blog post</a> by <a href="https://twitter.com/JulieReve">Julie Reveillaud</a>, and a <a href="https://www.mbl.edu/news/mobile-dna-element-found-mosquito-parasite-has-potential-infectious-disease-control">press release</a> from the Marine Biological Laboratory.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Nature Communications</b>, 10(1):1051 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=The+Wolbachia+mobilome+in+Culex+pipiens+includes+a+putative+plasmid" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/s41467-019-08973-w" target="_blank">doi:10.1038/s41467-019-08973-w</a></span>
</div>

<a id="10.7717/peerj.7548">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.7717/peerj.7548"></div>
<div class="__dimensions_badge_embed__" data-doi="10.7717/peerj.7548" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.7717/peerj.7548" target="_new">Genome-resolved insights into a novel Spiroplasma symbiont of the Wheat Stem Sawfly (Cephus cinctus)</a></span>
    <span class="pub-authors">Yeoman CJ, Brutscher LM, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Esen ÖC</span>, Ibaoglu F, Fowler C, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Wanner K, Weaver DK</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/spiroplasma-pangenome.png"><img src="/images/pubs/spiroplasma-pangenome.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A study that uses (1) <b>genome-resolved metagenomics</b> to reconstruct a population genome from fly metagenomes that resolves to the genus <a href="https://en.wikipedia.org/wiki/Spiroplasma" target="_blank"><i>Spiroplasma</i></a>, (2) <b>pangenomics</b> to put this genome in the context of other <i>Spiroplasma</i> genomes, (3) <b>phylogenomics</b> to infer ancestral relationships between <i>Spiroplasma</i> genomes, and (4) includes an ANI-based distance estimation between all genomes for comprehensive reporting.</span><br><span style="display: inline-block; padding-bottom: 5px;">- It is a particularly good example that demonstrates how pangenomics can reveal appropriate targets for high-resolution phylogenomics.</span><br><span style="display: inline-block; padding-bottom: 5px;">- A fully reproducible bioinformatics workflow for this multi'omics analysis is <a href="http://merenlab.org/data/spiroplasma-pangenome/" target="_blank">here</a>. Anvi'o databases to interactively reproduce and explore the Spiroplasma pangenome is also <a href="http://doi.org/10.6084/m9.figshare.8201852" target="_blank">available</a>.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>PeerJ</b>, 7(e7548) | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Genome-resolved+insights+into+a+novel+Spiroplasma+symbiont+of+the+Wheat+Stem+Sawfly+(Cephus+cinctus)" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.7717/peerj.7548" target="_blank">doi:10.7717/peerj.7548</a></span>
</div>

<a id="10.1038/s41564-019-0494-6">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/s41564-019-0494-6"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/s41564-019-0494-6" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1038/s41564-019-0494-6" target="_new">Global phylogeography and ancient evolution of the widespread human gut virus crAssphage</a></span>
    <span class="pub-authors">Edwards RA, Vega AA, Norman HM, Ohaeri M, Levi K, Dinsdale EA, Cinek O, Aziz RK, McNair K, Barr JJ, Bibby K, Brouns SJJ, Cazares A, Jonge PAd, Desnues C, Muñoz SLD, Fineran PC, Kurilshikov A, Lavigne R, Mazankova K, McCarthy DT, Nobrega FL, Muñoz AR, Tapia G, Trefault N, Tyakht AV, Vinuesa P, Wagemans J, Zhernakova A, Aarestrup FM, Ahmadov G, Alassaf A, Anton J, Asangba A, Billings EK, Cantu VA, Carlton JM, Cazares D, Cho G, Condeff T, Cortés P, Cranfield M, Cuevas DA, Iglesia RDl, Decewicz P, Doane MP, Dominy NJ, Dziewit L, Elwasila BM, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Franz C, Fu J, Garcia-Aljaro C, Ghedin E, Gulino KM, Haggerty JM, Head SR, Hendriksen RS, Hill C, Hyöty H, Ilina EN, Irwin MT, Jeffries TC, Jofre J, Junge RE, Kelley ST, Mirzaei MK, Kowalewski M, Kumaresan D, Leigh SR, Lipson D, Lisitsyna ES, Llagostera M, Maritz JM, Marr LC, McCann A, Molshanski-Mor S, Monteiro S, Moreira-Grez B, Morris M, Mugisha L, Muniesa M, Neve H, Nguyen N, Nigro OD, Nilsson AS, O’Connell T, Odeh R, Oliver A, Piuri M, II AJP, Qimron U, Quan Z, Rainetova P, Ramírez-Rojas A, Raya R, Reasor K, Rice GAO, Rossi A, Santos R, Shimashita J, Stachler EN, Stene LC, Strain R, Stumpf R, Torres PJ, Twaddle A, Ibekwe MU, Villagra N, Wandro S, White B, Whiteley A, Whiteson KL, Wijmenga C, Zambrano MM, Zschach H, Dutilh BE</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/crAssphage.png"><img src="/images/pubs/crAssphage.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A monumental effort lead by <a href="https://twitter.com/linsalrob" target="_blank">Rob Edwards</a> and <a href="https://twitter.com/BEDutilh" target="_blank">Bas Dutilh</a> to explore with the help of 114 scientists from around the world the global phylogeorgraphy and evolution of <a href="https://www.nature.com/articles/ncomms5498" target="_blank">crAssphage</a>, one of the most numerous viruses in the human gut that infect bacteria.</span><br><span style="display: inline-block; padding-bottom: 5px;">- A curated list of press coverage of this study is available <a href="https://edwards.sdsu.edu/research/press-about-global-phylogeography-of-crassphage/" target="_blank">on Rob Edwards' web site</a>.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Nature Microbiology</b>, 4(10):1727 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Global+phylogeography+and+ancient+evolution+of+the+widespread+human+gut+virus+crAssphage" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/s41564-019-0494-6" target="_blank">doi:10.1038/s41564-019-0494-6</a></span>
</div>

<a id="10.1038/s41564-019-0362-4">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/s41564-019-0362-4"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/s41564-019-0362-4" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1038/s41564-019-0362-4" target="_new">Co-occurring genomic capacity for anaerobic methane and dissimilatory sulfur metabolisms discovered in the Korarchaeota</a></span>
    <span class="pub-authors">McKay LJ, Dlakić M, Fields MW, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Delmont TO</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Jay ZJ, Klingelsmith KB, Rusch DB, Inskeep WP</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/mckay_and_inskeep.jpg"><img src="/images/pubs/mckay_and_inskeep.jpg" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- Yellowstone National Park, archaeal evolution, genome-resolved metagenomics, phylogenomics, and pangenomics (cool stuff all over).</span><br><span style="display: inline-block; padding-bottom: 5px;">- First description of a microbial population with both anaerobic methane and dissimilatory sulfur metabolisms.</span><br><span style="display: inline-block; padding-bottom: 5px;">- A news article by Montana State University with the photographs of cools scientists: <a href='http://www.montana.edu/news/18471'>'An organism in Yellowstone hot spring potentially linked to earliest life on Earth'</a></span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Nature Microbiology</b>, 4(4):614-622 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Co-occurring+genomic+capacity+for+anaerobic+methane+and+dissimilatory+sulfur+metabolisms+discovered+in+the+Korarchaeota" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/s41564-019-0362-4" target="_blank">doi:10.1038/s41564-019-0362-4</a></span>
</div>

<a id="10.1038/s41422-018-0127-2">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/s41422-018-0127-2"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/s41422-018-0127-2" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1038/s41422-018-0127-2" target="_new">Transcriptome-wide reprogramming of N 6-methyladenosine modification by the mouse microbiome</a></span>
    <span class="pub-authors">Wang X, Li Y, Chen W, Shi H, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Morozov A, He C, Luo G, Pan T</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/wang_et_al_m6A.png"><img src="/images/pubs/wang_et_al_m6A.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- N<sup>6</sup>-methyladenosine (m<sup>6</sup>A) is the most abundant mesenger RNA modification in mammalian cells, occurring at ~3 modified adenosine residues per transcript.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Using liquid chromatography/mass spectrometry, this study shows differential occurrence of m<sup>6</sup>A modifications in brain, liver, and intestinal cells between germ-free and conventional mice, demonstrating that the microbiome has a strong effect on host m<sup>6</sup>A mRNA modification.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Cell Research</b>, 29:167–170 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Transcriptome-wide+reprogramming+of+N+6-methyladenosine+modification+by+the+mouse+microbiome" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/s41422-018-0127-2" target="_blank">doi:10.1038/s41422-018-0127-2</a></span>
</div>

<a id="10.1126/scitranslmed.aau9356">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1126/scitranslmed.aau9356"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1126/scitranslmed.aau9356" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1126/scitranslmed.aau9356" target="_new">B cell superantigens in the human intestinal microbiota</a></span>
    <span class="pub-authors">Bunker JJ, Drees C, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Watson AR</span>, Plunkett CH, Nagler CR, Schneewind O, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Bendelac A</span>
    <span class="pub-journal"> 📚 <b>Science Translational Medicine</b>, 11(507):eaau9356 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=B+cell+superantigens+in+the+human+intestinal+microbiota" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1126/scitranslmed.aau9356" target="_blank">doi:10.1126/scitranslmed.aau9356</a></span>
</div>


## 2018

<a id="10.1038/s41467-018-07675-z">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/s41467-018-07675-z"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/s41467-018-07675-z" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1038/s41467-018-07675-z" target="_new">Microbiome characterization by high-throughput transfer RNA sequencing and modification analysis</a></span>
    <span class="pub-authors">Schwartz MH, Wang H, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Pan JN</span>, Clark WC, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Cui S</span>, Eckwahl MJ, Pan DW, Parisien M, Owens SM, Cheng BL, Martinez K, Xu J, Chang EB, Pan T<sup>‡</sup>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM<sup>‡</sup></span></span>
    <span class="pub-co-first-authors"><sup>‡</sup>Co-senior authors</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/schwartz_et_al_trna_seq.png"><img src="/images/pubs/schwartz_et_al_trna_seq.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- The first application of tRNA sequencing to environmental microbiomes (<a href="https://static-content.springer.com/esm/art%3A10.1038%2Fs41467-018-07675-z/MediaObjects/41467_2018_7675_MOESM2_ESM.pdf">peer reviews and responses</a>).</span><br><span style="display: inline-block; padding-bottom: 5px;">- Reveals taxon- and diet-dependent variations in tRNA modifications, and provides first <i>in situ</i> insights into 'metaepitranscriptomics' through tRNA gene expression dynamics and post-transcriptional modifications.</span><br><span style="display: inline-block; padding-bottom: 5px;">- <a href="https://www.uchicagomedicine.org/forefront/microbiome-articles/2018/december/new-rna-sequencing-strategy-provides-insight-into-microbiomes">New RNA sequencing strategy provides insight into microbiomes</a>, by <b>Matt Wood</b>.</span><br><span style="display: inline-block; padding-bottom: 5px;">- <a href="https://www.news-medical.net/news/20181217/RNA-sequencing-offers-novel-insights-into-the-microbiome.aspx">RNA sequencing offers novel insights into the microbiome</a>, by <b>Liji Thomas</b>, MD, and <b>Kate Anderton</b>, B.Sc.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Nature Communications</b>, 9(1):5353 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Microbiome+characterization+by+high-throughput+transfer+RNA+sequencing+and+modification+analysis" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/s41467-018-07675-z" target="_blank">doi:10.1038/s41467-018-07675-z</a></span>
</div>

<a id="10.1038/s41564-018-0176-9">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/s41564-018-0176-9"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/s41564-018-0176-9" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1038/s41564-018-0176-9" target="_new">Nitrogen-fixing populations of Planctomycetes and Proteobacteria are abundant in surface ocean metagenomes</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Delmont TO</span>, Quince C, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Shaiber A</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Esen ÖC</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Lee STM</span>, Rappé MS, McLellan SL, Lücker S, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span></span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/delmont_et_al_HBDs.png"><img src="/images/pubs/delmont_et_al_HBDs.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- First genomic evidence for abundant and widespread non-cyanobacterial nitrogen-fixing populations in the surface ocean.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Nearly 1,000 non-redundant, high-quality bacterial, archaeal, and eukaryotic population genomes from TARA Oceans metagenomes.</span><br><span style="display: inline-block; padding-bottom: 5px;">- A "behind the paper" <a href="https://naturemicrobiologycommunity.nature.com/users/113363-a-murat-eren-meren/posts/34040-microbiologists-vs-shotgun-metagenomes-surface-ocean" target="_blank">blog post</a> by Meren, a <a href="https://www.mbl.edu/news/large-scale-study-indicates-novel-abundant-nitrogen-fixing-microbes-surface-ocean" target="_blank">press release</a> by the MBL, and an extensive description of <a href="http://merenlab.org/data/2017_Delmont_et_al_HBDs/" target="_blank">the bioinformatics workflow</a>.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Nature Microbiology</b>, 3(7):804-813 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Nitrogen-fixing+populations+of+Planctomycetes+and+Proteobacteria+are+abundant+in+surface+ocean+metagenomes" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/s41564-018-0176-9" target="_blank">doi:10.1038/s41564-018-0176-9</a></span>
</div>

<a id="10.7717/peerj.4320">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.7717/peerj.4320"></div>
<div class="__dimensions_badge_embed__" data-doi="10.7717/peerj.4320" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.7717/peerj.4320" target="_new">Linking pangenomes and metagenomes -- the Prochlorococcus metapangenome</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Delmont TO</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span></span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/delmont_eren_metapangenomics.jpeg"><img src="/images/pubs/delmont_eren_metapangenomics.jpeg" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A big-data study in which a pangenome of 31 <i>Prochlorococcus</i> isolates meets 31 billion Tara Oceans metagenomic sequences (<a href="https://peerj.com/articles/4320/reviews/">Peer-review history</a>).</span><br><span style="display: inline-block; padding-bottom: 5px;">- Metapangenomes reveal to what extent genes that may be linked to the ecology and fitness of microbes are conserved within a phylogenetic clade.</span><br><span style="display: inline-block; padding-bottom: 5px;">- <a href="http://merenlab.org/data/prochlorococcus-metapangenome" target="_blank">Reproducible bioinformatics workflow</a>.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>PeerJ</b>, 6(e4320) | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Linking+pangenomes+and+metagenomes+--+the+Prochlorococcus+metapangenome" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.7717/peerj.4320" target="_blank">doi:10.7717/peerj.4320</a></span>
</div>

<a id="10.1186/s40168-018-0474-8">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1186/s40168-018-0474-8"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1186/s40168-018-0474-8" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1186/s40168-018-0474-8" target="_new">Gut microbes contribute to variation in solid organ transplant outcomes in mice</a></span>
    <span class="pub-authors">McIntosh CM, Chen L, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Shaiber A</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM<sup>‡</sup></span>, Alegre M<sup>‡</sup></span>
    <span class="pub-co-first-authors"><sup>‡</sup>Co-senior authors</span>
    <div class="pub-info-no-image">
    <div class="pub-highlights-no-image">
    <span style="display: inline-block; padding-bottom: 5px;">- Two groups of genetically identical mice with different microbial community structures differ in their response to skin transplants from genetically identical sources: one group rejects the skin faster than the other.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Transferring poop from slow-rejecting mice into fast-rejecting mice turns fast-rejecting mice into slow-rejecting mice.</span><br><span style="display: inline-block; padding-bottom: 5px;">- These suggest that differences in 'resident microbiome' in healthy individuals <i>may</i> contribute to interpersonal variability in graft outcomes, and fecal microbiota transplantation <i>may</i> play a therapeutic role to reverse that phenotype.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Basically, "we do not have a mechanistic understanding, but this is definitely very interesting".</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Microbiome</b>, 6(96) | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Gut+microbes+contribute+to+variation+in+solid+organ+transplant+outcomes+in+mice" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1186/s40168-018-0474-8" target="_blank">doi:10.1186/s40168-018-0474-8</a></span>
</div>

<a id="10.1038/s41586-018-0125-z">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/s41586-018-0125-z"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/s41586-018-0125-z" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1038/s41586-018-0125-z" target="_new">Microbial signals drive pre-leukaemic myeloproliferation in a Tet2-deficient host</a></span>
    <span class="pub-authors">Meisel M, Hinterleitner R, Pacis A, Chen L, Earley ZM, Mayassi T, Pierre JF, Ernest JD, Galipeau HJ, Thuille N, Bouziat R, Buscarlet M, Ringus DL, Wang Y, Li Y, Dinh V, Kim SM, McDonald BD, Zurenski MA, Musch MW, Furtado GC, Lira SA, Baier G, Chang EB, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Weber CR, Busque L, Godley LA, Verdú EF, Barreiro LB, Jabri B</span>
    <div class="pub-info-no-image">
    <div class="pub-highlights-no-image">
    <span style="display: inline-block; padding-bottom: 5px;">- TET2 deficiency ➡️  deteriorating small intestinal barrier ➡️  bacterial translocation ➡️  incrased IL-6 signalling ➡️  pre-leukemic myeloproliferation (a leukemia precursor).</span><br><span style="display: inline-block; padding-bottom: 5px;">- ScienceDaily: <a href='https://www.sciencedaily.com/releases/2018/05/180516131226.htm' target='_blank'>Under certain conditions, bacterial signals set the stage for leukemia</a>.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Nature</b>, 557(7706):580–584 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Microbial+signals+drive+pre-leukaemic+myeloproliferation+in+a+Tet2-deficient+host" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/s41586-018-0125-z" target="_blank">doi:10.1038/s41586-018-0125-z</a></span>
</div>


## 2017

<a id="10.1186/s40168-017-0270-x">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1186/s40168-017-0270-x"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1186/s40168-017-0270-x" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1186/s40168-017-0270-x" target="_new">Tracking microbial colonization in fecal microbiota transplantation experiments via genome-resolved metagenomics</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Lee STM<sup>☯</sup></span>, Kahn SA<sup>☯</sup>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Delmont TO</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Shaiber A</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Esen ÖC</span>, Hubert NA, Morrison HG, Antonopoulos DA, Rubin DT, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span></span>
    <span class="pub-co-first-authors"><sup>☯</sup>Co-first authors</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/lee_fmt.png"><img src="/images/pubs/lee_fmt.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- An FMT study with metagenome-assembled genomes (see <a href="http://merenlab.org/data/#lee-et-al-2017--fecal-transplants" target="_blank">public data</a>).</span><br><span style="display: inline-block; padding-bottom: 5px;">- Bacteroidales: high-colonization rate. Clostridiales: low colonization rate. Colonization success is negatively correlated with the number of genes related to sporulation.</span><br><span style="display: inline-block; padding-bottom: 5px;">- MAGs with the same taxonomy showed different colonization properties, highlighting the importance of high-resolution analyses.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Populations colonized both recipients were also prevalent in the HMP cohort (and the ones that did not, distribute sporadically across the HMP cohort).</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Microbiome</b>, 5(50) | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Tracking+microbial+colonization+in+fecal+microbiota+transplantation+experiments+via+genome-resolved+metagenomics" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1186/s40168-017-0270-x" target="_blank">doi:10.1186/s40168-017-0270-x</a></span>
</div>

<a id="10.1073/pnas.1712186114">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1073/pnas.1712186114"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1073/pnas.1712186114" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1073/pnas.1712186114" target="_new">Simulations predict microbial responses in the environment? This environment disagrees retrospectively</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Delmont TO</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span></span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="http://merenlab.org/images/miscellaneous/2017-10-16-reply-to-probst-et-al/Figure_01.png"><img src="http://merenlab.org/images/miscellaneous/2017-10-16-reply-to-probst-et-al/Figure_01.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A letter that re-analyzes some of the findings published in <a href="http://www.pnas.org/content/114/28/7432.abstract" target="_blank">Hu et al.</a></span><br><span style="display: inline-block; padding-bottom: 5px;">- Here is a <a href="http://www.pnas.org/content/early/2017/10/09/1712466114.extract" target="_blank">rebuttal from Probst et al.</a> challenging our findings in this letter.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Here is <a href="http://merenlab.org/2017/10/16/reply-to-probst-et-al/" target="_blank">our response to Probst et al.</a>, and <a href="http://merenlab.org/2017/11/25/DWH-O-desum-v2/" target="_blank">the recovery of DWH O. Desum v2</a>.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Proceedings of the National Academy of Sciences</b>, 114(43):E8947-E8949 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Simulations+predict+microbial+responses+in+the+environment?+This+environment+disagrees+retrospectively" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1073/pnas.1712186114" target="_blank">doi:10.1073/pnas.1712186114</a></span>
</div>

<a id="10.1038/s41467-017-01228-6">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/s41467-017-01228-6"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/s41467-017-01228-6" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1038/s41467-017-01228-6" target="_new">Genomic variation in microbial populations inhabiting the marine subseafloor at deep-sea hydrothermal vents</a></span>
    <span class="pub-authors">Anderson RE, <span class="pub-collaborator-author" title='A key collaborator of the lab at the time of publication'>Reveillaud J</span>, Reddington E, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Delmont TO</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, McDermott JM, Seewald JS, Huber JA</span>
    <span class="pub-journal"> 📚 <b>Nature Communications</b>, 8(1):1114 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Genomic+variation+in+microbial+populations+inhabiting+the+marine+subseafloor+at+deep-sea+hydrothermal+vents" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/s41467-017-01228-6" target="_blank">doi:10.1038/s41467-017-01228-6</a></span>
</div>

<a id="10.1186/s13059-017-1309-9">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1186/s13059-017-1309-9"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1186/s13059-017-1309-9" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1186/s13059-017-1309-9" target="_new">DESMAN: a new tool for de novo extraction of strains from metagenomes</a></span>
    <span class="pub-authors">Quince C, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Delmont TO</span>, Raguideau S, Alneberg J, Darling AE, Collins G, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span></span>
    <span class="pub-journal"> 📚 <b>Genome Biology</b>, 18(1):181 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=DESMAN:+a+new+tool+for+de+novo+extraction+of+strains+from+metagenomes" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1186/s13059-017-1309-9" target="_blank">doi:10.1186/s13059-017-1309-9</a></span>
</div>

<a id="10.1038/nbt.3893">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/nbt.3893"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/nbt.3893" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1038/nbt.3893" target="_new">Minimum information about a single amplified genome (MISAG) and a metagenome-assembled genome (MIMAG) of bacteria and archaea</a></span>
    <span class="pub-authors">Bowers RM, Kyrpides NC, Stepanauskas R, Harmon-Smith M, Doud D, Reddy TBK, Schulz F, Jarett J, Rivers AR, Eloe-Fadrosh EA, Tringe SG, Ivanova NN, Copeland A, Clum A, Becraft ED, Malmstrom RR, Birren B, Podar M, Bork P, Weinstock GM, Garrity GM, Dodsworth JA, Yooseph S, Sutton G, Glöckner FO, Gilbert JA, Nelson WC, Hallam SJ, Jungbluth SP, Ettema TJG, Tighe S, Konstantinidis KT, Liu W, Baker BJ, Rattei T, Eisen JA, Hedlund B, McMahon KD, Fierer N, Knight R, Finn R, Cochrane G, Karsch-Mizrachi I, Tyson GW, Rinke C, Consortium TGS, Lapidus A, Meyer F, Yilmaz P, Parks DH, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Schriml L, Banfield JF, Hugenholtz P, Woyke T</span>
    <span class="pub-journal"> 📚 <b>Nature biotechnology</b>, 35(8):725 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Minimum+information+about+a+single+amplified+genome+(MISAG)+and+a+metagenome-assembled+genome+(MIMAG)+of+bacteria+and+archaea" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/nbt.3893" target="_blank">doi:10.1038/nbt.3893</a></span>
</div>

<a id="10.1016/j.celrep.2017.06.060">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1016/j.celrep.2017.06.060"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1016/j.celrep.2017.06.060" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1016/j.celrep.2017.06.060" target="_new">Peripartum antibiotics promote gut dysbiosis, loss of immune tolerance, and inflammatory bowel disease in genetically prone offspring</a></span>
    <span class="pub-authors">Miyoshi J, Bobe AM, Miyoshi S, Huang Y, Hubert N, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Delmont TO</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Leone V, Chang EB</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/miyoshi_et_al_anvio_16S.png"><img src="/images/pubs/miyoshi_et_al_anvio_16S.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- Antibiotics during pregnancy promote offspring gut dysbiosis, immune dysfunction, and IBD.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Antibiotics given after the developmental period do not increase IBD.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Antibiotic-perturbed maternal microbiota likely contribute to neonatal gut dysbiosis.</span><br><span style="display: inline-block; padding-bottom: 5px;">- <a href="https://sciencelife.uchospitals.edu/2017/07/11/antibiotics-taken-late-in-pregnancy-can-increase-risk-for-inflammatory-bowel-diseases-in-offspring/" target="_blank">Press release</a>.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Cell Reports</b>, 20(2):491-504 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Peripartum+antibiotics+promote+gut+dysbiosis,+loss+of+immune+tolerance,+and+inflammatory+bowel+disease+in+genetically+prone+offspring" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1016/j.celrep.2017.06.060" target="_blank">doi:10.1016/j.celrep.2017.06.060</a></span>
</div>


## 2016

<a id="10.7717/peerj.1839">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.7717/peerj.1839"></div>
<div class="__dimensions_badge_embed__" data-doi="10.7717/peerj.1839" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.7717/peerj.1839" target="_new">Identifying contamination with advanced visualization and analysis practices: metagenomic approaches for eukaryotic genome assemblies</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Delmont TO</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span></span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/delmont_tardigrade.png"><img src="/images/pubs/delmont_tardigrade.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- A holistic approach to visualize and curate genomic and metagenomic assemblies.</span><br><span style="display: inline-block; padding-bottom: 5px;">- A re-analysis of the first released Tardigrade genome reveals a likely symbiont among other contaminants.</span><br><span style="display: inline-block; padding-bottom: 5px;">- A practical approach to estimate the number bacterial genomes in an assembly.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>PeerJ</b>, 4:e1839 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Identifying+contamination+with+advanced+visualization+and+analysis+practices:+metagenomic+approaches+for+eukaryotic+genome+assemblies" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.7717/peerj.1839" target="_blank">doi:10.7717/peerj.1839</a></span>
</div>

<a id="10.1128/mBio.01713-16">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1128/mBio.01713-16"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1128/mBio.01713-16" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1128/mBio.01713-16" target="_new">Patient-specific bacteroides genome variants in pouchitis</a></span>
    <span class="pub-authors">Vineis JH, Ringus DL, Morrison HG, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Delmont TO</span>, Dalal S, Raffals LH, Antonopoulos DA, Rubin DT, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Chang EB, Sogin ML</span>
    <span class="pub-journal"> 📚 <b>mBio</b>, 7(6):e01713-16 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Patient-specific+bacteroides+genome+variants+in+pouchitis" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1128/mBio.01713-16" target="_blank">doi:10.1128/mBio.01713-16</a></span>
</div>

<a id="10.3389/fmicb.2016.01318">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.3389/fmicb.2016.01318"></div>
<div class="__dimensions_badge_embed__" data-doi="10.3389/fmicb.2016.01318" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.3389/fmicb.2016.01318" target="_new">New insights into microbial ecology through subtle nucleotide variation</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Sogin ML, Maignien L</span>
    <span class="pub-journal"> 📚 <b>Frontiers in microbiology</b>, 7:1318 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=New+insights+into+microbial+ecology+through+subtle+nucleotide+variation" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.3389/fmicb.2016.01318" target="_blank">doi:10.3389/fmicb.2016.01318</a></span>
</div>

<a id="10.1093/femsre/fuw017">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1093/femsre/fuw017"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1093/femsre/fuw017" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1093/femsre/fuw017" target="_new">Millions of reads, thousands of taxa: microbial community structure and associations analyzed via marker genes</a></span>
    <span class="pub-authors">Bálint M, Bahram M, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Faust K, Fuhrman JA, Lindahl B, O'Hara RB, Öpik M, Sogin ML, Unterseher M, Tedersoo L</span>
    <span class="pub-journal"> 📚 <b>FEMS Microbiology Reviews</b>, 40(5):686-700 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Millions+of+reads,+thousands+of+taxa:+microbial+community+structure+and+associations+analyzed+via+marker+genes" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1093/femsre/fuw017" target="_blank">doi:10.1093/femsre/fuw017</a></span>
</div>


## 2015

<a id="10.7717/peerj.1319">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.7717/peerj.1319"></div>
<div class="__dimensions_badge_embed__" data-doi="10.7717/peerj.1319" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.7717/peerj.1319" target="_new">Anvi'o: an advanced analysis and visualization platform for ‘omics data</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Esen ÖC</span>, Quince C, Vineis JH, Morrison HG, Sogin ML, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Delmont TO</span></span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/eren_et_al_anvio.jpeg"><img src="/images/pubs/eren_et_al_anvio.jpeg" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- The very first description of <a href="http://anvio.org/" target="_blank">anvi'o</a>.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Demonstrating its abilities in genome-resolved metagenomics, and microbial populatio genetics via single-nucleotide variant analyses across metagenomes.</span><br><span style="display: inline-block; padding-bottom: 5px;">- Re-analysis of cultivar genomes, metagenomes, and metatranscriptomes associated with the Deepwater Horizon oil spill.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>PeerJ</b>, 3:e1319 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Anvi'o:+an+advanced+analysis+and+visualization+platform+for+‘omics+data" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.7717/peerj.1319" target="_blank">doi:10.7717/peerj.1319</a></span>
</div>

<a id="10.1038/ismej.2014.195">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/ismej.2014.195"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/ismej.2014.195" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1038/ismej.2014.195" target="_new">Minimum entropy decomposition: unsupervised oligotyping for sensitive partitioning of high-throughput marker gene sequences</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Morrison HG, Lescault PJ, <span class="pub-collaborator-author" title='A key collaborator of the lab at the time of publication'>Reveillaud J</span>, Vineis JH, Sogin ML</span>
    <span class="pub-journal"> 📚 <b>The ISME Journal</b>, 9(4):968 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Minimum+entropy+decomposition:+unsupervised+oligotyping+for+sensitive+partitioning+of+high-throughput+marker+gene+sequences" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/ismej.2014.195" target="_blank">doi:10.1038/ismej.2014.195</a></span>
</div>

<a id="10.1038/ismej.2014.97">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/ismej.2014.97"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/ismej.2014.97" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1038/ismej.2014.97" target="_new">A single genus in the gut microbiome reflects host preference and specificity</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Sogin ML, Morrison HG, Vineis JH, Fisher JC, Newton RJ, McLellan SL</span>
    <span class="pub-journal"> 📚 <b>The ISME Journal</b>, 9(1):90 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=A+single+genus+in+the+gut+microbiome+reflects+host+preference+and+specificity" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/ismej.2014.97" target="_blank">doi:10.1038/ismej.2014.97</a></span>
</div>

<a id="10.3389/fmicb.2015.00358">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.3389/fmicb.2015.00358"></div>
<div class="__dimensions_badge_embed__" data-doi="10.3389/fmicb.2015.00358" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.3389/fmicb.2015.00358" target="_new">Reconstructing rare soil microbial genomes using in situ enrichments and metagenomics</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Delmont TO</span>, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Maccario L, Prestat E, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Esen ÖC</span>, Pelletier E, Paslier DL, Simonet P, Vogel TM</span>
    <div class="pub-info">
    <div class="pub-featured-image">
    <a href="/images/pubs/delmont_rare_soil_genomes.png"><img src="/images/pubs/delmont_rare_soil_genomes.png" style="max-width: 100px; max-height: 80px; width: auto; border: none; height: auto; margin: 0 auto; display: block; transform: translateY(15%);"/></a>
    </div>
    <div class="pub-highlights">
    <span style="display: inline-block; padding-bottom: 5px;">- Application of <i>in situ</i> enrichments to shape the diversity of complex metagenomes prior to sequencing.</span><br><span style="display: inline-block; padding-bottom: 5px;">- The first successful attempt to assemble multiple near-complete bacterial genomes directly from a soil sample.</span><br><span style="display: inline-block; padding-bottom: 5px;">- The abundance of reconstructed genomes ranged from rare (<0.0001%) to relatively abundant (>0.01%) in pristine soil metagenomes.</span>
    </div>
    </div>
    <span class="pub-journal"> 📚 <b>Frontiers in microbiology</b>, 6:358 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Reconstructing+rare+soil+microbial+genomes+using+in+situ+enrichments+and+metagenomics" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.3389/fmicb.2015.00358" target="_blank">doi:10.3389/fmicb.2015.00358</a></span>
</div>

<a id="10.1038/ajg.2015.357">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/ajg.2015.357"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/ajg.2015.357" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1038/ajg.2015.357" target="_new">Extensive modulation of the fecal metagenome in children with Crohn’s disease during exclusive enteral nutrition</a></span>
    <span class="pub-authors">Quince C, Ijaz UZ, Loman N, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Saulnier D, Russell J, Haig SJ, Calus ST, Quick J, Barclay A, Bertz M, Blaut M, Hansen R, McGrogan P, Russell RK, Edwards CA, Gerasimidis K</span>
    <span class="pub-journal"> 📚 <b>The American journal of gastroenterology</b>, 110(12):1718 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Extensive+modulation+of+the+fecal+metagenome+in+children+with+Crohn’s+disease+during+exclusive+enteral+nutrition" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/ajg.2015.357" target="_blank">doi:10.1038/ajg.2015.357</a></span>
</div>

<a id="10.1128/mBio.02574-14">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1128/mBio.02574-14"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1128/mBio.02574-14" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1128/mBio.02574-14" target="_new">Sewage reflects the microbiomes of human populations</a></span>
    <span class="pub-authors">Newton RJ, McLellan SL, Dila DK, Vineis JH, Morrison HG, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Sogin ML</span>
    <span class="pub-journal"> 📚 <b>mBio</b>, 6(2):e02574-14 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Sewage+reflects+the+microbiomes+of+human+populations" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1128/mBio.02574-14" target="_blank">doi:10.1128/mBio.02574-14</a></span>
</div>


## 2014

<a id="10.1073/pnas.1409644111">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1073/pnas.1409644111"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1073/pnas.1409644111" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1073/pnas.1409644111" target="_new">Oligotyping analysis of the human oral microbiome</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Borisy GG, Huse SM, Welch aJLM</span>
    <span class="pub-journal"> 📚 <b>Proceedings of the National Academy of Sciences</b>, 111(28):E2875-E2884 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Oligotyping+analysis+of+the+human+oral+microbiome" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1073/pnas.1409644111" target="_blank">doi:10.1073/pnas.1409644111</a></span>
</div>

<a id="10.1016/j.tim.2014.08.002">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1016/j.tim.2014.08.002"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1016/j.tim.2014.08.002" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1016/j.tim.2014.08.002" target="_new">Discovering new indicators of fecal pollution</a></span>
    <span class="pub-authors">McLellan SL, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span></span>
    <span class="pub-journal"> 📚 <b>Trends in Microbiology</b>, 22(12):697-706 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Discovering+new+indicators+of+fecal+pollution" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1016/j.tim.2014.08.002" target="_blank">doi:10.1016/j.tim.2014.08.002</a></span>
</div>

<a id="10.3389/fmicb.2014.00568">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.3389/fmicb.2014.00568"></div>
<div class="__dimensions_badge_embed__" data-doi="10.3389/fmicb.2014.00568" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.3389/fmicb.2014.00568" target="_new">Dynamics of tongue microbial communities with single-nucleotide resolution using oligotyping</a></span>
    <span class="pub-authors">Welch JLM, Utter DR, Rossetti BJ, Welch DBM, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Borisy GG</span>
    <span class="pub-journal"> 📚 <b>Frontiers in microbiology</b>, 5:568 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Dynamics+of+tongue+microbial+communities+with+single-nucleotide+resolution+using+oligotyping" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.3389/fmicb.2014.00568" target="_blank">doi:10.3389/fmicb.2014.00568</a></span>
</div>

<a id="10.1371/journal.pone.0090785">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1371/journal.pone.0090785"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1371/journal.pone.0090785" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1371/journal.pone.0090785" target="_new">Humpback whale populations share a core skin bacterial community: towards a health index for marine mammals?</a></span>
    <span class="pub-authors">Apprill A, Robbins J, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Pack AA, <span class="pub-collaborator-author" title='A key collaborator of the lab at the time of publication'>Reveillaud J</span>, Mattila D, Moore M, Niemeyer M, Moore KMT, Mincer TJ</span>
    <span class="pub-journal"> 📚 <b>PLoS One</b>, 9(3):e90785 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Humpback+whale+populations+share+a+core+skin+bacterial+community:+towards+a+health+index+for+marine+mammals?" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1371/journal.pone.0090785" target="_blank">doi:10.1371/journal.pone.0090785</a></span>
</div>

<a id="10.1038/ismej.2013.227">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1038/ismej.2013.227"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1038/ismej.2013.227" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1038/ismej.2013.227" target="_new">Host-specificity among abundant and rare taxa in the sponge microbiome</a></span>
    <span class="pub-authors"><span class="pub-collaborator-author" title='A key collaborator of the lab at the time of publication'>Reveillaud J</span>, Maignien L, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Huber JA, Apprill A, Sogin ML, Vanreusel A</span>
    <span class="pub-journal"> 📚 <b>The ISME Journal</b>, 8(6):1198-209 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Host-specificity+among+abundant+and+rare+taxa+in+the+sponge+microbiome" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1038/ismej.2013.227" target="_blank">doi:10.1038/ismej.2013.227</a></span>
</div>

<a id="10.1128/mBio.00682-13">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1128/mBio.00682-13"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1128/mBio.00682-13" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1128/mBio.00682-13" target="_new">Ecological succession and stochastic variation in the assembly of Arabidopsis thaliana phyllosphere communities</a></span>
    <span class="pub-authors">Maignien L, DeForce EA, Chafee ME, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Simmons SL</span>
    <span class="pub-journal"> 📚 <b>mBio</b>, 5(1):e00682-13 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Ecological+succession+and+stochastic+variation+in+the+assembly+of+Arabidopsis+thaliana+phyllosphere+communities" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1128/mBio.00682-13" target="_blank">doi:10.1128/mBio.00682-13</a></span>
</div>


## 2013

<a id="10.1111/2041-210X.12114">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1111/2041-210X.12114"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1111/2041-210X.12114" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1111/2041-210X.12114" target="_new">Oligotyping: differentiating between closely related microbial taxa using 16S rRNA gene data</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Maignien L, Sul WJ, Murphy LG, Grim SL, Morrison HG, Sogin ML</span>
    <span class="pub-journal"> 📚 <b>Methods in Ecology and Evolution</b>, 4(12):1111-1119 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Oligotyping:+differentiating+between+closely+related+microbial+taxa+using+16S+rRNA+gene+data" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1111/2041-210X.12114" target="_blank">doi:10.1111/2041-210X.12114</a></span>
</div>

<a id="10.1371/journal.pone.0066643">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1371/journal.pone.0066643"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1371/journal.pone.0066643" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1371/journal.pone.0066643" target="_new">A filtering method to generate high quality short reads using Illumina paired-end technology</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Vineis JH, Morrison HG, Sogin ML</span>
    <span class="pub-journal"> 📚 <b>PLOS ONE</b>, 8(6):e66643 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=A+filtering+method+to+generate+high+quality+short+reads+using+Illumina+paired-end+technology" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1371/journal.pone.0066643" target="_blank">doi:10.1371/journal.pone.0066643</a></span>
</div>

<a id="10.1093/bib/bbt010">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1093/bib/bbt010"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1093/bib/bbt010" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1093/bib/bbt010" target="_new">DRISEE overestimates errors in metagenomic sequencing data</a></span>
    <span class="pub-authors"><span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Morrison HG, Huse SM, Sogin ML</span>
    <span class="pub-journal"> 📚 <b>Briefings in Bioinformatics</b>, 15(5):783-787 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=DRISEE+overestimates+errors+in+metagenomic+sequencing+data" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1093/bib/bbt010" target="_blank">doi:10.1093/bib/bbt010</a></span>
</div>

<a id="10.1111/1462-2920.12092">&nbsp;</a>
<div class="pub">
<div class='altmetric-embed' data-badge-type='donut' data-doi="10.1111/1462-2920.12092"></div>
<div class="__dimensions_badge_embed__" data-doi="10.1111/1462-2920.12092" data-hide-zero-citations="true" data-legend="hover-bottom" data-style="small_circle"></div>
    <span class="pub-title"><a href="https://doi.org/10.1111/1462-2920.12092" target="_new">Sewage reflects the distribution of human faecal Lachnospiraceae</a></span>
    <span class="pub-authors">McLellan SL, Newton RJ, Vandewalle JL, Shanks OC, Huse SM, <span class="pub-member-author" title='An official member of the lab at the time of publication'>Eren AM</span>, Sogin ML</span>
    <span class="pub-journal"> 📚 <b>Environmental Microbiology</b>, 15(8):2213-2227 | 🔍 <a href="http://scholar.google.com/scholar?hl=en&q=Sewage+reflects+the+distribution+of+human+faecal+Lachnospiraceae" target="_blank">Google Scholar</a> | 🔗 <a href="https://doi.org/10.1111/1462-2920.12092" target="_blank">doi:10.1111/1462-2920.12092</a></span>
</div>


